{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9d17e79b-d19a-4e8a-ab49-b66c96fa2fb0",
   "metadata": {},
   "source": [
    "# Comparing the HIT to GENCODE v47"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dde89d66-0df0-4452-9af2-09574f0e9407",
   "metadata": {},
   "source": [
    "Assigning HIT transcripts into categories based on their annotation status."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "937aafd0-26db-44ed-b07b-253decd20dfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/jferrer/jmidgley/annotation_project/lib64/python3.9/site-packages/sorted_nearest/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from collections import defaultdict\n",
    "import pyranges as pr\n",
    "import matplotlib as mpl\n",
    "from collections import defaultdict\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2cd93cdb-dc72-4ff6-b173-ecd99166ad75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11121</td>\n",
       "      <td>11211</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00004248723.1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00004248735.1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13453</td>\n",
       "      <td>14413</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00004248730.1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11125</td>\n",
       "      <td>11211</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832825.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00004248721.1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155000</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00001434974.2</td>\n",
       "      <td>3</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000354665.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155001</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00001544476.1</td>\n",
       "      <td>3</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155002</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00001436074.2</td>\n",
       "      <td>3</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000354554.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155003</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>3</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155004</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>3</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2155005 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source feature  start    end score strand frame  \\\n",
       "0          chr1   HAVANA    exon  11121  11211     .      +     .   \n",
       "1          chr1   HAVANA    exon  12010  12227     .      +     .   \n",
       "2          chr1   HAVANA    exon  12613  12721     .      +     .   \n",
       "3          chr1   HAVANA    exon  13453  14413     .      +     .   \n",
       "4          chr1   HAVANA    exon  11125  11211     .      +     .   \n",
       "...         ...      ...     ...    ...    ...   ...    ...   ...   \n",
       "2155000    chrM  ENSEMBL    exon  14149  14673     .      -     .   \n",
       "2155001    chrM  ENSEMBL    exon  14674  14742     .      -     .   \n",
       "2155002    chrM  ENSEMBL    exon  14747  15887     .      +     .   \n",
       "2155003    chrM  ENSEMBL    exon  15888  15953     .      +     .   \n",
       "2155004    chrM  ENSEMBL    exon  15956  16023     .      -     .   \n",
       "\n",
       "                   gene_id      transcript_id  ...            exon_id level  \\\n",
       "0        ENSG00000290825.2  ENST00000832824.1  ...  ENSE00004248723.1     2   \n",
       "1        ENSG00000290825.2  ENST00000832824.1  ...  ENSE00004248735.1     2   \n",
       "2        ENSG00000290825.2  ENST00000832824.1  ...  ENSE00003582793.1     2   \n",
       "3        ENSG00000290825.2  ENST00000832824.1  ...  ENSE00004248730.1     2   \n",
       "4        ENSG00000290825.2  ENST00000832825.1  ...  ENSE00004248721.1     2   \n",
       "...                    ...                ...  ...                ...   ...   \n",
       "2155000  ENSG00000198695.2  ENST00000361681.2  ...  ENSE00001434974.2     3   \n",
       "2155001  ENSG00000210194.1  ENST00000387459.1  ...  ENSE00001544476.1     3   \n",
       "2155002  ENSG00000198727.2  ENST00000361789.2  ...  ENSE00001436074.2     3   \n",
       "2155003  ENSG00000210195.2  ENST00000387460.2  ...  ENSE00001544475.2     3   \n",
       "2155004  ENSG00000210196.2  ENST00000387461.2  ...  ENSE00001544473.2     3   \n",
       "\n",
       "                        tag transcript_support_level havana_transcript  \\\n",
       "0                    TAGENE                      NaN               NaN   \n",
       "1                    TAGENE                      NaN               NaN   \n",
       "2                    TAGENE                      NaN               NaN   \n",
       "3                    TAGENE                      NaN               NaN   \n",
       "4                    TAGENE                      NaN               NaN   \n",
       "...                     ...                      ...               ...   \n",
       "2155000  appris_principal_1                      NaN               NaN   \n",
       "2155001   Ensembl_canonical                      NaN               NaN   \n",
       "2155002  appris_principal_1                      NaN               NaN   \n",
       "2155003   Ensembl_canonical                      NaN               NaN   \n",
       "2155004   Ensembl_canonical                      NaN               NaN   \n",
       "\n",
       "           hgnc_id  ont havana_gene         protein_id ccdsid  \n",
       "0              NaN  NaN         NaN                NaN    NaN  \n",
       "1              NaN  NaN         NaN                NaN    NaN  \n",
       "2              NaN  NaN         NaN                NaN    NaN  \n",
       "3              NaN  NaN         NaN                NaN    NaN  \n",
       "4              NaN  NaN         NaN                NaN    NaN  \n",
       "...            ...  ...         ...                ...    ...  \n",
       "2155000  HGNC:7462  NaN         NaN  ENSP00000354665.2    NaN  \n",
       "2155001  HGNC:7479  NaN         NaN                NaN    NaN  \n",
       "2155002  HGNC:7427  NaN         NaN  ENSP00000354554.2    NaN  \n",
       "2155003  HGNC:7499  NaN         NaN                NaN    NaN  \n",
       "2155004  HGNC:7494  NaN         NaN                NaN    NaN  \n",
       "\n",
       "[2155005 rows x 25 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_v47_df = pd.read_csv(\n",
    "    \"../data_raw/gencode.v47.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=str)\n",
    "gencode_v47_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05a82f7b-d762-42de-ba36-8153653f4561",
   "metadata": {
    "editable": true,
    "scrolled": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_classification</th>\n",
       "      <th>protein_id_orf_summary</th>\n",
       "      <th>Intron</th>\n",
       "      <th>Exon</th>\n",
       "      <th>5prime</th>\n",
       "      <th>3prime</th>\n",
       "      <th>closest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>14360</td>\n",
       "      <td>14829</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>PBT00037199</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Novel Protein coding not overlapping annotated...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>PBT00037199</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Novel Protein coding not overlapping annotated...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>15947</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>PBT00037199</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Novel Protein coding not overlapping annotated...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>PBT00037199</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Novel Protein coding not overlapping annotated...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>16854</td>\n",
       "      <td>17055</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>PBT00037199</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Novel Protein coding not overlapping annotated...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3967844</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>20582590</td>\n",
       "      <td>20582693</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>STRT02294670</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Novel Protein coding: internal additional pept...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3967845</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>20584474</td>\n",
       "      <td>20584524</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>STRT02294670</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Novel Protein coding: internal additional pept...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3967846</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>20588024</td>\n",
       "      <td>20588105</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>STRT02294670</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Novel Protein coding: internal additional pept...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3967847</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>20589484</td>\n",
       "      <td>20589575</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>STRT02294670</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Novel Protein coding: internal additional pept...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3967848</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>20592341</td>\n",
       "      <td>20592624</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>STRT02294670</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Novel Protein coding: internal additional pept...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3967849 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname     source feature     start       end score strand frame  \\\n",
       "0          chr1     PacBio    exon     14360     14829     .      -     .   \n",
       "1          chr1     PacBio    exon     14970     15038     .      -     .   \n",
       "2          chr1     PacBio    exon     15796     15947     .      -     .   \n",
       "3          chr1     PacBio    exon     16607     16765     .      -     .   \n",
       "4          chr1     PacBio    exon     16854     17055     .      -     .   \n",
       "...         ...        ...     ...       ...       ...   ...    ...   ...   \n",
       "3967844    chrY  StringTie    exon  20582590  20582693     .      +     .   \n",
       "3967845    chrY  StringTie    exon  20584474  20584524     .      +     .   \n",
       "3967846    chrY  StringTie    exon  20588024  20588105     .      +     .   \n",
       "3967847    chrY  StringTie    exon  20589484  20589575     .      +     .   \n",
       "3967848    chrY  StringTie    exon  20592341  20592624     .      +     .   \n",
       "\n",
       "                 gene_id transcript_id gene_name       gene_type  \\\n",
       "0        ENSG00000227232   PBT00037199    WASH7P      pseudogene   \n",
       "1        ENSG00000227232   PBT00037199    WASH7P      pseudogene   \n",
       "2        ENSG00000227232   PBT00037199    WASH7P      pseudogene   \n",
       "3        ENSG00000227232   PBT00037199    WASH7P      pseudogene   \n",
       "4        ENSG00000227232   PBT00037199    WASH7P      pseudogene   \n",
       "...                  ...           ...       ...             ...   \n",
       "3967844  ENSG00000198692  STRT02294670    EIF1AY  protein_coding   \n",
       "3967845  ENSG00000198692  STRT02294670    EIF1AY  protein_coding   \n",
       "3967846  ENSG00000198692  STRT02294670    EIF1AY  protein_coding   \n",
       "3967847  ENSG00000198692  STRT02294670    EIF1AY  protein_coding   \n",
       "3967848  ENSG00000198692  STRT02294670    EIF1AY  protein_coding   \n",
       "\n",
       "                                 transcript_classification  \\\n",
       "0        Novel Protein coding not overlapping annotated...   \n",
       "1        Novel Protein coding not overlapping annotated...   \n",
       "2        Novel Protein coding not overlapping annotated...   \n",
       "3        Novel Protein coding not overlapping annotated...   \n",
       "4        Novel Protein coding not overlapping annotated...   \n",
       "...                                                    ...   \n",
       "3967844  Novel Protein coding: internal additional pept...   \n",
       "3967845  Novel Protein coding: internal additional pept...   \n",
       "3967846  Novel Protein coding: internal additional pept...   \n",
       "3967847  Novel Protein coding: internal additional pept...   \n",
       "3967848  Novel Protein coding: internal additional pept...   \n",
       "\n",
       "        protein_id_orf_summary Intron Exon 5prime 3prime closest  \n",
       "0                          NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "1                          NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "2                          NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "3                          NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "4                          NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "...                        ...    ...  ...    ...    ...     ...  \n",
       "3967844                    NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "3967845                    NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "3967846                    NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "3967847                    NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "3967848                    NaN    NaN  NaN    NaN    NaN     NaN  \n",
       "\n",
       "[3967849 rows x 19 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HIT_hg38_df = pd.read_csv(\n",
    "    \"../data_raw/HIT_hg38.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=str)\n",
    "HIT_hg38_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2f915a5-c19d-4080-b1d0-85e8d592c556",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number unique transcripts\n",
      "HIT: 376582\n",
      "Number unique genes\n",
      "HIT: 22129\n"
     ]
    }
   ],
   "source": [
    "print(\"Number unique transcripts\")\n",
    "print(\"HIT:\", len(HIT_hg38_df[\"transcript_id\"].unique()))\n",
    "print(\"Number unique genes\")\n",
    "print(\"HIT:\", len(HIT_hg38_df[\"gene_id\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4bf7d227-70d5-40fe-912a-13a6593fab88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove duplicated tx (these duplicated transcripts were identified based on GffCompare output from HIT vs GENCODE)\n",
    "dup_tx = pd.read_csv(\"../data_processed/duplicated_transcripts.tsv\", sep=\"\\t\", dtype=str)[\"transcript_id\"].unique()\n",
    "HIT_hg38_df = HIT_hg38_df[~HIT_hg38_df[\"transcript_id\"].isin(dup_tx)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cc438985-52cc-43b2-8f3a-c180ee329571",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number unique transcripts\n",
      "HIT: 376460\n",
      "Number unique genes\n",
      "HIT: 22128\n"
     ]
    }
   ],
   "source": [
    "print(\"Number unique transcripts\")\n",
    "print(\"HIT:\", len(HIT_hg38_df[\"transcript_id\"].unique()))\n",
    "print(\"Number unique genes\")\n",
    "print(\"HIT:\", len(HIT_hg38_df[\"gene_id\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5fc9e4f8-ceaf-4518-97ee-dd7ef887a9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove non-standard chr\n",
    "standard_chromosomes = {f'chr{i}' for i in range(1, 23)}.union({'chrX', 'chrY'})\n",
    "HIT_hg38_df = HIT_hg38_df[HIT_hg38_df['seqname'].isin(standard_chromosomes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "00354927-8bb1-4205-badb-4a2f8b4e716d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number unique transcripts\n",
      "GENCODE: 385659\n",
      "HIT: 376343\n",
      "Number unique genes\n",
      "GENCODE: 78724\n",
      "HIT: 22119\n"
     ]
    }
   ],
   "source": [
    "print(\"Number unique transcripts\")\n",
    "print(\"GENCODE:\", len(gencode_v47_df[\"transcript_id\"].unique()))\n",
    "print(\"HIT:\", len(HIT_hg38_df[\"transcript_id\"].unique()))\n",
    "print(\"Number unique genes\")\n",
    "print(\"GENCODE:\", len(gencode_v47_df[\"gene_id\"].unique()))\n",
    "print(\"HIT:\", len(HIT_hg38_df[\"gene_id\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6e01fac8-1f0c-4ce0-9415-6df8058c1fd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build exon dictionaries for each chromosome\n",
    "def build_transcript_dict(df):\n",
    "    chr_dict = defaultdict(list)\n",
    "    grouped = df.groupby(\"transcript_id\")\n",
    "    \n",
    "    for tid, group in grouped:\n",
    "        seqname = group[\"seqname\"].iloc[0]\n",
    "        exon_coords = sorted(list(zip(group[\"start\"].astype(int), group[\"end\"].astype(int))))\n",
    "        chr_dict[seqname].append((tid, exon_coords))\n",
    "    return chr_dict\n",
    "\n",
    "HIT_hg38_df_exons = build_transcript_dict(HIT_hg38_df)\n",
    "gencode_v47_df_exons = build_transcript_dict(gencode_v47_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eb256c44-d089-4006-80db-51b0a198a284",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find transcripts fully annotated in GENCODE\n",
    "def exon_overlap(exons1, exons2, tolerance=100):\n",
    "    for i, ((s1, e1), (s2, e2)) in enumerate(zip(exons1, exons2)):\n",
    "        if i == 0:  # first exon\n",
    "            if abs(s1 - s2) > tolerance or e1 != e2:\n",
    "                return False\n",
    "        elif i == len(exons1) - 1:  # last exon\n",
    "            if abs(e1 - e2) > tolerance or s1 != s2:\n",
    "                return False\n",
    "        else:  # internal exon\n",
    "            if s1 != s2 or e1 != e2:\n",
    "                return False\n",
    "    return True\n",
    "\n",
    "\n",
    "# Find new splice variants with ALL exons annotated (GENCODE transcript may contain additional exons)\n",
    "def subset_match(HIT_exons, gencode_exons, tolerance=100):\n",
    "    g_idx = 0\n",
    "    for i, (hit_start, hit_end) in enumerate(HIT_exons):\n",
    "        found = False\n",
    "        while g_idx < len(gencode_exons):\n",
    "            gen_start, gen_end = gencode_exons[g_idx]\n",
    "            if i == 0:\n",
    "                # First HIT exon — allow tolerance only on start\n",
    "                if abs(hit_start - gen_start) <= tolerance and hit_end == gen_end:\n",
    "                    found = True\n",
    "                    g_idx += 1\n",
    "                    break\n",
    "            elif i == len(HIT_exons) - 1:\n",
    "                # Last HIT exon — allow tolerance only on end\n",
    "                if abs(hit_end - gen_end) <= tolerance and hit_start == gen_start:\n",
    "                    found = True\n",
    "                    g_idx += 1\n",
    "                    break\n",
    "            else:\n",
    "                # Internal exon — exact match\n",
    "                if hit_start == gen_start and hit_end == gen_end:\n",
    "                    found = True\n",
    "                    g_idx += 1\n",
    "                    break\n",
    "\n",
    "            g_idx += 1\n",
    "        if not found:\n",
    "            return False\n",
    "\n",
    "    return True\n",
    "\n",
    "\n",
    "\n",
    "def found_exon(HIT_exon, gencode_exon_dict, position, nr_HIT_exons):\n",
    "    hit_start, hit_end = HIT_exon\n",
    "\n",
    "    # Always allow an internal exact match\n",
    "    candidate_keys = [(hit_start, hit_end, 'internal')]\n",
    "\n",
    "    if position == 0:\n",
    "        # HIT first exon: allow tolerant 'first' + internal exact\n",
    "        candidate_keys.append((hit_start, hit_end, 'first'))\n",
    "    elif position == nr_HIT_exons - 1:\n",
    "        # HIT last exon: allow tolerant 'last'  + internal exact\n",
    "        candidate_keys.append((hit_start, hit_end, 'last'))\n",
    "    else:\n",
    "        # HIT internal exon: allow exact matches to GENCODE first/last \n",
    "        candidate_keys.extend([\n",
    "            (hit_start, hit_end, 'first'),   \n",
    "            (hit_start, hit_end, 'last')     \n",
    "        ])\n",
    "\n",
    "    matched_tids = set()\n",
    "    for k in candidate_keys:\n",
    "        tids = gencode_exon_dict.get(k)\n",
    "        if tids:\n",
    "            matched_tids |= tids\n",
    "\n",
    "    return (len(matched_tids) > 0, matched_tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "27fb5e6f-75dc-469e-80f1-b1a9f94cbfec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a lookup set of GENCODE exons which includes a tolerance window for the start and end of a tx\n",
    "def build_tolerant_gencode_set(transcripts, tolerance=100):\n",
    "    exon_dict = defaultdict(set)\n",
    "    for tid, exons in transcripts:\n",
    "        for i, (start, end) in enumerate(exons):\n",
    "            if i == 0:\n",
    "                for delta in range(-tolerance, tolerance + 1):\n",
    "                    exon_dict[(start + delta, end, 'first')].add(tid)\n",
    "            elif i == len(exons) - 1:\n",
    "                for delta in range(-tolerance, tolerance + 1):\n",
    "                    exon_dict[(start, end + delta, 'last')].add(tid)\n",
    "            else:\n",
    "                exon_dict[(start, end, 'internal')].add(tid)\n",
    "    return exon_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "02594388-b0e6-48e1-a02d-ad8bc234eff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge exons that are directly adjacent to each other\n",
    "def merge_adjacent_exons(exons, tid):\n",
    "    if not exons:\n",
    "        return []\n",
    "    \n",
    "    # Sort exons by start coordinate \n",
    "    exons = sorted(exons, key=lambda x: x[0])\n",
    "    merged = [exons[0]]\n",
    "    merged_count = 0\n",
    "    \n",
    "    for start, end in exons[1:]:\n",
    "        last_start, last_end = merged[-1]\n",
    "        \n",
    "        # If current exon starts directly after the last one ends\n",
    "        if start == last_end + 1:\n",
    "            # Merge them\n",
    "            merged[-1] = (last_start, end)\n",
    "            merged_count += 1\n",
    "        else:\n",
    "            merged.append((start, end))\n",
    "\n",
    "    return merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5fee2cee-e51b-4e03-a158-64896ee95759",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                          \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr18\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr20\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr7\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr9\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrX\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                           \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrY\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                        \r"
     ]
    }
   ],
   "source": [
    "matching_transcripts = {}\n",
    "subset_transcripts = {}\n",
    "partially_unannotated_transcripts = {}\n",
    "fully_unannotated_transcripts = {}\n",
    "new_combo_annotated_exons = {}\n",
    "\n",
    "for chr_name in HIT_hg38_df_exons:\n",
    "    print(chr_name)\n",
    "    if chr_name not in gencode_v47_df_exons:\n",
    "        continue\n",
    "    \n",
    "    HIT_hg38_transcript_list = HIT_hg38_df_exons[chr_name]\n",
    "    gencode_v47_transcript_list = gencode_v47_df_exons[chr_name]\n",
    "    \n",
    "    # Group GENCODE transcripts by exon count\n",
    "    gencode_by_exon_count = defaultdict(list)\n",
    "    for tid, exons in gencode_v47_transcript_list:\n",
    "        gencode_by_exon_count[len(exons)].append((tid, exons))\n",
    "\n",
    "    # Build fast exon lookup set (for finding individual annotated exons)\n",
    "    gencode_exon_set = build_tolerant_gencode_set(gencode_v47_transcript_list)\n",
    "\n",
    "    \n",
    "    for HIT_tid, HIT_exons in tqdm(HIT_hg38_transcript_list, desc=f\"Transcripts in {chr_name}\", leave=False):\n",
    "        HIT_exons = merge_adjacent_exons(HIT_exons, HIT_tid)\n",
    "        matched = False\n",
    "        len_hit = len(HIT_exons)\n",
    "        \n",
    "        # Try to match entire transcript exactly\n",
    "        for gencode_tid, gencode_exons in gencode_by_exon_count[len_hit]:\n",
    "            if exon_overlap(HIT_exons, gencode_exons):\n",
    "                matching_transcripts[HIT_tid] = {\n",
    "                    'chr': chr_name,\n",
    "                    'exons': HIT_exons,\n",
    "                    'matched_to': gencode_exons,\n",
    "                    'matched_ref_id': gencode_tid\n",
    "                }\n",
    "                matched = True\n",
    "                break\n",
    "\n",
    "        if matched:\n",
    "            continue\n",
    "            \n",
    "        # Try subset match (HIT tx contained in GENCODE tx)\n",
    "        for longer_len in range(len_hit + 1, max(gencode_by_exon_count.keys()) + 1):\n",
    "            for gencode_tid, gencode_exons in gencode_by_exon_count[longer_len]:\n",
    "                # Skip if exon coordinates are out of possible range\n",
    "                if HIT_exons[0][0] > gencode_exons[-1][1] or HIT_exons[-1][1] < gencode_exons[0][0]:\n",
    "                    continue\n",
    "                if subset_match(HIT_exons, gencode_exons):\n",
    "                    subset_transcripts[HIT_tid] = {\n",
    "                        'chr': chr_name,\n",
    "                        'exons': HIT_exons,\n",
    "                        'matched_to': gencode_exons,\n",
    "                        'matched_ref_id': gencode_tid\n",
    "                    }\n",
    "                    matched = True\n",
    "                    break\n",
    "            if matched:\n",
    "                break\n",
    "            \n",
    "        if not matched:\n",
    "            # If entire tx can't be matched, try to match individual exons\n",
    "            annotated_info = [\n",
    "                found_exon(exon, gencode_exon_set, i, len_hit)\n",
    "                for i, exon in enumerate(HIT_exons)\n",
    "                ]\n",
    "            annotated_flags = [flag for flag, _ in annotated_info]\n",
    "            matched_tids = set().union(*[tids for flag, tids in annotated_info if flag])\n",
    "     \n",
    "            # No exons are annotated\n",
    "            if all(not flag for flag in annotated_flags):\n",
    "                fully_unannotated_transcripts[HIT_tid] = {\n",
    "                    'chr': chr_name,\n",
    "                    'exons': HIT_exons\n",
    "                }\n",
    "            # All exons are annotated (new splice variant)\n",
    "            elif all(annotated_flags):\n",
    "                ref_match_counts = defaultdict(int)\n",
    "                for flag, tids in annotated_info:\n",
    "                    for tid in tids:\n",
    "                        ref_match_counts[tid] += 1\n",
    "\n",
    "                # Find the ref ID with the most matches\n",
    "                if ref_match_counts:\n",
    "                    # Get the highest match count\n",
    "                    max_count = max(ref_match_counts.values())\n",
    "    \n",
    "                    # Find all ref IDs with that count\n",
    "                    tied_ref_ids = [tid for tid, count in ref_match_counts.items() if count == max_count]\n",
    "    \n",
    "                    if len(tied_ref_ids) == 1:\n",
    "                        # No tie: take the sole best match\n",
    "                        best_ref_id = tied_ref_ids[0]\n",
    "                    else:\n",
    "                        # Tie: resolve by comparing tx lengths\n",
    "                        hit_exons = dict(HIT_hg38_transcript_list)[HIT_tid]\n",
    "                        hit_length = hit_exons[-1][1] - hit_exons[0][0]\n",
    "        \n",
    "                        ref_lengths = {}\n",
    "                        for tid in tied_ref_ids:\n",
    "                            ref_exons = dict(gencode_v47_transcript_list)[tid]\n",
    "                            ref_lengths[tid] = ref_exons[-1][1] - ref_exons[0][0]\n",
    "        \n",
    "                        # Choose the ref ID with length closest to HIT\n",
    "                        best_ref_id = min(ref_lengths.items(), key=lambda x: abs(x[1] - hit_length))[0]\n",
    "                else:\n",
    "                    best_ref_id = None\n",
    "\n",
    "                # Save result\n",
    "                new_combo_annotated_exons[HIT_tid] = {\n",
    "                    'chr': chr_name,\n",
    "                    'exons': HIT_exons,\n",
    "                    'matched_ref_id': list(matched_tids),\n",
    "                    'best_match_ref_id': best_ref_id\n",
    "                }\n",
    "            # Some exons annotated\n",
    "            elif any(not flag for flag in annotated_flags):\n",
    "                unannotated_exons = [\n",
    "                    exon for exon, is_annotated in zip(HIT_exons, annotated_flags) if not is_annotated\n",
    "                ]\n",
    "                # Count how many times each ref ID appears across annotated exons\n",
    "                ref_match_counts = defaultdict(int)\n",
    "                for flag, tids in annotated_info:\n",
    "                    if flag:\n",
    "                        for tid in tids:\n",
    "                            ref_match_counts[tid] += 1\n",
    "\n",
    "                # Find the ref ID with the most matches\n",
    "                if ref_match_counts:\n",
    "                    # Get the highest match count\n",
    "                    max_count = max(ref_match_counts.values())\n",
    "    \n",
    "                    # Find all ref IDs with that count\n",
    "                    tied_ref_ids = [tid for tid, count in ref_match_counts.items() if count == max_count]\n",
    "    \n",
    "                    if len(tied_ref_ids) == 1:\n",
    "                        # No tie: take the sole best match\n",
    "                        best_ref_id = tied_ref_ids[0]\n",
    "                    else:\n",
    "                        # Tie: resolve by comparing tx lengths\n",
    "                        hit_exons = dict(HIT_hg38_transcript_list)[HIT_tid]\n",
    "                        hit_length = hit_exons[-1][1] - hit_exons[0][0]\n",
    "        \n",
    "                        ref_lengths = {}\n",
    "                        for tid in tied_ref_ids:\n",
    "                            ref_exons = dict(gencode_v47_transcript_list)[tid]\n",
    "                            ref_lengths[tid] = ref_exons[-1][1] - ref_exons[0][0]\n",
    "        \n",
    "                        # Choose the ref ID with length closest to HIT\n",
    "                        best_ref_id = min(ref_lengths.items(), key=lambda x: abs(x[1] - hit_length))[0]\n",
    "                else:\n",
    "                    best_ref_id = None\n",
    "\n",
    "                # Save result\n",
    "                partially_unannotated_transcripts[HIT_tid] = {\n",
    "                    'chr': chr_name,\n",
    "                    'unannotated_exons': unannotated_exons,\n",
    "                    'exons': HIT_exons,\n",
    "                    'matched_ref_id': list(matched_tids),\n",
    "                    'best_match_ref_id': best_ref_id\n",
    "                }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f331a89b-c88f-451e-ac1c-edbfd9b5f741",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "38891"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matching_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f648783f-eff7-4e27-be0d-8fbdac0602f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25891"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(subset_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c8477cce-634f-4f09-addd-726e9165fdd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "257815"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(partially_unannotated_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4077820c-2d7d-4d1c-a132-71a6ffa43a3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17261"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(fully_unannotated_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "17abbc6b-f458-4a91-924d-fc4b6caa6dbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36485"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(new_combo_annotated_exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "637f94b3-fb72-4d70-91dd-637b52e7e71f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376343"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matching_transcripts) + len(subset_transcripts) + len(partially_unannotated_transcripts) + len(fully_unannotated_transcripts) + len(new_combo_annotated_exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "26d47790-df76-46e9-8dee-fbf3a9b649d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376343"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(len(v) for v in HIT_hg38_df_exons.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cf89413-35d0-4ed6-959e-65a5261b3185",
   "metadata": {},
   "source": [
    "### Split up partially unannotated tx into overlapping & non-overlapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "55dfaef1-3d23-4851-af15-1e8092c95cbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert dictionary of partially unannotated exons into PyRanges\n",
    "rows = []\n",
    "for tid, info in partially_unannotated_transcripts.items():\n",
    "    for exon in info[\"unannotated_exons\"]:\n",
    "        start, end = exon\n",
    "        rows.append({\n",
    "            \"Chromosome\": info[\"chr\"],\n",
    "            \"Start\": start,\n",
    "            \"End\": end,\n",
    "            \"transcript_id\": tid\n",
    "        })\n",
    "\n",
    "partially_unannotated_df = pd.DataFrame(rows)\n",
    "# Add strand info\n",
    "strand_info = HIT_hg38_df[['transcript_id', 'strand']].drop_duplicates(subset='transcript_id')\n",
    "strand_info = strand_info.rename(columns={'strand': 'Strand'})\n",
    "\n",
    "# Merge \n",
    "partially_unannotated_df = partially_unannotated_df.merge(\n",
    "    strand_info,\n",
    "    on='transcript_id',\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "partially_unannotated_pr = pr.PyRanges(partially_unannotated_df)\n",
    "\n",
    "\n",
    "gencode_pr = pr.PyRanges(\n",
    "    gencode_v47_df.rename(columns={\n",
    "        \"seqname\": \"Chromosome\",\n",
    "        \"start\": \"Start\",\n",
    "        \"end\": \"End\",\n",
    "        \"strand\": \"Strand\"\n",
    "    })[[\"Chromosome\", \"Start\", \"End\", \"Strand\", \"transcript_id\"]]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "33a375b5-64fc-4e01-a39e-09f0fd469196",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join the unannotated exon df with gencode exons\n",
    "overlapping_exon_pu = partially_unannotated_pr.join(gencode_pr)\n",
    "overlapping_exon_pu = overlapping_exon_pu[overlapping_exon_pu.Strand == overlapping_exon_pu.Strand_b]\n",
    "overlapping_exon_pu = overlapping_exon_pu.df\n",
    "overlapping_exon_pu = overlapping_exon_pu.rename(columns={\n",
    "    \"transcript_id\": \"HIT_transcript_id\",\n",
    "    \"transcript_id_b\": \"gencode_transcript_id\"\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c8aa5400-de8b-47fa-99e9-ee20ccc303d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign each exon as overlapping or non-overlapping\n",
    "overlapping_coords_pu = set(\n",
    "    tuple(row) for row in overlapping_exon_pu[[\"Chromosome\", \"Start\", \"End\"]].values\n",
    ")\n",
    "\n",
    "partially_unannotated_df[\"is_overlapping\"] = partially_unannotated_df.apply(\n",
    "    lambda row: (row[\"Chromosome\"], row[\"Start\"], row[\"End\"]) in overlapping_coords_pu,\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4446582e-5a34-4b00-ae3d-40ac07299e11",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "# Save df for downstream analysis of novel exons\n",
    "with open(\"../data_processed/partially_unannotated_df.pkl\", \"wb\") as f:\n",
    "    pickle.dump(partially_unannotated_df, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9b3c9769-9aa1-4539-847a-a34261576d11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split up partially annotated tx into those with exon overlap and those not overlapping gencode exon\n",
    "non_overlapping_ids_pu = set(\n",
    "    partially_unannotated_df.loc[partially_unannotated_df[\"is_overlapping\"] == False, \"transcript_id\"]\n",
    ")\n",
    "overlapping_ids_pu = set(partially_unannotated_df[\"transcript_id\"]) - non_overlapping_ids_pu\n",
    "\n",
    "nonoverlapping_partially_unannotated = {}\n",
    "overlapping_partially_unannotated = {}\n",
    "\n",
    "nonoverlapping_partially_unannotated = {\n",
    "    tid: val for tid, val in partially_unannotated_transcripts.items()\n",
    "    if tid in non_overlapping_ids_pu\n",
    "}\n",
    "\n",
    "overlapping_partially_unannotated = {\n",
    "    tid: val for tid, val in partially_unannotated_transcripts.items()\n",
    "    if tid in overlapping_ids_pu\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4328a1d0-9acd-4afd-86d6-4cdd53351612",
   "metadata": {},
   "source": [
    "### Split up fully unannotated tx into overlapping & non-overlapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c5f1637e-ab27-4de0-81a8-3a60cc1e618a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now do the same for the fully unannotated tx\n",
    "rows = []\n",
    "for tid, info in fully_unannotated_transcripts.items():\n",
    "    for exon in info[\"exons\"]:\n",
    "        start, end = exon\n",
    "        rows.append({\n",
    "            \"Chromosome\": info[\"chr\"],\n",
    "            \"Start\": start,\n",
    "            \"End\": end,\n",
    "            \"transcript_id\": tid\n",
    "        })\n",
    "\n",
    "fully_unannotated_df = pd.DataFrame(rows)\n",
    "\n",
    "# Merge with strand info\n",
    "fully_unannotated_df = fully_unannotated_df.merge(\n",
    "    strand_info,\n",
    "    on='transcript_id',\n",
    "    how='left'\n",
    ")\n",
    "fully_unannotated_pr = pr.PyRanges(fully_unannotated_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "eb442dd7-c635-4305-85bd-3cb77347e3f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# join the unannotated exon df with gencode exons\n",
    "overlapping_exon_fu = fully_unannotated_pr.join(gencode_pr)\n",
    "overlapping_exon_fu = overlapping_exon_fu[overlapping_exon_fu.Strand == overlapping_exon_fu.Strand_b]\n",
    "overlapping_exon_fu = overlapping_exon_fu.df\n",
    "overlapping_exon_fu = overlapping_exon_fu.rename(columns={\n",
    "    \"transcript_id\": \"HIT_transcript_id\",\n",
    "    \"transcript_id_b\": \"gencode_transcript_id\"\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "35ba6088-1449-42b1-9440-4de5290045e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute size of overlap\n",
    "overlapping_exon_fu[\"overlap_size\"] = (\n",
    "    overlapping_exon_fu[[\"Start\", \"End\", \"Start_b\", \"End_b\"]]\n",
    "    .apply(lambda row: max(0, min(row[\"End\"], row[\"End_b\"]) - max(row[\"Start\"], row[\"Start_b\"])), axis=1)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b25124bb-dabc-45a5-892d-fb9fca108586",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign each exon as overlapping or non-overlapping\n",
    "overlapping_coords_fu = set(\n",
    "    tuple(row) for row in overlapping_exon_fu[[\"Chromosome\", \"Start\", \"End\"]].values\n",
    ")\n",
    "\n",
    "fully_unannotated_df[\"is_overlapping\"] = fully_unannotated_df.apply(\n",
    "    lambda row: (row[\"Chromosome\"], row[\"Start\"], row[\"End\"]) in overlapping_coords_fu,\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "42a6fc46-2530-4d33-ba21-37ed609ae951",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute total overlap size for each HIT-gencode transcript pair\n",
    "tx_overlap = (\n",
    "    overlapping_exon_fu\n",
    "    .groupby([\"HIT_transcript_id\", \"gencode_transcript_id\"])[\"overlap_size\"]\n",
    "    .sum()\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Select the gencode transcript with the highest total overlap per HIT transcript\n",
    "best_tx_overlap = (\n",
    "    tx_overlap\n",
    "    .sort_values(by=[\"overlap_size\", \"gencode_transcript_id\"], ascending=[False, True])\n",
    "    .drop_duplicates(subset=[\"HIT_transcript_id\"])\n",
    "    .rename(columns={\n",
    "        \"HIT_transcript_id\": \"transcript_id\",\n",
    "        \"gencode_transcript_id\": \"largest_overlap_gencode_id\"\n",
    "    })[[\"transcript_id\", \"largest_overlap_gencode_id\"]]\n",
    ")\n",
    "\n",
    "# Merge best transcript-level overlap into unannotated_df\n",
    "fully_unannotated_df = fully_unannotated_df.merge(\n",
    "    best_tx_overlap,\n",
    "    on=\"transcript_id\",\n",
    "    how=\"left\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "042056c8-ce22-4d27-b41f-0e49a2309d09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split up fully unannotated tx into those with exon overlap and those not overlapping gencode exon\n",
    "overlapping_ids_fu = set(\n",
    "    fully_unannotated_df.loc[fully_unannotated_df[\"is_overlapping\"] == True, \"transcript_id\"]\n",
    ")\n",
    "non_overlapping_ids_fu = set(fully_unannotated_df[\"transcript_id\"]) - overlapping_ids_fu\n",
    "\n",
    "\n",
    "nonoverlapping_fully_unannotated = {}\n",
    "overlapping_fully_unannotated = {}\n",
    "\n",
    "nonoverlapping_fully_unannotated = {\n",
    "    tid: val for tid, val in fully_unannotated_transcripts.items()\n",
    "    if tid in non_overlapping_ids_fu\n",
    "}\n",
    "\n",
    "overlapping_fully_unannotated = {\n",
    "    tid: val for tid, val in fully_unannotated_transcripts.items()\n",
    "    if tid in overlapping_ids_fu\n",
    "}\n",
    "\n",
    "for tid in overlapping_fully_unannotated:\n",
    "    # Get all gencode IDs with the largest overlaps for this transcript\n",
    "    gencode_ids = fully_unannotated_df.loc[\n",
    "        fully_unannotated_df[\"transcript_id\"] == tid, \"largest_overlap_gencode_id\"\n",
    "    ].dropna().values[0]\n",
    "\n",
    "    # Add as a new key to the transcript entry\n",
    "    overlapping_fully_unannotated[tid][\"largest_overlap_gencode_id\"] = gencode_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2541c31d-472c-4745-bb28-4ce82792d2fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dict_to_df(d, label):\n",
    "    df = pd.DataFrame.from_dict(d, orient='index')\n",
    "    df.reset_index(inplace=True)\n",
    "    df.rename(columns={'index': 'transcript_id'}, inplace=True)\n",
    "    df['category'] = label\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bb2f38fe-d49e-4420-b1c0-c1914816049d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert dictionaries into df\n",
    "df_nonoverlap_partial = dict_to_df(nonoverlapping_partially_unannotated, \"Transcripts with annotated exons and non-overlapping unannotated exons\")\n",
    "df_overlap_partial = dict_to_df(overlapping_partially_unannotated, \"Transcripts with annotated exons and only overlapping unannotated exons\")\n",
    "df_unannotated_overlap = dict_to_df(overlapping_fully_unannotated, \"Fully unannotated transcripts with overlapping exons\")\n",
    "df_unannotated_nonoverlap = dict_to_df(nonoverlapping_fully_unannotated, \"Fully unannotated transcripts without any overlapping exons\")\n",
    "df_combo = dict_to_df(new_combo_annotated_exons, \"Unannotated transcripts composed entirely of annotated exons\")\n",
    "df_match = dict_to_df(matching_transcripts, \"GENCODE-annotated transcripts\")\n",
    "df_subset = dict_to_df(subset_transcripts, \"Unannotated transcripts fully contained within GENCODE transcripts\")\n",
    "\n",
    "\n",
    "combined_df = pd.concat([df_match, df_subset, df_nonoverlap_partial, df_overlap_partial, df_unannotated_overlap, df_unannotated_nonoverlap, df_combo], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d7984081-2af6-41b7-b648-937f5d57efbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add strand info\n",
    "combined_df = combined_df.merge(\n",
    "    strand_info[[\"transcript_id\", \"Strand\"]],\n",
    "    how='left',\n",
    "    on='transcript_id',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "fc8d106b-4188-4bc4-8e12-318d8fe8ec18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "category\n",
       "Transcripts with annotated exons and only overlapping unannotated exons    193904\n",
       "Transcripts with annotated exons and non-overlapping unannotated exons      63911\n",
       "GENCODE-annotated transcripts                                               38891\n",
       "Unannotated transcripts composed entirely of annotated exons                36485\n",
       "Unannotated transcripts fully contained within GENCODE transcripts          25891\n",
       "Fully unannotated transcripts with overlapping exons                        13970\n",
       "Fully unannotated transcripts without any overlapping exons                  3291\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"category\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4fcc5dc5-2f25-42ca-9b63-912da468098c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_tx_category(df):\n",
    "    tx_category = df['category']\n",
    "    if tx_category == 'GENCODE-annotated transcripts':\n",
    "        summarized_category = \"GENCODE-annotated transcripts\"\n",
    "        ref_id = df['matched_ref_id']\n",
    "        \n",
    "    elif tx_category == \"Unannotated transcripts fully contained within GENCODE transcripts\":\n",
    "        summarized_category = \"Unannotated transcripts with all exons annotated\"\n",
    "        ref_id = df['matched_ref_id']\n",
    "        \n",
    "    elif tx_category == \"Unannotated transcripts composed entirely of annotated exons\":\n",
    "        summarized_category = \"Unannotated transcripts with all exons annotated\"\n",
    "        ref_id = df['best_match_ref_id']\n",
    "        \n",
    "    elif tx_category in [\"Transcripts with annotated exons and only overlapping unannotated exons\", \"Transcripts with annotated exons and non-overlapping unannotated exons\"]:\n",
    "        summarized_category = \"Unannotated transcripts with annotated and unannotated exons\"\n",
    "        ref_id = df['best_match_ref_id']\n",
    "        \n",
    "    elif tx_category == \"Fully unannotated transcripts with overlapping exons\":\n",
    "        summarized_category = \"Unannotated transcripts without any annotated exons\"\n",
    "        ref_id = df['largest_overlap_gencode_id']\n",
    "        \n",
    "    elif tx_category == \"Fully unannotated transcripts without any overlapping exons\":\n",
    "        summarized_category = \"Unannotated transcripts without any annotated exons\"\n",
    "        ref_id = 'None'\n",
    "    \n",
    "    else:\n",
    "        summarized_category = 'None'\n",
    "        ref_id = 'None'\n",
    "        \n",
    "    return summarized_category, ref_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d1fbb52f-51c8-48f7-808b-14cebc56aad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df[['summarized_category', 'ref_id']] = combined_df.apply(summarize_tx_category, axis=1, result_type='expand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "92d6a11f-5be9-4236-970e-188efaaaad27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(61077274, 61077628), (61077916, 61079203)]</td>\n",
       "      <td>[(61077326, 61077628), (61077916, 61079231)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000699992.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160261734, 160261922), (160281430, 160281892)]</td>\n",
       "      <td>[(160261682, 160261922), (160281430, 160281935)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000756371.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCT00000022754.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(23191893, 23195001), (23217291, 23217499)]</td>\n",
       "      <td>[(23191895, 23195001), (23217291, 23217502)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id   chr                                              exons  \\\n",
       "0  ENCT00000006334.1  chr1       [(61077274, 61077628), (61077916, 61079203)]   \n",
       "1  ENCT00000013056.1  chr1   [(160261734, 160261922), (160281430, 160281892)]   \n",
       "2  ENCT00000019620.1  chr1  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3  ENCT00000022754.1  chr1       [(23191893, 23195001), (23217291, 23217499)]   \n",
       "4  ENCT00000039204.1  chr1   [(233319835, 233321155), (233327229, 233327357)]   \n",
       "\n",
       "                                          matched_to  \\\n",
       "0       [(61077326, 61077628), (61077916, 61079231)]   \n",
       "1   [(160261682, 160261922), (160281430, 160281935)]   \n",
       "2  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3       [(23191895, 23195001), (23217291, 23217502)]   \n",
       "4   [(233319834, 233321155), (233327229, 233327455)]   \n",
       "\n",
       "                        category unannotated_exons Strand  \\\n",
       "0  GENCODE-annotated transcripts               NaN      +   \n",
       "1  GENCODE-annotated transcripts               NaN      +   \n",
       "2  GENCODE-annotated transcripts               NaN      +   \n",
       "3  GENCODE-annotated transcripts               NaN      -   \n",
       "4  GENCODE-annotated transcripts               NaN      -   \n",
       "\n",
       "             summarized_category             ref_id  \n",
       "0  GENCODE-annotated transcripts  ENST00000699992.1  \n",
       "1  GENCODE-annotated transcripts  ENST00000756371.1  \n",
       "2  GENCODE-annotated transcripts  ENST00000676153.1  \n",
       "3  GENCODE-annotated transcripts  ENST00000374619.2  \n",
       "4  GENCODE-annotated transcripts  ENST00000771587.1  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df = combined_df.drop(columns=[\"matched_ref_id\", \"best_match_ref_id\", \"largest_overlap_gencode_id\"])\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60353d6d-e498-432d-9f65-b9432464e29a",
   "metadata": {},
   "source": [
    "## Add gene name and type info from Ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1a4288b9-d601-420f-b119-679ce9e71761",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0  ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "1  ENSG00000211459      ENSG00000211459.2      ENST00000389680   \n",
       "2  ENSG00000210077      ENSG00000210077.1      ENST00000387342   \n",
       "3  ENSG00000210082      ENSG00000210082.2      ENST00000387347   \n",
       "4  ENSG00000209082      ENSG00000209082.1      ENST00000386347   \n",
       "\n",
       "  Transcript stable ID version Gene name Gene type  \n",
       "0            ENST00000387314.1     MT-TF   Mt_tRNA  \n",
       "1            ENST00000389680.2   MT-RNR1   Mt_rRNA  \n",
       "2            ENST00000387342.1     MT-TV   Mt_tRNA  \n",
       "3            ENST00000387347.2   MT-RNR2   Mt_rRNA  \n",
       "4            ENST00000386347.1    MT-TL1   Mt_tRNA  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names = pd.read_csv('../data_raw/mart_export.txt',\n",
    "                           sep='\\t')\n",
    "gene_names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "89aa08b1-3fe1-41d8-8927-7e93189eb9e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df['stripped_ref_id'] = combined_df['ref_id'].str.replace(r'\\.\\d+$', '', regex=True)\n",
    "\n",
    "\n",
    "combined_df = combined_df.merge(\n",
    "    gene_names[[\"Transcript stable ID\", \"Gene stable ID\", \"Gene name\", \"Gene type\"]],\n",
    "    how='left',\n",
    "    left_on='stripped_ref_id',\n",
    "    right_on='Transcript stable ID'\n",
    ").drop(columns=\"Transcript stable ID\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1a01dac4-faed-4c25-9679-de0dce6d4367",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene type\n",
       "protein_coding                        330745\n",
       "lncRNA                                 40993\n",
       "transcribed_unprocessed_pseudogene       482\n",
       "processed_pseudogene                     364\n",
       "transcribed_unitary_pseudogene           136\n",
       "TEC                                       59\n",
       "unprocessed_pseudogene                    56\n",
       "transcribed_processed_pseudogene          53\n",
       "IG_V_gene                                 38\n",
       "snRNA                                     29\n",
       "misc_RNA                                  16\n",
       "miRNA                                     15\n",
       "snoRNA                                    14\n",
       "TR_C_gene                                  5\n",
       "scaRNA                                     5\n",
       "unitary_pseudogene                         5\n",
       "ribozyme                                   1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"Gene type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8f0f1c4e-e743-4a6e-81c6-45df1aef9652",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(20239)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"Gene name\"].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9701eed3-92e6-49dc-8466-c0cbb5111f3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(3327)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"Gene type\"].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "eb871c09-6e82-40d9-9457-abbf40ffbd99",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "      <td>ENST00000771587</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>ENST00000289779.3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160997957, 160998906), (160999043, 160999091...</td>\n",
       "      <td>[(160997957, 160998906), (160999043, 160999091...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000289779.7</td>\n",
       "      <td>ENST00000289779</td>\n",
       "      <td>ENSG00000270149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>ENST00000362058.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(16629670, 16631109), (16631530, 16631613), (...</td>\n",
       "      <td>[(16629670, 16631109), (16631530, 16631613), (...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000362058.2</td>\n",
       "      <td>ENST00000362058</td>\n",
       "      <td>ENSG00000291072</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1482</th>\n",
       "      <td>ENST00000412483.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(234212606, 234214107), (234214868, 234215088)]</td>\n",
       "      <td>[(234212606, 234214107), (234214868, 234215088)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000412483.1</td>\n",
       "      <td>ENST00000412483</td>\n",
       "      <td>ENSG00000233332</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1504</th>\n",
       "      <td>ENST00000415019.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(158197922, 158199835), (158203793, 158203877)]</td>\n",
       "      <td>[(158197922, 158199835), (158203793, 158203877)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000415019.1</td>\n",
       "      <td>ENST00000415019</td>\n",
       "      <td>ENSG00000176320</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376259</th>\n",
       "      <td>STRT02285555</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(102599494, 102599847), (102601464, 102601509...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000466616.6</td>\n",
       "      <td>ENST00000466616</td>\n",
       "      <td>ENSG00000271147</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376300</th>\n",
       "      <td>ENCT00000484963.1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(11161704, 11163137), (11212744, 11212802), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000796795.1</td>\n",
       "      <td>ENST00000796795</td>\n",
       "      <td>ENSG00000291032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376328</th>\n",
       "      <td>STRT02292412</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(12406182, 12406500), (12406796, 12406937), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000737339.1</td>\n",
       "      <td>ENST00000737339</td>\n",
       "      <td>ENSG00000291034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376330</th>\n",
       "      <td>STRT02293232</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(11167047, 11167217), (11212744, 11212802), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000796810.1</td>\n",
       "      <td>ENST00000796810</td>\n",
       "      <td>ENSG00000291032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376331</th>\n",
       "      <td>STRT02293605</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(12382179, 12382327), (12398680, 12398817), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000737331.1</td>\n",
       "      <td>ENST00000737331</td>\n",
       "      <td>ENSG00000291034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20239 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id   chr  \\\n",
       "4       ENCT00000039204.1  chr1   \n",
       "121     ENST00000289779.3  chr1   \n",
       "516     ENST00000362058.2  chr1   \n",
       "1482    ENST00000412483.1  chr1   \n",
       "1504    ENST00000415019.1  chr1   \n",
       "...                   ...   ...   \n",
       "376259       STRT02285555  chrX   \n",
       "376300  ENCT00000484963.1  chrY   \n",
       "376328       STRT02292412  chrY   \n",
       "376330       STRT02293232  chrY   \n",
       "376331       STRT02293605  chrY   \n",
       "\n",
       "                                                    exons  \\\n",
       "4        [(233319835, 233321155), (233327229, 233327357)]   \n",
       "121     [(160997957, 160998906), (160999043, 160999091...   \n",
       "516     [(16629670, 16631109), (16631530, 16631613), (...   \n",
       "1482     [(234212606, 234214107), (234214868, 234215088)]   \n",
       "1504     [(158197922, 158199835), (158203793, 158203877)]   \n",
       "...                                                   ...   \n",
       "376259  [(102599494, 102599847), (102601464, 102601509...   \n",
       "376300  [(11161704, 11163137), (11212744, 11212802), (...   \n",
       "376328  [(12406182, 12406500), (12406796, 12406937), (...   \n",
       "376330  [(11167047, 11167217), (11212744, 11212802), (...   \n",
       "376331  [(12382179, 12382327), (12398680, 12398817), (...   \n",
       "\n",
       "                                               matched_to  \\\n",
       "4        [(233319834, 233321155), (233327229, 233327455)]   \n",
       "121     [(160997957, 160998906), (160999043, 160999091...   \n",
       "516     [(16629670, 16631109), (16631530, 16631613), (...   \n",
       "1482     [(234212606, 234214107), (234214868, 234215088)]   \n",
       "1504     [(158197922, 158199835), (158203793, 158203877)]   \n",
       "...                                                   ...   \n",
       "376259                                                NaN   \n",
       "376300                                                NaN   \n",
       "376328                                                NaN   \n",
       "376330                                                NaN   \n",
       "376331                                                NaN   \n",
       "\n",
       "                                                 category unannotated_exons  \\\n",
       "4                           GENCODE-annotated transcripts               NaN   \n",
       "121                         GENCODE-annotated transcripts               NaN   \n",
       "516                         GENCODE-annotated transcripts               NaN   \n",
       "1482                        GENCODE-annotated transcripts               NaN   \n",
       "1504                        GENCODE-annotated transcripts               NaN   \n",
       "...                                                   ...               ...   \n",
       "376259  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376300  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376328  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376330  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376331  Unannotated transcripts composed entirely of a...               NaN   \n",
       "\n",
       "       Strand                               summarized_category  \\\n",
       "4           -                     GENCODE-annotated transcripts   \n",
       "121         -                     GENCODE-annotated transcripts   \n",
       "516         -                     GENCODE-annotated transcripts   \n",
       "1482        -                     GENCODE-annotated transcripts   \n",
       "1504        -                     GENCODE-annotated transcripts   \n",
       "...       ...                                               ...   \n",
       "376259      +  Unannotated transcripts with all exons annotated   \n",
       "376300      -  Unannotated transcripts with all exons annotated   \n",
       "376328      -  Unannotated transcripts with all exons annotated   \n",
       "376330      -  Unannotated transcripts with all exons annotated   \n",
       "376331      -  Unannotated transcripts with all exons annotated   \n",
       "\n",
       "                   ref_id  stripped_ref_id   Gene stable ID Gene name  \\\n",
       "4       ENST00000771587.1  ENST00000771587  ENSG00000300424       NaN   \n",
       "121     ENST00000289779.7  ENST00000289779  ENSG00000270149       NaN   \n",
       "516     ENST00000362058.2  ENST00000362058  ENSG00000291072       NaN   \n",
       "1482    ENST00000412483.1  ENST00000412483  ENSG00000233332       NaN   \n",
       "1504    ENST00000415019.1  ENST00000415019  ENSG00000176320       NaN   \n",
       "...                   ...              ...              ...       ...   \n",
       "376259  ENST00000466616.6  ENST00000466616  ENSG00000271147       NaN   \n",
       "376300  ENST00000796795.1  ENST00000796795  ENSG00000291032       NaN   \n",
       "376328  ENST00000737339.1  ENST00000737339  ENSG00000291034       NaN   \n",
       "376330  ENST00000796810.1  ENST00000796810  ENSG00000291032       NaN   \n",
       "376331  ENST00000737331.1  ENST00000737331  ENSG00000291034       NaN   \n",
       "\n",
       "             Gene type  \n",
       "4               lncRNA  \n",
       "121     protein_coding  \n",
       "516             lncRNA  \n",
       "1482            lncRNA  \n",
       "1504            lncRNA  \n",
       "...                ...  \n",
       "376259          lncRNA  \n",
       "376300          lncRNA  \n",
       "376328          lncRNA  \n",
       "376330          lncRNA  \n",
       "376331          lncRNA  \n",
       "\n",
       "[20239 rows x 13 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[combined_df['Gene name'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7783dc97-b6bc-4689-9d10-a1b0f943da86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ENST00000351839', 'ENST00000391947', 'ENST00000636394',\n",
       "       'ENST00000334318', 'ENST00000418951', 'ENST00000617484',\n",
       "       'ENST00000612355', 'ENST00000458258', 'ENST00000756245',\n",
       "       'ENST00000797550'], dtype=object)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No ref gene could be assigned but tx have exon overlap with GENCODE\n",
    "combined_df[(combined_df[\"Gene type\"].isna()) & (combined_df[\"category\"] != \"Fully unannotated transcripts without any overlapping exons\")][\"stripped_ref_id\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e1d714f8-c4a1-42e8-b13f-c6d314b43878",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual assignment\n",
    "manual_gene_info = {\n",
    "    \"ENST00000351839\": {\"Gene stable ID\": \"ENSG00000165119\", \"Gene name\": \"HNRNPK\", \"Gene type\": \"protein_coding\"},\n",
    "    \"ENST00000636394\": {\"Gene stable ID\": \"ENSG00000279170\", \"Gene name\": \"TSTD3\", \"Gene type\": \"protein_coding\"},\n",
    "    \"ENST00000617484\": {\"Gene stable ID\": \"ENSG00000109819\", \"Gene name\": \"PPARGC1A\", \"Gene type\": \"protein_coding\"},\n",
    "    \"ENST00000756245\": {\"Gene stable ID\": \"ENSG00000298529\", \"Gene name\": \"ENSG00000298529\", \"Gene type\": \"lncRNA\"},\n",
    "\n",
    "    \"ENST00000391947\": {\"Gene stable ID\": \"ENSG00000198625\", \"Gene name\": \"MDM4\", \"Gene type\": \"protein_coding\"},\n",
    "    \"ENST00000334318\": {\"Gene stable ID\": \"ENSG00000119314\", \"Gene name\": \"PTBP3\", \"Gene type\": \"protein_coding\"},\n",
    "    \"ENST00000612355\": {\"Gene stable ID\": \"ENSG00000109819\", \"Gene name\": \"PPARGC1A\", \"Gene type\": \"protein_coding\"},\n",
    "    \"ENST00000797550\": {\"Gene stable ID\": \"ENSG00000303858\", \"Gene name\": \"ENSG00000303858\", \"Gene type\": \"lncRNA\"},\n",
    "\n",
    "    \"ENST00000418951\": {\"Gene stable ID\": \"ENSG00000099968\", \"Gene name\": \"BCL2L13\", \"Gene type\": \"protein_coding\"},\n",
    "    \"ENST00000458258\": {\"Gene stable ID\": \"ENSG00000119314\", \"Gene name\": \"PTBP3\", \"Gene type\": \"protein_coding\"} \n",
    "}\n",
    "\n",
    "for ref_id, gene_info in manual_gene_info.items():\n",
    "    mask = combined_df[\"stripped_ref_id\"] == ref_id\n",
    "    for col, value in gene_info.items():\n",
    "        combined_df.loc[mask, col] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "8bb6a265-ac16-49a8-99c7-49cd206353f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['None'], dtype=object)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[(combined_df[\"Gene type\"].isna()) & (combined_df[\"category\"] != \"Fully unannotated transcripts with non-overlapping exons\")][\"stripped_ref_id\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6f5c74c3-6114-4e31-b7b8-03122ceb531f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(61077274, 61077628), (61077916, 61079203)]</td>\n",
       "      <td>[(61077326, 61077628), (61077916, 61079231)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000699992.1</td>\n",
       "      <td>ENST00000699992</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160261734, 160261922), (160281430, 160281892)]</td>\n",
       "      <td>[(160261682, 160261922), (160281430, 160281935)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000756371.1</td>\n",
       "      <td>ENST00000756371</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>ENST00000676153</td>\n",
       "      <td>ENSG00000133019</td>\n",
       "      <td>CHRM3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCT00000022754.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(23191893, 23195001), (23217291, 23217499)]</td>\n",
       "      <td>[(23191895, 23195001), (23217291, 23217502)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>ENST00000374619</td>\n",
       "      <td>ENSG00000179546</td>\n",
       "      <td>HTR1D</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "      <td>ENST00000771587</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id   chr                                              exons  \\\n",
       "0  ENCT00000006334.1  chr1       [(61077274, 61077628), (61077916, 61079203)]   \n",
       "1  ENCT00000013056.1  chr1   [(160261734, 160261922), (160281430, 160281892)]   \n",
       "2  ENCT00000019620.1  chr1  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3  ENCT00000022754.1  chr1       [(23191893, 23195001), (23217291, 23217499)]   \n",
       "4  ENCT00000039204.1  chr1   [(233319835, 233321155), (233327229, 233327357)]   \n",
       "\n",
       "                                          matched_to  \\\n",
       "0       [(61077326, 61077628), (61077916, 61079231)]   \n",
       "1   [(160261682, 160261922), (160281430, 160281935)]   \n",
       "2  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3       [(23191895, 23195001), (23217291, 23217502)]   \n",
       "4   [(233319834, 233321155), (233327229, 233327455)]   \n",
       "\n",
       "                        category unannotated_exons Strand  \\\n",
       "0  GENCODE-annotated transcripts               NaN      +   \n",
       "1  GENCODE-annotated transcripts               NaN      +   \n",
       "2  GENCODE-annotated transcripts               NaN      +   \n",
       "3  GENCODE-annotated transcripts               NaN      -   \n",
       "4  GENCODE-annotated transcripts               NaN      -   \n",
       "\n",
       "             summarized_category             ref_id  stripped_ref_id  \\\n",
       "0  GENCODE-annotated transcripts  ENST00000699992.1  ENST00000699992   \n",
       "1  GENCODE-annotated transcripts  ENST00000756371.1  ENST00000756371   \n",
       "2  GENCODE-annotated transcripts  ENST00000676153.1  ENST00000676153   \n",
       "3  GENCODE-annotated transcripts  ENST00000374619.2  ENST00000374619   \n",
       "4  GENCODE-annotated transcripts  ENST00000771587.1  ENST00000771587   \n",
       "\n",
       "    Gene stable ID Gene name       Gene type   gene_category  \n",
       "0  ENSG00000162599      NFIA  protein_coding  protein_coding  \n",
       "1  ENSG00000228606  DCAF8-DT          lncRNA          lncRNA  \n",
       "2  ENSG00000133019     CHRM3  protein_coding  protein_coding  \n",
       "3  ENSG00000179546     HTR1D  protein_coding  protein_coding  \n",
       "4  ENSG00000300424       NaN          lncRNA          lncRNA  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_type_map = {\n",
    "    'protein_coding': 'protein_coding',\n",
    "    'lncRNA': 'lncRNA'\n",
    "}\n",
    "\n",
    "# Apply the mapping to make a new category column\n",
    "combined_df['gene_category'] = combined_df['Gene type'].map(gene_type_map)\n",
    "\n",
    "# Fill the rest: NaN --> 'unmapped', others --> 'other_noncoding'\n",
    "combined_df['gene_category'] = combined_df['gene_category'].fillna(\n",
    "    combined_df['Gene type'].apply(lambda x: 'unmapped' if pd.isna(x) else 'other_noncoding')\n",
    ")\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6c97796a-ce0f-4fe8-9de5-1a2ff64fba85",
   "metadata": {},
   "outputs": [],
   "source": [
    "perfect_matches = combined_df[combined_df['category'] == \"GENCODE-annotated transcripts\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8ca76ec9-bbbd-4762-9813-235fb6679eb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11512</th>\n",
       "      <td>ENST00000204566.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>[(64963021, 64963736), (64965320, 64965460), (...</td>\n",
       "      <td>[(64963022, 64963736), (64965320, 64965460), (...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000204566.7</td>\n",
       "      <td>ENST00000204566</td>\n",
       "      <td>ENSG00000090487</td>\n",
       "      <td>SPG21</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12724</th>\n",
       "      <td>MICT00000118180.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>[(64963022, 64963736), (64965320, 64965460), (...</td>\n",
       "      <td>[(64963022, 64963736), (64965320, 64965460), (...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000204566.7</td>\n",
       "      <td>ENST00000204566</td>\n",
       "      <td>ENSG00000090487</td>\n",
       "      <td>SPG21</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           transcript_id    chr  \\\n",
       "11512  ENST00000204566.2  chr15   \n",
       "12724  MICT00000118180.1  chr15   \n",
       "\n",
       "                                                   exons  \\\n",
       "11512  [(64963021, 64963736), (64965320, 64965460), (...   \n",
       "12724  [(64963022, 64963736), (64965320, 64965460), (...   \n",
       "\n",
       "                                              matched_to  \\\n",
       "11512  [(64963022, 64963736), (64965320, 64965460), (...   \n",
       "12724  [(64963022, 64963736), (64965320, 64965460), (...   \n",
       "\n",
       "                            category unannotated_exons Strand  \\\n",
       "11512  GENCODE-annotated transcripts               NaN      -   \n",
       "12724  GENCODE-annotated transcripts               NaN      -   \n",
       "\n",
       "                 summarized_category             ref_id  stripped_ref_id  \\\n",
       "11512  GENCODE-annotated transcripts  ENST00000204566.7  ENST00000204566   \n",
       "12724  GENCODE-annotated transcripts  ENST00000204566.7  ENST00000204566   \n",
       "\n",
       "        Gene stable ID Gene name       Gene type   gene_category  \n",
       "11512  ENSG00000090487     SPG21  protein_coding  protein_coding  \n",
       "12724  ENSG00000090487     SPG21  protein_coding  protein_coding  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find cases where multiple transcripts have same GENCODE perfect match (redundancy)\n",
    "perfect_matches[(perfect_matches['ref_id'].duplicated(keep=False))].sort_values(\"ref_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "283f7243-70cd-45ec-949b-0a95e77087c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop the transcript with worse match to GENCODE ref\n",
    "combined_df = combined_df[combined_df['transcript_id'] != 'ENST00000204566.2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d441f9f7-1743-4ccb-a212-a179d6c4ffeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill NA gene names with the gene ID\n",
    "combined_df['Gene name'] = combined_df['Gene name'].fillna(combined_df['Gene stable ID'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c794b9-dadb-44e8-91de-f5b36edb27ee",
   "metadata": {},
   "source": [
    "## Add transcript expression info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "62099965-f294-44f7-bc36-5f9c2abc2a32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>tx_start</th>\n",
       "      <th>tx_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Gene class</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>...</th>\n",
       "      <th>low_tpm</th>\n",
       "      <th>peak</th>\n",
       "      <th>TSS expression high glucose (Salmon, TPM)</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>Distance to closest Gencode TSS</th>\n",
       "      <th>name</th>\n",
       "      <th>new_1st_exon</th>\n",
       "      <th>new_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000423562.1</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>...</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14359</td>\n",
       "      <td>29350</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037199</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>...</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14359</td>\n",
       "      <td>29350</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037200</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>...</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14360</td>\n",
       "      <td>29364</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037201</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>...</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29346</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037202</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>...</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  tx_start  tx_end strand      transcript_id          gene_id  \\\n",
       "0       chr1     14362   29370      -  ENST00000423562.1  ENSG00000227232   \n",
       "1       chr1     14359   29350      -        PBT00037199  ENSG00000227232   \n",
       "2       chr1     14359   29350      -        PBT00037200  ENSG00000227232   \n",
       "3       chr1     14360   29364      -        PBT00037201  ENSG00000227232   \n",
       "4       chr1     14362   29346      -        PBT00037202  ENSG00000227232   \n",
       "\n",
       "  gene_name Gene class   gene_type transcript_type  ...    low_tpm    peak  \\\n",
       "0    WASH7P   misc RNA  pseudogene  new_transcript  ...  17.856425  Robust   \n",
       "1    WASH7P   misc RNA  pseudogene  new_transcript  ...  17.856425  Robust   \n",
       "2    WASH7P   misc RNA  pseudogene  new_transcript  ...  17.856425  Robust   \n",
       "3    WASH7P   misc RNA  pseudogene  new_transcript  ...  17.856425  Robust   \n",
       "4    WASH7P   misc RNA  pseudogene  new_transcript  ...  17.856425  Robust   \n",
       "\n",
       "   TSS expression high glucose (Salmon, TPM)  Relative TSS usage    TSS_type  \\\n",
       "0                                  27.348529                 1.0  unique TSS   \n",
       "1                                  27.348529                 1.0  unique TSS   \n",
       "2                                  27.348529                 1.0  unique TSS   \n",
       "3                                  27.348529                 1.0  unique TSS   \n",
       "4                                  27.348529                 1.0  unique TSS   \n",
       "\n",
       "  distance_to_closest_TSS  Distance to closest Gencode TSS  \\\n",
       "0                     211                           >100bp   \n",
       "1                     211                           >100bp   \n",
       "2                     211                           >100bp   \n",
       "3                     211                           >100bp   \n",
       "4                     211                           >100bp   \n",
       "\n",
       "                      name  new_1st_exon                 new_name  \n",
       "0  chr1:29336-29359_Peak_1           yes  chr1:29336-29359_Peak_1  \n",
       "1  chr1:29336-29359_Peak_1           yes  chr1:29336-29359_Peak_1  \n",
       "2  chr1:29336-29359_Peak_1           yes  chr1:29336-29359_Peak_1  \n",
       "3  chr1:29336-29359_Peak_1           yes  chr1:29336-29359_Peak_1  \n",
       "4  chr1:29336-29359_Peak_1           yes  chr1:29336-29359_Peak_1  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm_df = pd.read_csv( '../data_processed/TSS_info_w_lift.tsv', sep='\\t')\n",
    "tpm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a6b5884d-ee50-4e28-90cd-3c713e00fe0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "d9f2b292-45b6-4494-b93e-ff9cdc5c5deb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add peak_name to combined_df\n",
    "combined_df = combined_df.merge(\n",
    "    tpm_df[['transcript_id', 'new_name', 'peak']].rename(columns={'transcript_id': 'merge_key', 'new_name': 'peak_name', 'peak': 'peak_type'}),\n",
    "    how='left',\n",
    "    left_on='transcript_id',\n",
    "    right_on='merge_key'\n",
    ").drop(columns='merge_key')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "5774af66-91f2-4ddf-b577-6ea7ef2abaed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign new Gene ID/name to unmapped tx\n",
    "mask = combined_df['Gene stable ID'].isna()\n",
    "\n",
    "peak_ids = combined_df.loc[mask, 'peak_name'].str.extract(r'(Peak_\\d+)', expand=False)\n",
    "combined_df.loc[mask, 'Gene stable ID'] = peak_ids\n",
    "combined_df.loc[mask, 'Gene name'] = peak_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "bb112976-8f2b-4010-83f6-fe02fa77b5eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, chr, exons, matched_to, category, unannotated_exons, Strand, summarized_category, ref_id, stripped_ref_id, Gene stable ID, Gene name, Gene type, gene_category, peak_name, peak_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[combined_df[\"Gene name\"].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "74fa75e1-a3b1-45a7-be71-d63fcaa21202",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add back the TPMs per transcript \n",
    "tpm_df['mean_TPM_ref'] = tpm_df['relative_tx_exp'] * tpm_df['gene_exp_ref'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c3526e7b-c394-4c6a-adf5-eefb50c193ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>tx_start</th>\n",
       "      <th>tx_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Gene class</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>protein_id_orf_summary</th>\n",
       "      <th>transcript_classification</th>\n",
       "      <th>Transcript expression high glucose (TPM)</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>CAGE_ID</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>high_tpm</th>\n",
       "      <th>low_tpm</th>\n",
       "      <th>peak</th>\n",
       "      <th>TSS expression high glucose (Salmon, TPM)</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>Distance to closest Gencode TSS</th>\n",
       "      <th>name</th>\n",
       "      <th>new_1st_exon</th>\n",
       "      <th>new_name</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000423562.1</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>UniProtKB=B3KS13</td>\n",
       "      <td>Known Protein coding</td>\n",
       "      <td>0.219906</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>chr1:29336-29359:-</td>\n",
       "      <td>29336</td>\n",
       "      <td>29359</td>\n",
       "      <td>17.58083</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>0.219906</td>\n",
       "      <td>-2.185038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14359</td>\n",
       "      <td>29350</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037199</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novel Protein coding not overlapping annotated...</td>\n",
       "      <td>3.422413</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>0.125141</td>\n",
       "      <td>chr1:29336-29359:-</td>\n",
       "      <td>29336</td>\n",
       "      <td>29359</td>\n",
       "      <td>17.58083</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>3.422413</td>\n",
       "      <td>1.775014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14359</td>\n",
       "      <td>29350</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037200</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>Closest_UniProtKB=A9QQ17, percentage_UniProtKB...</td>\n",
       "      <td>Novel Protein coding not overlapping annotated...</td>\n",
       "      <td>0.172042</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>chr1:29336-29359:-</td>\n",
       "      <td>29336</td>\n",
       "      <td>29359</td>\n",
       "      <td>17.58083</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>0.172042</td>\n",
       "      <td>-2.539164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14360</td>\n",
       "      <td>29364</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037201</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non coding</td>\n",
       "      <td>0.071035</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>0.002597</td>\n",
       "      <td>chr1:29336-29359:-</td>\n",
       "      <td>29336</td>\n",
       "      <td>29359</td>\n",
       "      <td>17.58083</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>0.071035</td>\n",
       "      <td>-3.815321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29346</td>\n",
       "      <td>-</td>\n",
       "      <td>PBT00037202</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>misc RNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>new_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non coding</td>\n",
       "      <td>0.322260</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>0.011783</td>\n",
       "      <td>chr1:29336-29359:-</td>\n",
       "      <td>29336</td>\n",
       "      <td>29359</td>\n",
       "      <td>17.58083</td>\n",
       "      <td>17.856425</td>\n",
       "      <td>Robust</td>\n",
       "      <td>27.348529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>211</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>yes</td>\n",
       "      <td>chr1:29336-29359_Peak_1</td>\n",
       "      <td>0.322260</td>\n",
       "      <td>-1.633703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  tx_start  tx_end strand      transcript_id          gene_id  \\\n",
       "0       chr1     14362   29370      -  ENST00000423562.1  ENSG00000227232   \n",
       "1       chr1     14359   29350      -        PBT00037199  ENSG00000227232   \n",
       "2       chr1     14359   29350      -        PBT00037200  ENSG00000227232   \n",
       "3       chr1     14360   29364      -        PBT00037201  ENSG00000227232   \n",
       "4       chr1     14362   29346      -        PBT00037202  ENSG00000227232   \n",
       "\n",
       "  gene_name Gene class   gene_type transcript_type  \\\n",
       "0    WASH7P   misc RNA  pseudogene  new_transcript   \n",
       "1    WASH7P   misc RNA  pseudogene  new_transcript   \n",
       "2    WASH7P   misc RNA  pseudogene  new_transcript   \n",
       "3    WASH7P   misc RNA  pseudogene  new_transcript   \n",
       "4    WASH7P   misc RNA  pseudogene  new_transcript   \n",
       "\n",
       "                              protein_id_orf_summary  \\\n",
       "0                                   UniProtKB=B3KS13   \n",
       "1                                                NaN   \n",
       "2  Closest_UniProtKB=A9QQ17, percentage_UniProtKB...   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "                           transcript_classification  \\\n",
       "0                               Known Protein coding   \n",
       "1  Novel Protein coding not overlapping annotated...   \n",
       "2  Novel Protein coding not overlapping annotated...   \n",
       "3                                         Non coding   \n",
       "4                                         Non coding   \n",
       "\n",
       "   Transcript expression high glucose (TPM)  gene_exp_ref  relative_tx_exp  \\\n",
       "0                                  0.219906     27.348529         0.008041   \n",
       "1                                  3.422413     27.348529         0.125141   \n",
       "2                                  0.172042     27.348529         0.006291   \n",
       "3                                  0.071035     27.348529         0.002597   \n",
       "4                                  0.322260     27.348529         0.011783   \n",
       "\n",
       "              CAGE_ID  TSS_start  TSS_end  high_tpm    low_tpm    peak  \\\n",
       "0  chr1:29336-29359:-      29336    29359  17.58083  17.856425  Robust   \n",
       "1  chr1:29336-29359:-      29336    29359  17.58083  17.856425  Robust   \n",
       "2  chr1:29336-29359:-      29336    29359  17.58083  17.856425  Robust   \n",
       "3  chr1:29336-29359:-      29336    29359  17.58083  17.856425  Robust   \n",
       "4  chr1:29336-29359:-      29336    29359  17.58083  17.856425  Robust   \n",
       "\n",
       "   TSS expression high glucose (Salmon, TPM)  Relative TSS usage    TSS_type  \\\n",
       "0                                  27.348529                 1.0  unique TSS   \n",
       "1                                  27.348529                 1.0  unique TSS   \n",
       "2                                  27.348529                 1.0  unique TSS   \n",
       "3                                  27.348529                 1.0  unique TSS   \n",
       "4                                  27.348529                 1.0  unique TSS   \n",
       "\n",
       "   distance_to_closest_TSS Distance to closest Gencode TSS  \\\n",
       "0                      211                          >100bp   \n",
       "1                      211                          >100bp   \n",
       "2                      211                          >100bp   \n",
       "3                      211                          >100bp   \n",
       "4                      211                          >100bp   \n",
       "\n",
       "                      name new_1st_exon                 new_name  \\\n",
       "0  chr1:29336-29359_Peak_1          yes  chr1:29336-29359_Peak_1   \n",
       "1  chr1:29336-29359_Peak_1          yes  chr1:29336-29359_Peak_1   \n",
       "2  chr1:29336-29359_Peak_1          yes  chr1:29336-29359_Peak_1   \n",
       "3  chr1:29336-29359_Peak_1          yes  chr1:29336-29359_Peak_1   \n",
       "4  chr1:29336-29359_Peak_1          yes  chr1:29336-29359_Peak_1   \n",
       "\n",
       "   mean_TPM_ref  log2_tx_exp  \n",
       "0      0.219906    -2.185038  \n",
       "1      3.422413     1.775014  \n",
       "2      0.172042    -2.539164  \n",
       "3      0.071035    -3.815321  \n",
       "4      0.322260    -1.633703  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove rows with no expression and log transform\n",
    "tpm_df = tpm_df[tpm_df['mean_TPM_ref'] != 0].copy()\n",
    "tpm_df['log2_tx_exp'] = np.log2(tpm_df['mean_TPM_ref'])\n",
    "tpm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f876c40a-8ff1-4d8d-8a21-597b2ea03658",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge tx tpm info with comparison results\n",
    "combined_df = combined_df.merge(\n",
    "    tpm_df[['transcript_id', 'mean_TPM_ref', 'log2_tx_exp']].rename(columns={'transcript_id': 'merge_key'}),\n",
    "    how='left',\n",
    "    left_on='transcript_id',\n",
    "    right_on='merge_key'\n",
    ").drop(columns='merge_key')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "dcec4e83-6bc4-441d-b20e-bd3c81394204",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376342"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(combined_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "05d08c2d-8b46-4f92-a1c9-61b882fbd6cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the gene expression (sum of all the transcripts from the same gene) \n",
    "f = lambda x: x.sum()\n",
    "combined_df['gene_exp_ref'] = combined_df.groupby('Gene stable ID')['mean_TPM_ref'].transform(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f5a23aec-6c96-44fa-ada5-eb2a22946fe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the transcript relative expression \n",
    "combined_df['relative_tx_exp'] = combined_df['mean_TPM_ref'] / combined_df['gene_exp_ref']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e575dd89-229d-4a56-9a21-55e9c3749d83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the sum of TPM expression of all the transcripts from one gene starting from the same TSS\n",
    "f = lambda x: x.sum()\n",
    "combined_df['TSS_TPM_per_gene'] = combined_df.groupby(['Gene stable ID','peak_name'])['mean_TPM_ref'].transform(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "945ef5fd-19c8-4bc0-ac51-b25c7ea5c8ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df['Relative TSS usage'] = combined_df['TSS_TPM_per_gene']/combined_df['gene_exp_ref']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "2b3774fd-fe43-4449-ae42-a0f83241ccb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, chr, exons, matched_to, category, unannotated_exons, Strand, summarized_category, ref_id, stripped_ref_id, Gene stable ID, Gene name, Gene type, gene_category, peak_name, peak_type, mean_TPM_ref, log2_tx_exp, gene_exp_ref, relative_tx_exp, TSS_TPM_per_gene, Relative TSS usage]\n",
       "Index: []"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# control that no gene has an total relative expression above 1\n",
    "combined_df[(combined_df['Relative TSS usage'] > 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "5008238b-b4dd-44b2-83f4-075e011d2b6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_rank</th>\n",
       "      <th>TSS_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(61077274, 61077628), (61077916, 61079203)]</td>\n",
       "      <td>[(61077326, 61077628), (61077916, 61079231)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000699992.1</td>\n",
       "      <td>ENST00000699992</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:61077221-61077380_Peak_3111</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.178215</td>\n",
       "      <td>-2.488310</td>\n",
       "      <td>4.820350</td>\n",
       "      <td>0.036971</td>\n",
       "      <td>0.213765</td>\n",
       "      <td>0.044346</td>\n",
       "      <td>4.0</td>\n",
       "      <td>minor TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160261734, 160261922), (160281430, 160281892)]</td>\n",
       "      <td>[(160261682, 160261922), (160281430, 160281935)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000756371.1</td>\n",
       "      <td>ENST00000756371</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:160261726-160261763_Peak_5604</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.358414</td>\n",
       "      <td>-1.480300</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>0.481821</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>ENST00000676153</td>\n",
       "      <td>ENSG00000133019</td>\n",
       "      <td>CHRM3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:239386576-239386587_Peak_7900</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.468766</td>\n",
       "      <td>-1.093061</td>\n",
       "      <td>3.127170</td>\n",
       "      <td>0.149901</td>\n",
       "      <td>1.421490</td>\n",
       "      <td>0.454561</td>\n",
       "      <td>1.0</td>\n",
       "      <td>main TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCT00000022754.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(23191893, 23195001), (23217291, 23217499)]</td>\n",
       "      <td>[(23191895, 23195001), (23217291, 23217502)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>ENST00000374619</td>\n",
       "      <td>ENSG00000179546</td>\n",
       "      <td>HTR1D</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:23217498-23217499_Peak_1212</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.064644</td>\n",
       "      <td>-3.951339</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>0.553728</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "      <td>ENST00000771587</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:233327323-233327461_Peak_7654</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.616169</td>\n",
       "      <td>-0.698603</td>\n",
       "      <td>2.291033</td>\n",
       "      <td>0.268948</td>\n",
       "      <td>2.093885</td>\n",
       "      <td>0.913948</td>\n",
       "      <td>1.0</td>\n",
       "      <td>main TSS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id   chr                                              exons  \\\n",
       "0  ENCT00000006334.1  chr1       [(61077274, 61077628), (61077916, 61079203)]   \n",
       "1  ENCT00000013056.1  chr1   [(160261734, 160261922), (160281430, 160281892)]   \n",
       "2  ENCT00000019620.1  chr1  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3  ENCT00000022754.1  chr1       [(23191893, 23195001), (23217291, 23217499)]   \n",
       "4  ENCT00000039204.1  chr1   [(233319835, 233321155), (233327229, 233327357)]   \n",
       "\n",
       "                                          matched_to  \\\n",
       "0       [(61077326, 61077628), (61077916, 61079231)]   \n",
       "1   [(160261682, 160261922), (160281430, 160281935)]   \n",
       "2  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3       [(23191895, 23195001), (23217291, 23217502)]   \n",
       "4   [(233319834, 233321155), (233327229, 233327455)]   \n",
       "\n",
       "                        category unannotated_exons Strand  \\\n",
       "0  GENCODE-annotated transcripts               NaN      +   \n",
       "1  GENCODE-annotated transcripts               NaN      +   \n",
       "2  GENCODE-annotated transcripts               NaN      +   \n",
       "3  GENCODE-annotated transcripts               NaN      -   \n",
       "4  GENCODE-annotated transcripts               NaN      -   \n",
       "\n",
       "             summarized_category             ref_id  stripped_ref_id  \\\n",
       "0  GENCODE-annotated transcripts  ENST00000699992.1  ENST00000699992   \n",
       "1  GENCODE-annotated transcripts  ENST00000756371.1  ENST00000756371   \n",
       "2  GENCODE-annotated transcripts  ENST00000676153.1  ENST00000676153   \n",
       "3  GENCODE-annotated transcripts  ENST00000374619.2  ENST00000374619   \n",
       "4  GENCODE-annotated transcripts  ENST00000771587.1  ENST00000771587   \n",
       "\n",
       "    Gene stable ID        Gene name       Gene type   gene_category  \\\n",
       "0  ENSG00000162599             NFIA  protein_coding  protein_coding   \n",
       "1  ENSG00000228606         DCAF8-DT          lncRNA          lncRNA   \n",
       "2  ENSG00000133019            CHRM3  protein_coding  protein_coding   \n",
       "3  ENSG00000179546            HTR1D  protein_coding  protein_coding   \n",
       "4  ENSG00000300424  ENSG00000300424          lncRNA          lncRNA   \n",
       "\n",
       "                            peak_name   peak_type  mean_TPM_ref  log2_tx_exp  \\\n",
       "0    chr1:61077221-61077380_Peak_3111  Permissive      0.178215    -2.488310   \n",
       "1  chr1:160261726-160261763_Peak_5604  Permissive      0.358414    -1.480300   \n",
       "2  chr1:239386576-239386587_Peak_7900      Robust      0.468766    -1.093061   \n",
       "3    chr1:23217498-23217499_Peak_1212  Permissive      0.064644    -3.951339   \n",
       "4  chr1:233327323-233327461_Peak_7654  Permissive      0.616169    -0.698603   \n",
       "\n",
       "   gene_exp_ref  relative_tx_exp  TSS_TPM_per_gene  Relative TSS usage  \\\n",
       "0      4.820350         0.036971          0.213765            0.044346   \n",
       "1      0.743874         0.481821          0.743874            1.000000   \n",
       "2      3.127170         0.149901          1.421490            0.454561   \n",
       "3      0.116743         0.553728          0.116743            1.000000   \n",
       "4      2.291033         0.268948          2.093885            0.913948   \n",
       "\n",
       "   TSS_rank    TSS_type  \n",
       "0       4.0   minor TSS  \n",
       "1       1.0  unique TSS  \n",
       "2       1.0    main TSS  \n",
       "3       1.0  unique TSS  \n",
       "4       1.0    main TSS  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"TSS_rank\"] = combined_df.groupby(\"Gene stable ID\")[\"Relative TSS usage\"].rank(method=\"dense\", ascending=False)\n",
    "\n",
    "def get_TSS_type(row):\n",
    "    if row[\"Relative TSS usage\"] == 1:\n",
    "        return \"unique TSS\"\n",
    "    elif row[\"TSS_rank\"] == 1:\n",
    "        return \"main TSS\"\n",
    "    elif row[\"TSS_rank\"] == 2:\n",
    "        return \"secondary TSS\"\n",
    "    else:\n",
    "        return \"minor TSS\"\n",
    "\n",
    "combined_df[\"TSS_type\"] = combined_df.apply(get_TSS_type, axis=1)\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "5e05dec9-51f7-4fc4-b9d7-678e37dd9bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df = combined_df.drop(columns=[\"TSS_rank\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53208243-b452-4f83-b7e5-dd3020c83809",
   "metadata": {},
   "source": [
    "## Finding unannotated TSSs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "2abd7639-37c0-464a-a848-8fbe261b6803",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(61077274, 61077628), (61077916, 61079203)]</td>\n",
       "      <td>[(61077326, 61077628), (61077916, 61079231)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000699992.1</td>\n",
       "      <td>ENST00000699992</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:61077221-61077380_Peak_3111</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.178215</td>\n",
       "      <td>-2.488310</td>\n",
       "      <td>4.820350</td>\n",
       "      <td>0.036971</td>\n",
       "      <td>0.213765</td>\n",
       "      <td>0.044346</td>\n",
       "      <td>minor TSS</td>\n",
       "      <td>61077221</td>\n",
       "      <td>61077380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160261734, 160261922), (160281430, 160281892)]</td>\n",
       "      <td>[(160261682, 160261922), (160281430, 160281935)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000756371.1</td>\n",
       "      <td>ENST00000756371</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:160261726-160261763_Peak_5604</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.358414</td>\n",
       "      <td>-1.480300</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>0.481821</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>160261726</td>\n",
       "      <td>160261763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>ENST00000676153</td>\n",
       "      <td>ENSG00000133019</td>\n",
       "      <td>CHRM3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:239386576-239386587_Peak_7900</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.468766</td>\n",
       "      <td>-1.093061</td>\n",
       "      <td>3.127170</td>\n",
       "      <td>0.149901</td>\n",
       "      <td>1.421490</td>\n",
       "      <td>0.454561</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>239386576</td>\n",
       "      <td>239386587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCT00000022754.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(23191893, 23195001), (23217291, 23217499)]</td>\n",
       "      <td>[(23191895, 23195001), (23217291, 23217502)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>ENST00000374619</td>\n",
       "      <td>ENSG00000179546</td>\n",
       "      <td>HTR1D</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:23217498-23217499_Peak_1212</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.064644</td>\n",
       "      <td>-3.951339</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>0.553728</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>23217498</td>\n",
       "      <td>23217499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "      <td>ENST00000771587</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:233327323-233327461_Peak_7654</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.616169</td>\n",
       "      <td>-0.698603</td>\n",
       "      <td>2.291033</td>\n",
       "      <td>0.268948</td>\n",
       "      <td>2.093885</td>\n",
       "      <td>0.913948</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>233327323</td>\n",
       "      <td>233327461</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id   chr                                              exons  \\\n",
       "0  ENCT00000006334.1  chr1       [(61077274, 61077628), (61077916, 61079203)]   \n",
       "1  ENCT00000013056.1  chr1   [(160261734, 160261922), (160281430, 160281892)]   \n",
       "2  ENCT00000019620.1  chr1  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3  ENCT00000022754.1  chr1       [(23191893, 23195001), (23217291, 23217499)]   \n",
       "4  ENCT00000039204.1  chr1   [(233319835, 233321155), (233327229, 233327357)]   \n",
       "\n",
       "                                          matched_to  \\\n",
       "0       [(61077326, 61077628), (61077916, 61079231)]   \n",
       "1   [(160261682, 160261922), (160281430, 160281935)]   \n",
       "2  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3       [(23191895, 23195001), (23217291, 23217502)]   \n",
       "4   [(233319834, 233321155), (233327229, 233327455)]   \n",
       "\n",
       "                        category unannotated_exons Strand  \\\n",
       "0  GENCODE-annotated transcripts               NaN      +   \n",
       "1  GENCODE-annotated transcripts               NaN      +   \n",
       "2  GENCODE-annotated transcripts               NaN      +   \n",
       "3  GENCODE-annotated transcripts               NaN      -   \n",
       "4  GENCODE-annotated transcripts               NaN      -   \n",
       "\n",
       "             summarized_category             ref_id  stripped_ref_id  \\\n",
       "0  GENCODE-annotated transcripts  ENST00000699992.1  ENST00000699992   \n",
       "1  GENCODE-annotated transcripts  ENST00000756371.1  ENST00000756371   \n",
       "2  GENCODE-annotated transcripts  ENST00000676153.1  ENST00000676153   \n",
       "3  GENCODE-annotated transcripts  ENST00000374619.2  ENST00000374619   \n",
       "4  GENCODE-annotated transcripts  ENST00000771587.1  ENST00000771587   \n",
       "\n",
       "    Gene stable ID        Gene name       Gene type   gene_category  \\\n",
       "0  ENSG00000162599             NFIA  protein_coding  protein_coding   \n",
       "1  ENSG00000228606         DCAF8-DT          lncRNA          lncRNA   \n",
       "2  ENSG00000133019            CHRM3  protein_coding  protein_coding   \n",
       "3  ENSG00000179546            HTR1D  protein_coding  protein_coding   \n",
       "4  ENSG00000300424  ENSG00000300424          lncRNA          lncRNA   \n",
       "\n",
       "                            peak_name   peak_type  mean_TPM_ref  log2_tx_exp  \\\n",
       "0    chr1:61077221-61077380_Peak_3111  Permissive      0.178215    -2.488310   \n",
       "1  chr1:160261726-160261763_Peak_5604  Permissive      0.358414    -1.480300   \n",
       "2  chr1:239386576-239386587_Peak_7900      Robust      0.468766    -1.093061   \n",
       "3    chr1:23217498-23217499_Peak_1212  Permissive      0.064644    -3.951339   \n",
       "4  chr1:233327323-233327461_Peak_7654  Permissive      0.616169    -0.698603   \n",
       "\n",
       "   gene_exp_ref  relative_tx_exp  TSS_TPM_per_gene  Relative TSS usage  \\\n",
       "0      4.820350         0.036971          0.213765            0.044346   \n",
       "1      0.743874         0.481821          0.743874            1.000000   \n",
       "2      3.127170         0.149901          1.421490            0.454561   \n",
       "3      0.116743         0.553728          0.116743            1.000000   \n",
       "4      2.291033         0.268948          2.093885            0.913948   \n",
       "\n",
       "     TSS_type  TSS_start    TSS_end  \n",
       "0   minor TSS   61077221   61077380  \n",
       "1  unique TSS  160261726  160261763  \n",
       "2    main TSS  239386576  239386587  \n",
       "3  unique TSS   23217498   23217499  \n",
       "4    main TSS  233327323  233327461  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract peak start and end positions\n",
    "combined_df[[\"TSS_start\", \"TSS_end\"]] = (combined_df[\"peak_name\"].str.extract(r\":(\\d+)-(\\d+)\", expand=True)).astype(\"Int64\")\n",
    "\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c9bf5d80-ac17-492b-adeb-140bbb281326",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>level</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11121</td>\n",
       "      <td>14413</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16-260</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11125</td>\n",
       "      <td>14405</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16-261</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11410</td>\n",
       "      <td>14413</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832826.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16-262</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11411</td>\n",
       "      <td>14413</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832827.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16-263</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11426</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>ENST00000832828.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L16-264</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source     feature  start    end score strand frame  \\\n",
       "0  chr1  HAVANA  transcript  11121  14413     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11125  14405     .      +     .   \n",
       "2  chr1  HAVANA  transcript  11410  14413     .      +     .   \n",
       "3  chr1  HAVANA  transcript  11411  14413     .      +     .   \n",
       "4  chr1  HAVANA  transcript  11426  14409     .      +     .   \n",
       "\n",
       "             gene_id      transcript_id gene_type gene_name transcript_type  \\\n",
       "0  ENSG00000290825.2  ENST00000832824.1    lncRNA  DDX11L16          lncRNA   \n",
       "1  ENSG00000290825.2  ENST00000832825.1    lncRNA  DDX11L16          lncRNA   \n",
       "2  ENSG00000290825.2  ENST00000832826.1    lncRNA  DDX11L16          lncRNA   \n",
       "3  ENSG00000290825.2  ENST00000832827.1    lncRNA  DDX11L16          lncRNA   \n",
       "4  ENSG00000290825.2  ENST00000832828.1    lncRNA  DDX11L16          lncRNA   \n",
       "\n",
       "  transcript_name level     tag transcript_support_level havana_transcript  \\\n",
       "0    DDX11L16-260     2  TAGENE                      NaN               NaN   \n",
       "1    DDX11L16-261     2  TAGENE                      NaN               NaN   \n",
       "2    DDX11L16-262     2  TAGENE                      NaN               NaN   \n",
       "3    DDX11L16-263     2  TAGENE                      NaN               NaN   \n",
       "4    DDX11L16-264     2  TAGENE                      NaN               NaN   \n",
       "\n",
       "  hgnc_id  ont havana_gene protein_id ccdsid  \n",
       "0     NaN  NaN         NaN        NaN    NaN  \n",
       "1     NaN  NaN         NaN        NaN    NaN  \n",
       "2     NaN  NaN         NaN        NaN    NaN  \n",
       "3     NaN  NaN         NaN        NaN    NaN  \n",
       "4     NaN  NaN         NaN        NaN    NaN  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load gencode v47 transcripts\n",
    "gencode_v47_transcripts = pd.read_csv(\"../data_raw/gencode.v47.transcript.tsv\", sep=\"\\t\", dtype=str)\n",
    "gencode_v47_transcripts[\"start\"] = gencode_v47_transcripts[\"start\"].astype(int)\n",
    "gencode_v47_transcripts[\"end\"] = gencode_v47_transcripts[\"end\"].astype(int)\n",
    "\n",
    "gencode_v47_transcripts = gencode_v47_transcripts.rename(columns={\"seqname\": \"chr\"})\n",
    "\n",
    "gencode_v47_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "71057536-eeb4-4c1d-9de0-f9e6ce59e419",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 376342/376342 [02:52<00:00, 2176.32it/s]\n"
     ]
    }
   ],
   "source": [
    "# For each TSS find the closest GENCODE tx start\n",
    "transcript_groups = {\n",
    "    (chr, strand): df for (chr, strand), df in gencode_v47_transcripts.groupby([\"chr\", \"strand\"])\n",
    "}\n",
    "\n",
    "def compute_closest_distance_and_transcript(row):\n",
    "    key = (row[\"chr\"], row[\"Strand\"])\n",
    "    matches = transcript_groups.get(key, None)\n",
    "\n",
    "    if matches is None or matches.empty or pd.isna(row[\"TSS_start\"]) or pd.isna(row[\"TSS_end\"]):\n",
    "        return pd.Series([pd.NA, pd.NA])\n",
    "\n",
    "    tss_midpoint = (row[\"TSS_start\"] + row[\"TSS_end\"]) // 2\n",
    "    \n",
    "    if row[\"Strand\"] == \"+\":\n",
    "        distances = np.minimum(np.minimum((matches[\"start\"] - row[\"TSS_start\"]).abs(), (matches[\"start\"] - row[\"TSS_end\"]).abs()), (matches[\"start\"] - tss_midpoint).abs())\n",
    "    elif row[\"Strand\"] == \"-\":\n",
    "        distances = np.minimum(np.minimum((matches[\"end\"] - row[\"TSS_start\"]).abs(), (matches[\"end\"] - row[\"TSS_end\"]).abs()), (matches[\"end\"] - tss_midpoint).abs())\n",
    "    else:\n",
    "        return pd.Series([pd.NA, pd.NA])\n",
    "\n",
    "    min_idx = distances.idxmin()\n",
    "    min_distance = distances.loc[min_idx]\n",
    "    closest_transcript = matches.loc[min_idx, \"transcript_id\"]\n",
    "\n",
    "    return pd.Series([min_distance, closest_transcript])\n",
    "\n",
    "\n",
    "tqdm.pandas()\n",
    "\n",
    "combined_df[[\"distance_to_closest_TSS\", \"closest_TSS_ref_id\"]] = combined_df.progress_apply(compute_closest_distance_and_transcript, axis=1)\n",
    "combined_df[\"distance_to_closest_TSS\"] = combined_df[\"distance_to_closest_TSS\"].astype(\"Int64\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "08506494-71e4-4377-bb70-29563a248fe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = lambda x: \"<=100bp\" if x <= 100 else \">100bp\"\n",
    "combined_df['TSS distance category'] = combined_df['distance_to_closest_TSS'].apply(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b1b03688-ec21-4545-8d26-4bb33e204c98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>closest_TSS_ref_id</th>\n",
       "      <th>TSS distance category</th>\n",
       "      <th>TSS_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(61077274, 61077628), (61077916, 61079203)]</td>\n",
       "      <td>[(61077326, 61077628), (61077916, 61079231)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000699992.1</td>\n",
       "      <td>ENST00000699992</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:61077221-61077380_Peak_3111</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.178215</td>\n",
       "      <td>-2.488310</td>\n",
       "      <td>4.820350</td>\n",
       "      <td>0.036971</td>\n",
       "      <td>0.213765</td>\n",
       "      <td>0.044346</td>\n",
       "      <td>minor TSS</td>\n",
       "      <td>61077221</td>\n",
       "      <td>61077380</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000699964.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P3111_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160261734, 160261922), (160281430, 160281892)]</td>\n",
       "      <td>[(160261682, 160261922), (160281430, 160281935)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000756371.1</td>\n",
       "      <td>ENST00000756371</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:160261726-160261763_Peak_5604</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.358414</td>\n",
       "      <td>-1.480300</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>0.481821</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>160261726</td>\n",
       "      <td>160261763</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000442130.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P5604_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>ENST00000676153</td>\n",
       "      <td>ENSG00000133019</td>\n",
       "      <td>CHRM3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:239386576-239386587_Peak_7900</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.468766</td>\n",
       "      <td>-1.093061</td>\n",
       "      <td>3.127170</td>\n",
       "      <td>0.149901</td>\n",
       "      <td>1.421490</td>\n",
       "      <td>0.454561</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>239386576</td>\n",
       "      <td>239386587</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P7900_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCT00000022754.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(23191893, 23195001), (23217291, 23217499)]</td>\n",
       "      <td>[(23191895, 23195001), (23217291, 23217502)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>ENST00000374619</td>\n",
       "      <td>ENSG00000179546</td>\n",
       "      <td>HTR1D</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:23217498-23217499_Peak_1212</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.064644</td>\n",
       "      <td>-3.951339</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>0.553728</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>23217498</td>\n",
       "      <td>23217499</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P1212_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "      <td>ENST00000771587</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:233327323-233327461_Peak_7654</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.616169</td>\n",
       "      <td>-0.698603</td>\n",
       "      <td>2.291033</td>\n",
       "      <td>0.268948</td>\n",
       "      <td>2.093885</td>\n",
       "      <td>0.913948</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>233327323</td>\n",
       "      <td>233327461</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000771586.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P7654_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id   chr                                              exons  \\\n",
       "0  ENCT00000006334.1  chr1       [(61077274, 61077628), (61077916, 61079203)]   \n",
       "1  ENCT00000013056.1  chr1   [(160261734, 160261922), (160281430, 160281892)]   \n",
       "2  ENCT00000019620.1  chr1  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3  ENCT00000022754.1  chr1       [(23191893, 23195001), (23217291, 23217499)]   \n",
       "4  ENCT00000039204.1  chr1   [(233319835, 233321155), (233327229, 233327357)]   \n",
       "\n",
       "                                          matched_to  \\\n",
       "0       [(61077326, 61077628), (61077916, 61079231)]   \n",
       "1   [(160261682, 160261922), (160281430, 160281935)]   \n",
       "2  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3       [(23191895, 23195001), (23217291, 23217502)]   \n",
       "4   [(233319834, 233321155), (233327229, 233327455)]   \n",
       "\n",
       "                        category unannotated_exons Strand  \\\n",
       "0  GENCODE-annotated transcripts               NaN      +   \n",
       "1  GENCODE-annotated transcripts               NaN      +   \n",
       "2  GENCODE-annotated transcripts               NaN      +   \n",
       "3  GENCODE-annotated transcripts               NaN      -   \n",
       "4  GENCODE-annotated transcripts               NaN      -   \n",
       "\n",
       "             summarized_category             ref_id  stripped_ref_id  \\\n",
       "0  GENCODE-annotated transcripts  ENST00000699992.1  ENST00000699992   \n",
       "1  GENCODE-annotated transcripts  ENST00000756371.1  ENST00000756371   \n",
       "2  GENCODE-annotated transcripts  ENST00000676153.1  ENST00000676153   \n",
       "3  GENCODE-annotated transcripts  ENST00000374619.2  ENST00000374619   \n",
       "4  GENCODE-annotated transcripts  ENST00000771587.1  ENST00000771587   \n",
       "\n",
       "    Gene stable ID        Gene name       Gene type   gene_category  \\\n",
       "0  ENSG00000162599             NFIA  protein_coding  protein_coding   \n",
       "1  ENSG00000228606         DCAF8-DT          lncRNA          lncRNA   \n",
       "2  ENSG00000133019            CHRM3  protein_coding  protein_coding   \n",
       "3  ENSG00000179546            HTR1D  protein_coding  protein_coding   \n",
       "4  ENSG00000300424  ENSG00000300424          lncRNA          lncRNA   \n",
       "\n",
       "                            peak_name   peak_type  mean_TPM_ref  log2_tx_exp  \\\n",
       "0    chr1:61077221-61077380_Peak_3111  Permissive      0.178215    -2.488310   \n",
       "1  chr1:160261726-160261763_Peak_5604  Permissive      0.358414    -1.480300   \n",
       "2  chr1:239386576-239386587_Peak_7900      Robust      0.468766    -1.093061   \n",
       "3    chr1:23217498-23217499_Peak_1212  Permissive      0.064644    -3.951339   \n",
       "4  chr1:233327323-233327461_Peak_7654  Permissive      0.616169    -0.698603   \n",
       "\n",
       "   gene_exp_ref  relative_tx_exp  TSS_TPM_per_gene  Relative TSS usage  \\\n",
       "0      4.820350         0.036971          0.213765            0.044346   \n",
       "1      0.743874         0.481821          0.743874            1.000000   \n",
       "2      3.127170         0.149901          1.421490            0.454561   \n",
       "3      0.116743         0.553728          0.116743            1.000000   \n",
       "4      2.291033         0.268948          2.093885            0.913948   \n",
       "\n",
       "     TSS_type  TSS_start    TSS_end  distance_to_closest_TSS  \\\n",
       "0   minor TSS   61077221   61077380                        6   \n",
       "1  unique TSS  160261726  160261763                        0   \n",
       "2    main TSS  239386576  239386587                        8   \n",
       "3  unique TSS   23217498   23217499                        3   \n",
       "4    main TSS  233327323  233327461                        3   \n",
       "\n",
       "  closest_TSS_ref_id TSS distance category TSS_name  \n",
       "0  ENST00000699964.1               <=100bp  P3111_A  \n",
       "1  ENST00000442130.1               <=100bp  P5604_A  \n",
       "2  ENST00000676153.1               <=100bp  P7900_A  \n",
       "3  ENST00000374619.2               <=100bp  P1212_A  \n",
       "4  ENST00000771586.1               <=100bp  P7654_A  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create TSS name based on annotation status\n",
    "combined_df[\"TSS_name\"] = (\n",
    "    \"P\" \n",
    "    + combined_df[\"peak_name\"].str.extract(r\"Peak_(\\d+)\", expand=False)\n",
    "    + np.where(\n",
    "        combined_df[\"TSS distance category\"] == \"<=100bp\",\n",
    "        \"_A\",\n",
    "        \"_U\"\n",
    "    )\n",
    ")\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "958e3980-381f-442e-a0c5-dc8d3665de7e",
   "metadata": {},
   "source": [
    "## Collapsing TSS groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "e1d8d6bf-7bd7-4270-8716-694e5b6f4c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "unannotated_genes = combined_df[combined_df[\"category\"] == \"Fully unannotated transcripts without any overlapping exons\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "fff97b4b-e1bf-4380-a7f1-a67095618256",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>closest_TSS_ref_id</th>\n",
       "      <th>TSS distance category</th>\n",
       "      <th>TSS_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336566</th>\n",
       "      <td>ENCT00000000219.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(1162877, 1163719), (1163993, 1164042)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chr1:1162789-1162942_Peak_111</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.018685</td>\n",
       "      <td>-5.741955</td>\n",
       "      <td>0.339126</td>\n",
       "      <td>0.055098</td>\n",
       "      <td>0.339126</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>1162789</td>\n",
       "      <td>1162942</td>\n",
       "      <td>4162</td>\n",
       "      <td>ENST00000384997.3</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P111_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336567</th>\n",
       "      <td>ENCT00000000276.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(1349559, 1355076), (1355263, 1355740), (1357...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_145</td>\n",
       "      <td>Peak_145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chr1:1349517-1349586_Peak_145</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.836916</td>\n",
       "      <td>-0.256846</td>\n",
       "      <td>0.867767</td>\n",
       "      <td>0.964448</td>\n",
       "      <td>0.867767</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>1349517</td>\n",
       "      <td>1349586</td>\n",
       "      <td>679</td>\n",
       "      <td>ENST00000806659.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P145_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336568</th>\n",
       "      <td>ENCT00000002194.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(17439849, 17449625)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_910</td>\n",
       "      <td>Peak_910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chr1:17439844-17439937_Peak_910</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.085408</td>\n",
       "      <td>-3.549480</td>\n",
       "      <td>0.085408</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.085408</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>17439844</td>\n",
       "      <td>17439937</td>\n",
       "      <td>20004</td>\n",
       "      <td>ENST00000835437.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P910_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336569</th>\n",
       "      <td>ENCT00000002302.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(19485741, 19489397)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_996</td>\n",
       "      <td>Peak_996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chr1:19485730-19485780_Peak_996</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.114061</td>\n",
       "      <td>-3.132117</td>\n",
       "      <td>0.168185</td>\n",
       "      <td>0.678191</td>\n",
       "      <td>0.168185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>19485730</td>\n",
       "      <td>19485780</td>\n",
       "      <td>1327</td>\n",
       "      <td>ENST00000648702.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P996_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336570</th>\n",
       "      <td>ENCT00000002410.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(20508236, 20510041)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chr1:20508218-20508233_Peak_1039</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.059096</td>\n",
       "      <td>-4.080791</td>\n",
       "      <td>0.059096</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.059096</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>20508218</td>\n",
       "      <td>20508233</td>\n",
       "      <td>17756</td>\n",
       "      <td>ENST00000443195.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P1039_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339852</th>\n",
       "      <td>STRT02284050</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(74034396, 74034874)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_79446</td>\n",
       "      <td>Peak_79446</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chrX:74034838-74034839_Peak_79446</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>1.715846</td>\n",
       "      <td>0.778920</td>\n",
       "      <td>1.715846</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.715846</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>74034838</td>\n",
       "      <td>74034839</td>\n",
       "      <td>21893</td>\n",
       "      <td>ENST00000638437.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P79446_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339853</th>\n",
       "      <td>STRT02289282</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(51805614, 51805945)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_79008</td>\n",
       "      <td>Peak_79008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chrX:51805624-51805625_Peak_79008</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.318309</td>\n",
       "      <td>-1.651502</td>\n",
       "      <td>0.318309</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.318309</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>51805624</td>\n",
       "      <td>51805625</td>\n",
       "      <td>2617</td>\n",
       "      <td>ENST00000375772.7</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P79008_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339854</th>\n",
       "      <td>STRT02289408</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(57617393, 57618659)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_79202</td>\n",
       "      <td>Peak_79202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chrX:57617355-57617356_Peak_79202</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.129846</td>\n",
       "      <td>-2.945122</td>\n",
       "      <td>0.129846</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.129846</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>57617355</td>\n",
       "      <td>57617356</td>\n",
       "      <td>23432</td>\n",
       "      <td>ENST00000818758.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P79202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339855</th>\n",
       "      <td>STRT02290336</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(19990302, 19990805)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_78445</td>\n",
       "      <td>Peak_78445</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chrX:19990802-19990828_Peak_78445</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.101023</td>\n",
       "      <td>-3.307240</td>\n",
       "      <td>0.260752</td>\n",
       "      <td>0.387430</td>\n",
       "      <td>0.260752</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>19990802</td>\n",
       "      <td>19990828</td>\n",
       "      <td>233</td>\n",
       "      <td>ENST00000379682.9</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P78445_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339856</th>\n",
       "      <td>PBT00442330</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(18987729, 18991364)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any over...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts without any annotated ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>chrY:18991177-18991443_Peak_80755</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.372451</td>\n",
       "      <td>-1.424877</td>\n",
       "      <td>0.372451</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.372451</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>18991177</td>\n",
       "      <td>18991443</td>\n",
       "      <td>1266</td>\n",
       "      <td>ENST00000382840.3</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P80755_U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3291 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id   chr  \\\n",
       "336566  ENCT00000000219.1  chr1   \n",
       "336567  ENCT00000000276.1  chr1   \n",
       "336568  ENCT00000002194.1  chr1   \n",
       "336569  ENCT00000002302.1  chr1   \n",
       "336570  ENCT00000002410.1  chr1   \n",
       "...                   ...   ...   \n",
       "339852       STRT02284050  chrX   \n",
       "339853       STRT02289282  chrX   \n",
       "339854       STRT02289408  chrX   \n",
       "339855       STRT02290336  chrX   \n",
       "339856        PBT00442330  chrY   \n",
       "\n",
       "                                                    exons matched_to  \\\n",
       "336566           [(1162877, 1163719), (1163993, 1164042)]        NaN   \n",
       "336567  [(1349559, 1355076), (1355263, 1355740), (1357...        NaN   \n",
       "336568                             [(17439849, 17449625)]        NaN   \n",
       "336569                             [(19485741, 19489397)]        NaN   \n",
       "336570                             [(20508236, 20510041)]        NaN   \n",
       "...                                                   ...        ...   \n",
       "339852                             [(74034396, 74034874)]        NaN   \n",
       "339853                             [(51805614, 51805945)]        NaN   \n",
       "339854                             [(57617393, 57618659)]        NaN   \n",
       "339855                             [(19990302, 19990805)]        NaN   \n",
       "339856                             [(18987729, 18991364)]        NaN   \n",
       "\n",
       "                                                 category unannotated_exons  \\\n",
       "336566  Fully unannotated transcripts without any over...               NaN   \n",
       "336567  Fully unannotated transcripts without any over...               NaN   \n",
       "336568  Fully unannotated transcripts without any over...               NaN   \n",
       "336569  Fully unannotated transcripts without any over...               NaN   \n",
       "336570  Fully unannotated transcripts without any over...               NaN   \n",
       "...                                                   ...               ...   \n",
       "339852  Fully unannotated transcripts without any over...               NaN   \n",
       "339853  Fully unannotated transcripts without any over...               NaN   \n",
       "339854  Fully unannotated transcripts without any over...               NaN   \n",
       "339855  Fully unannotated transcripts without any over...               NaN   \n",
       "339856  Fully unannotated transcripts without any over...               NaN   \n",
       "\n",
       "       Strand                                summarized_category ref_id  \\\n",
       "336566      +  Unannotated transcripts without any annotated ...   None   \n",
       "336567      +  Unannotated transcripts without any annotated ...   None   \n",
       "336568      +  Unannotated transcripts without any annotated ...   None   \n",
       "336569      +  Unannotated transcripts without any annotated ...   None   \n",
       "336570      +  Unannotated transcripts without any annotated ...   None   \n",
       "...       ...                                                ...    ...   \n",
       "339852      -  Unannotated transcripts without any annotated ...   None   \n",
       "339853      +  Unannotated transcripts without any annotated ...   None   \n",
       "339854      +  Unannotated transcripts without any annotated ...   None   \n",
       "339855      -  Unannotated transcripts without any annotated ...   None   \n",
       "339856      -  Unannotated transcripts without any annotated ...   None   \n",
       "\n",
       "       stripped_ref_id Gene stable ID   Gene name Gene type gene_category  \\\n",
       "336566            None       Peak_111    Peak_111       NaN      unmapped   \n",
       "336567            None       Peak_145    Peak_145       NaN      unmapped   \n",
       "336568            None       Peak_910    Peak_910       NaN      unmapped   \n",
       "336569            None       Peak_996    Peak_996       NaN      unmapped   \n",
       "336570            None      Peak_1039   Peak_1039       NaN      unmapped   \n",
       "...                ...            ...         ...       ...           ...   \n",
       "339852            None     Peak_79446  Peak_79446       NaN      unmapped   \n",
       "339853            None     Peak_79008  Peak_79008       NaN      unmapped   \n",
       "339854            None     Peak_79202  Peak_79202       NaN      unmapped   \n",
       "339855            None     Peak_78445  Peak_78445       NaN      unmapped   \n",
       "339856            None     Peak_80755  Peak_80755       NaN      unmapped   \n",
       "\n",
       "                                peak_name   peak_type  mean_TPM_ref  \\\n",
       "336566      chr1:1162789-1162942_Peak_111      Robust      0.018685   \n",
       "336567      chr1:1349517-1349586_Peak_145  Permissive      0.836916   \n",
       "336568    chr1:17439844-17439937_Peak_910  Permissive      0.085408   \n",
       "336569    chr1:19485730-19485780_Peak_996  Permissive      0.114061   \n",
       "336570   chr1:20508218-20508233_Peak_1039  Permissive      0.059096   \n",
       "...                                   ...         ...           ...   \n",
       "339852  chrX:74034838-74034839_Peak_79446  Permissive      1.715846   \n",
       "339853  chrX:51805624-51805625_Peak_79008  Permissive      0.318309   \n",
       "339854  chrX:57617355-57617356_Peak_79202  Permissive      0.129846   \n",
       "339855  chrX:19990802-19990828_Peak_78445  Permissive      0.101023   \n",
       "339856  chrY:18991177-18991443_Peak_80755  Permissive      0.372451   \n",
       "\n",
       "        log2_tx_exp  gene_exp_ref  relative_tx_exp  TSS_TPM_per_gene  \\\n",
       "336566    -5.741955      0.339126         0.055098          0.339126   \n",
       "336567    -0.256846      0.867767         0.964448          0.867767   \n",
       "336568    -3.549480      0.085408         1.000000          0.085408   \n",
       "336569    -3.132117      0.168185         0.678191          0.168185   \n",
       "336570    -4.080791      0.059096         1.000000          0.059096   \n",
       "...             ...           ...              ...               ...   \n",
       "339852     0.778920      1.715846         1.000000          1.715846   \n",
       "339853    -1.651502      0.318309         1.000000          0.318309   \n",
       "339854    -2.945122      0.129846         1.000000          0.129846   \n",
       "339855    -3.307240      0.260752         0.387430          0.260752   \n",
       "339856    -1.424877      0.372451         1.000000          0.372451   \n",
       "\n",
       "        Relative TSS usage    TSS_type  TSS_start   TSS_end  \\\n",
       "336566                 1.0  unique TSS    1162789   1162942   \n",
       "336567                 1.0  unique TSS    1349517   1349586   \n",
       "336568                 1.0  unique TSS   17439844  17439937   \n",
       "336569                 1.0  unique TSS   19485730  19485780   \n",
       "336570                 1.0  unique TSS   20508218  20508233   \n",
       "...                    ...         ...        ...       ...   \n",
       "339852                 1.0  unique TSS   74034838  74034839   \n",
       "339853                 1.0  unique TSS   51805624  51805625   \n",
       "339854                 1.0  unique TSS   57617355  57617356   \n",
       "339855                 1.0  unique TSS   19990802  19990828   \n",
       "339856                 1.0  unique TSS   18991177  18991443   \n",
       "\n",
       "        distance_to_closest_TSS closest_TSS_ref_id TSS distance category  \\\n",
       "336566                     4162  ENST00000384997.3                >100bp   \n",
       "336567                      679  ENST00000806659.1                >100bp   \n",
       "336568                    20004  ENST00000835437.1                >100bp   \n",
       "336569                     1327  ENST00000648702.1                >100bp   \n",
       "336570                    17756  ENST00000443195.1                >100bp   \n",
       "...                         ...                ...                   ...   \n",
       "339852                    21893  ENST00000638437.1                >100bp   \n",
       "339853                     2617  ENST00000375772.7                >100bp   \n",
       "339854                    23432  ENST00000818758.1                >100bp   \n",
       "339855                      233  ENST00000379682.9                >100bp   \n",
       "339856                     1266  ENST00000382840.3                >100bp   \n",
       "\n",
       "        TSS_name  \n",
       "336566    P111_U  \n",
       "336567    P145_U  \n",
       "336568    P910_U  \n",
       "336569    P996_U  \n",
       "336570   P1039_U  \n",
       "...          ...  \n",
       "339852  P79446_U  \n",
       "339853  P79008_U  \n",
       "339854  P79202_U  \n",
       "339855  P78445_U  \n",
       "339856  P80755_U  \n",
       "\n",
       "[3291 rows x 29 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unannotated_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "7d834173-2203-450f-8949-f9edde03dda0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tx_coords(exon_list):\n",
    "    if not exon_list or not isinstance(exon_list, (list, tuple)):\n",
    "        return (None, None)\n",
    "    starts = [e[0] for e in exon_list]\n",
    "    ends = [e[1] for e in exon_list]\n",
    "    return (min(starts), max(ends))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "aba19213-cc70-4be9-8610-1bc38f899a67",
   "metadata": {},
   "outputs": [],
   "source": [
    "unannotated_genes[[\"tx_start\", \"tx_end\"]] = unannotated_genes[\"exons\"].apply(\n",
    "    lambda x: pd.Series(get_tx_coords(x))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ceafc4a8-9743-4723-a185-5ddd56899a8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tx_start</th>\n",
       "      <th>tx_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336566</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>ENCT00000000219.1</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1164042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336567</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_145</td>\n",
       "      <td>ENCT00000000276.1</td>\n",
       "      <td>1349559</td>\n",
       "      <td>1358279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336568</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_910</td>\n",
       "      <td>ENCT00000002194.1</td>\n",
       "      <td>17439849</td>\n",
       "      <td>17449625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336569</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_996</td>\n",
       "      <td>ENCT00000002302.1</td>\n",
       "      <td>19485741</td>\n",
       "      <td>19489397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336570</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>ENCT00000002410.1</td>\n",
       "      <td>20508236</td>\n",
       "      <td>20510041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339852</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_79446</td>\n",
       "      <td>STRT02284050</td>\n",
       "      <td>74034396</td>\n",
       "      <td>74034874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339853</th>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_79008</td>\n",
       "      <td>STRT02289282</td>\n",
       "      <td>51805614</td>\n",
       "      <td>51805945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339854</th>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_79202</td>\n",
       "      <td>STRT02289408</td>\n",
       "      <td>57617393</td>\n",
       "      <td>57618659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339855</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_78445</td>\n",
       "      <td>STRT02290336</td>\n",
       "      <td>19990302</td>\n",
       "      <td>19990805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339856</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>PBT00442330</td>\n",
       "      <td>18987729</td>\n",
       "      <td>18991364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3291 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr Strand   Gene name      transcript_id  tx_start    tx_end\n",
       "336566  chr1      +    Peak_111  ENCT00000000219.1   1162877   1164042\n",
       "336567  chr1      +    Peak_145  ENCT00000000276.1   1349559   1358279\n",
       "336568  chr1      +    Peak_910  ENCT00000002194.1  17439849  17449625\n",
       "336569  chr1      +    Peak_996  ENCT00000002302.1  19485741  19489397\n",
       "336570  chr1      +   Peak_1039  ENCT00000002410.1  20508236  20510041\n",
       "...      ...    ...         ...                ...       ...       ...\n",
       "339852  chrX      -  Peak_79446       STRT02284050  74034396  74034874\n",
       "339853  chrX      +  Peak_79008       STRT02289282  51805614  51805945\n",
       "339854  chrX      +  Peak_79202       STRT02289408  57617393  57618659\n",
       "339855  chrX      -  Peak_78445       STRT02290336  19990302  19990805\n",
       "339856  chrY      -  Peak_80755        PBT00442330  18987729  18991364\n",
       "\n",
       "[3291 rows x 6 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_coords = unannotated_genes[[\"chr\", \"Strand\", \"Gene name\", \"transcript_id\", \"tx_start\", \"tx_end\"]]\n",
    "transcript_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "88c663d2-d3a5-479c-8274-54ca633e62b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1704"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_coords[\"Gene name\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "739fe7d3-549c-4b15-bec6-b2dd2a31e4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>min_start</th>\n",
       "      <th>max_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>20508236</td>\n",
       "      <td>20510041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1166499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1187</td>\n",
       "      <td>22342372</td>\n",
       "      <td>22353088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1234</td>\n",
       "      <td>23555118</td>\n",
       "      <td>23556240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1238</td>\n",
       "      <td>23559643</td>\n",
       "      <td>23560425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1699</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80211</td>\n",
       "      <td>126472363</td>\n",
       "      <td>126472699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80214</td>\n",
       "      <td>129382617</td>\n",
       "      <td>129382957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1701</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80261</td>\n",
       "      <td>130401109</td>\n",
       "      <td>130401897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1702</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80530</td>\n",
       "      <td>153927441</td>\n",
       "      <td>153928228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1703</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>18987729</td>\n",
       "      <td>18991364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1704 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr Strand   Gene_name  min_start    max_end\n",
       "0     chr1      +   Peak_1039   20508236   20510041\n",
       "1     chr1      +    Peak_111    1162877    1166499\n",
       "2     chr1      +   Peak_1187   22342372   22353088\n",
       "3     chr1      +   Peak_1234   23555118   23556240\n",
       "4     chr1      +   Peak_1238   23559643   23560425\n",
       "...    ...    ...         ...        ...        ...\n",
       "1699  chrX      -  Peak_80211  126472363  126472699\n",
       "1700  chrX      -  Peak_80214  129382617  129382957\n",
       "1701  chrX      -  Peak_80261  130401109  130401897\n",
       "1702  chrX      -  Peak_80530  153927441  153928228\n",
       "1703  chrY      -  Peak_80755   18987729   18991364\n",
       "\n",
       "[1704 rows x 5 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_coords = (\n",
    "    transcript_coords\n",
    "    .groupby([\"chr\", \"Strand\", \"Gene name\"], as_index=False)\n",
    "    .agg(min_start=(\"tx_start\", \"min\"), max_end=(\"tx_end\", \"max\"))\n",
    ")\n",
    "gene_coords = gene_coords.rename(columns={\"Gene name\": \"Gene_name\"})\n",
    "gene_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "72b41a0a-ee8f-4bc9-823d-c8ad172998b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>group_id</th>\n",
       "      <th>group_start</th>\n",
       "      <th>group_end</th>\n",
       "      <th>Genes</th>\n",
       "      <th>group_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1166499</td>\n",
       "      <td>[Peak_111]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>1349550</td>\n",
       "      <td>1358279</td>\n",
       "      <td>[Peak_145]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1375573</td>\n",
       "      <td>1378310</td>\n",
       "      <td>[Peak_150]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>2189680</td>\n",
       "      <td>2191740</td>\n",
       "      <td>[Peak_264]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>2199524</td>\n",
       "      <td>2202776</td>\n",
       "      <td>[Peak_274]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1603</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1604</td>\n",
       "      <td>126472363</td>\n",
       "      <td>126472699</td>\n",
       "      <td>[Peak_80211]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1604</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1605</td>\n",
       "      <td>129382617</td>\n",
       "      <td>129382957</td>\n",
       "      <td>[Peak_80214]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1605</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1606</td>\n",
       "      <td>130401109</td>\n",
       "      <td>130401897</td>\n",
       "      <td>[Peak_80261]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1607</td>\n",
       "      <td>153927441</td>\n",
       "      <td>153928228</td>\n",
       "      <td>[Peak_80530]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1607</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1608</td>\n",
       "      <td>18987729</td>\n",
       "      <td>18991364</td>\n",
       "      <td>[Peak_80755]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1608 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr Strand  group_id  group_start  group_end         Genes  group_size\n",
       "0     chr1      +         1      1162877    1166499    [Peak_111]           1\n",
       "1     chr1      +         2      1349550    1358279    [Peak_145]           1\n",
       "2     chr1      +         3      1375573    1378310    [Peak_150]           1\n",
       "3     chr1      +         4      2189680    2191740    [Peak_264]           1\n",
       "4     chr1      +         5      2199524    2202776    [Peak_274]           1\n",
       "...    ...    ...       ...          ...        ...           ...         ...\n",
       "1603  chrX      -      1604    126472363  126472699  [Peak_80211]           1\n",
       "1604  chrX      -      1605    129382617  129382957  [Peak_80214]           1\n",
       "1605  chrX      -      1606    130401109  130401897  [Peak_80261]           1\n",
       "1606  chrX      -      1607    153927441  153928228  [Peak_80530]           1\n",
       "1607  chrY      -      1608     18987729   18991364  [Peak_80755]           1\n",
       "\n",
       "[1608 rows x 7 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort first\n",
    "gene_coords = gene_coords.sort_values([\"chr\", \"Strand\", \"min_start\"]).reset_index(drop=True)\n",
    "\n",
    "# Compute running maximum end per chromosome/strand\n",
    "gene_coords[\"running_max_end\"] = (\n",
    "    gene_coords.groupby([\"chr\", \"Strand\"])[\"max_end\"]\n",
    "    .cummax()\n",
    ")\n",
    "\n",
    "# A new group starts when the current start > previous running max_end\n",
    "gene_coords[\"prev_running_max_end\"] = gene_coords.groupby([\"chr\", \"Strand\"])[\"running_max_end\"].shift()\n",
    "gene_coords[\"new_group\"] = (\n",
    "    (gene_coords[\"chr\"] != gene_coords[\"chr\"].shift()) |\n",
    "    (gene_coords[\"Strand\"] != gene_coords[\"Strand\"].shift()) |\n",
    "    (gene_coords[\"min_start\"] > gene_coords[\"prev_running_max_end\"])\n",
    ").astype(int)\n",
    "\n",
    "gene_coords[\"group_id\"] = gene_coords[\"new_group\"].cumsum()\n",
    "\n",
    "# Summarize groups\n",
    "groups = (\n",
    "    gene_coords.groupby([\"chr\", \"Strand\", \"group_id\"])\n",
    "    .agg(\n",
    "        group_start=(\"min_start\", \"min\"),\n",
    "        group_end=(\"max_end\", \"max\"),\n",
    "        Genes=(\"Gene_name\", list),\n",
    "        group_size=(\"Gene_name\", \"count\")\n",
    "    )\n",
    "    .reset_index()\n",
    ")\n",
    "groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "f02639eb-a343-4cc1-b2c6-8fb0fd651bd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>group_id</th>\n",
       "      <th>group_start</th>\n",
       "      <th>group_end</th>\n",
       "      <th>Genes</th>\n",
       "      <th>group_size</th>\n",
       "      <th>new_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>chr8</td>\n",
       "      <td>+</td>\n",
       "      <td>1463</td>\n",
       "      <td>117244748</td>\n",
       "      <td>117311898</td>\n",
       "      <td>[Peak_74199, Peak_74200, Peak_74201, Peak_7420...</td>\n",
       "      <td>5</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>567</td>\n",
       "      <td>57564434</td>\n",
       "      <td>57607224</td>\n",
       "      <td>[Peak_29184, Peak_29170, Peak_29174, Peak_2917...</td>\n",
       "      <td>5</td>\n",
       "      <td>Peak_29170-29184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>1009</td>\n",
       "      <td>122184040</td>\n",
       "      <td>122253386</td>\n",
       "      <td>[Peak_54073, Peak_54086, Peak_54100, Peak_54102]</td>\n",
       "      <td>4</td>\n",
       "      <td>Peak_54073-54102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>182</td>\n",
       "      <td>97772165</td>\n",
       "      <td>97842556</td>\n",
       "      <td>[Peak_10465, Peak_10467, Peak_10468, Peak_10469]</td>\n",
       "      <td>4</td>\n",
       "      <td>Peak_10465-10469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>572</td>\n",
       "      <td>85445590</td>\n",
       "      <td>85465053</td>\n",
       "      <td>[Peak_30027, Peak_30030, Peak_30031]</td>\n",
       "      <td>3</td>\n",
       "      <td>Peak_30027-30031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>544</td>\n",
       "      <td>75266030</td>\n",
       "      <td>75269181</td>\n",
       "      <td>[Peak_29846]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>543</td>\n",
       "      <td>72221946</td>\n",
       "      <td>72228455</td>\n",
       "      <td>[Peak_29730]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>542</td>\n",
       "      <td>71809311</td>\n",
       "      <td>71813524</td>\n",
       "      <td>[Peak_29702]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>541</td>\n",
       "      <td>70946388</td>\n",
       "      <td>70948148</td>\n",
       "      <td>[Peak_29657]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>553</td>\n",
       "      <td>605155</td>\n",
       "      <td>611191</td>\n",
       "      <td>[Peak_27341]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_27341</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1608 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr Strand  group_id  group_start  group_end  \\\n",
       "1462   chr8      +      1463    117244748  117311898   \n",
       "566   chr16      -       567     57564434   57607224   \n",
       "1008   chr3      +      1009    122184040  122253386   \n",
       "181   chr10      +       182     97772165   97842556   \n",
       "571   chr16      -       572     85445590   85465053   \n",
       "...     ...    ...       ...          ...        ...   \n",
       "543   chr16      +       544     75266030   75269181   \n",
       "542   chr16      +       543     72221946   72228455   \n",
       "541   chr16      +       542     71809311   71813524   \n",
       "540   chr16      +       541     70946388   70948148   \n",
       "552   chr16      -       553       605155     611191   \n",
       "\n",
       "                                                  Genes  group_size  \\\n",
       "1462  [Peak_74199, Peak_74200, Peak_74201, Peak_7420...           5   \n",
       "566   [Peak_29184, Peak_29170, Peak_29174, Peak_2917...           5   \n",
       "1008   [Peak_54073, Peak_54086, Peak_54100, Peak_54102]           4   \n",
       "181    [Peak_10465, Peak_10467, Peak_10468, Peak_10469]           4   \n",
       "571                [Peak_30027, Peak_30030, Peak_30031]           3   \n",
       "...                                                 ...         ...   \n",
       "543                                        [Peak_29846]           1   \n",
       "542                                        [Peak_29730]           1   \n",
       "541                                        [Peak_29702]           1   \n",
       "540                                        [Peak_29657]           1   \n",
       "552                                        [Peak_27341]           1   \n",
       "\n",
       "         new_gene_name  \n",
       "1462  Peak_74199-74203  \n",
       "566   Peak_29170-29184  \n",
       "1008  Peak_54073-54102  \n",
       "181   Peak_10465-10469  \n",
       "571   Peak_30027-30031  \n",
       "...                ...  \n",
       "543         Peak_29846  \n",
       "542         Peak_29730  \n",
       "541         Peak_29702  \n",
       "540         Peak_29657  \n",
       "552         Peak_27341  \n",
       "\n",
       "[1608 rows x 8 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def make_combined_name(genes):\n",
    "    if len(genes) == 1:\n",
    "        return genes[0]\n",
    "    else:\n",
    "        # Extract numeric parts and sort\n",
    "        nums = sorted([int(g.split(\"_\")[1]) for g in genes])\n",
    "        return f\"Peak_{nums[0]}-{nums[-1]}\"\n",
    "\n",
    "# Apply to create new column\n",
    "groups[\"new_gene_name\"] = groups[\"Genes\"].apply(make_combined_name)\n",
    "\n",
    "groups.sort_values(by=\"group_size\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "6fd7577e-1fe1-4409-90f4-4d4403fac5ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create gene name mapping\n",
    "gene_to_new = {}\n",
    "for _, row in groups.iterrows():\n",
    "    for gene in row[\"Genes\"]:\n",
    "        gene_to_new[gene] = row[\"new_gene_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "fb8a8d56-4f23-4b14-ba40-2fb0ef895a4d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Replace gene names\n",
    "mask = combined_df[\"Gene name\"].str.startswith(\"Peak\")\n",
    "combined_df.loc[mask, \"Gene name\"] = combined_df.loc[mask, \"Gene name\"].map(gene_to_new)\n",
    "combined_df.loc[mask, \"Gene stable ID\"] = combined_df.loc[mask, \"Gene stable ID\"].map(gene_to_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "47a4d7b2-6b3b-4a12-bbcd-ba8dfcdf9343",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22015"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"Gene stable ID\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "07125c75-af34-409c-9bb0-492245e6faa1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>closest_TSS_ref_id</th>\n",
       "      <th>TSS distance category</th>\n",
       "      <th>TSS_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(61077274, 61077628), (61077916, 61079203)]</td>\n",
       "      <td>[(61077326, 61077628), (61077916, 61079231)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000699992.1</td>\n",
       "      <td>ENST00000699992</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:61077221-61077380_Peak_3111</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.178215</td>\n",
       "      <td>-2.488310</td>\n",
       "      <td>4.820350</td>\n",
       "      <td>0.036971</td>\n",
       "      <td>0.213765</td>\n",
       "      <td>0.044346</td>\n",
       "      <td>minor TSS</td>\n",
       "      <td>61077221</td>\n",
       "      <td>61077380</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000699964.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P3111_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160261734, 160261922), (160281430, 160281892)]</td>\n",
       "      <td>[(160261682, 160261922), (160281430, 160281935)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000756371.1</td>\n",
       "      <td>ENST00000756371</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:160261726-160261763_Peak_5604</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.358414</td>\n",
       "      <td>-1.480300</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>0.481821</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>160261726</td>\n",
       "      <td>160261763</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000442130.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P5604_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>ENST00000676153</td>\n",
       "      <td>ENSG00000133019</td>\n",
       "      <td>CHRM3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:239386576-239386587_Peak_7900</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.468766</td>\n",
       "      <td>-1.093061</td>\n",
       "      <td>3.127170</td>\n",
       "      <td>0.149901</td>\n",
       "      <td>1.421490</td>\n",
       "      <td>0.454561</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>239386576</td>\n",
       "      <td>239386587</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P7900_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCT00000022754.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(23191893, 23195001), (23217291, 23217499)]</td>\n",
       "      <td>[(23191895, 23195001), (23217291, 23217502)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>ENST00000374619</td>\n",
       "      <td>ENSG00000179546</td>\n",
       "      <td>HTR1D</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:23217498-23217499_Peak_1212</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.064644</td>\n",
       "      <td>-3.951339</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>0.553728</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>23217498</td>\n",
       "      <td>23217499</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P1212_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "      <td>ENST00000771587</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:233327323-233327461_Peak_7654</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.616169</td>\n",
       "      <td>-0.698603</td>\n",
       "      <td>2.291033</td>\n",
       "      <td>0.268948</td>\n",
       "      <td>2.093885</td>\n",
       "      <td>0.913948</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>233327323</td>\n",
       "      <td>233327461</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000771586.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P7654_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376337</th>\n",
       "      <td>STRT02294272</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(12662355, 12662706), (12664640, 12664686), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000651177.1</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrY:12662270-12662588_Peak_80721</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.035458</td>\n",
       "      <td>-4.817728</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>0.004055</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>12662270</td>\n",
       "      <td>12662588</td>\n",
       "      <td>51</td>\n",
       "      <td>ENST00000417071.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P80721_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376338</th>\n",
       "      <td>STRT02294670</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(20575798, 20575887), (20582590, 20582693), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000382772.3</td>\n",
       "      <td>ENST00000382772</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrY:20575721-20575776_Peak_80778</td>\n",
       "      <td>Robust</td>\n",
       "      <td>3.717255</td>\n",
       "      <td>1.894237</td>\n",
       "      <td>20.208525</td>\n",
       "      <td>0.183945</td>\n",
       "      <td>20.208525</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>20575721</td>\n",
       "      <td>20575776</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000361365.7</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P80778_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376339</th>\n",
       "      <td>STRT02294844</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(12662506, 12662706), (12664640, 12664686), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000651177.1</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrY:12662270-12662588_Peak_80721</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.248549</td>\n",
       "      <td>-2.008399</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>0.028425</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>12662270</td>\n",
       "      <td>12662588</td>\n",
       "      <td>51</td>\n",
       "      <td>ENST00000417071.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P80721_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376340</th>\n",
       "      <td>STRT02294886</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(12662353, 12662706), (12664640, 12664686), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000651177.1</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrY:12662270-12662588_Peak_80721</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.033453</td>\n",
       "      <td>-4.901732</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>0.003826</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>12662270</td>\n",
       "      <td>12662588</td>\n",
       "      <td>51</td>\n",
       "      <td>ENST00000417071.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P80721_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376341</th>\n",
       "      <td>STRT02296304</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(12662506, 12662706), (12664640, 12664686), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts with all exons annotated</td>\n",
       "      <td>ENST00000651177.1</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrY:12662270-12662588_Peak_80721</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.020601</td>\n",
       "      <td>-5.601169</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>0.002356</td>\n",
       "      <td>8.744087</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>12662270</td>\n",
       "      <td>12662588</td>\n",
       "      <td>51</td>\n",
       "      <td>ENST00000417071.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P80721_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>376342 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id   chr  \\\n",
       "0       ENCT00000006334.1  chr1   \n",
       "1       ENCT00000013056.1  chr1   \n",
       "2       ENCT00000019620.1  chr1   \n",
       "3       ENCT00000022754.1  chr1   \n",
       "4       ENCT00000039204.1  chr1   \n",
       "...                   ...   ...   \n",
       "376337       STRT02294272  chrY   \n",
       "376338       STRT02294670  chrY   \n",
       "376339       STRT02294844  chrY   \n",
       "376340       STRT02294886  chrY   \n",
       "376341       STRT02296304  chrY   \n",
       "\n",
       "                                                    exons  \\\n",
       "0            [(61077274, 61077628), (61077916, 61079203)]   \n",
       "1        [(160261734, 160261922), (160281430, 160281892)]   \n",
       "2       [(239386568, 239387227), (239492709, 239492807...   \n",
       "3            [(23191893, 23195001), (23217291, 23217499)]   \n",
       "4        [(233319835, 233321155), (233327229, 233327357)]   \n",
       "...                                                   ...   \n",
       "376337  [(12662355, 12662706), (12664640, 12664686), (...   \n",
       "376338  [(20575798, 20575887), (20582590, 20582693), (...   \n",
       "376339  [(12662506, 12662706), (12664640, 12664686), (...   \n",
       "376340  [(12662353, 12662706), (12664640, 12664686), (...   \n",
       "376341  [(12662506, 12662706), (12664640, 12664686), (...   \n",
       "\n",
       "                                               matched_to  \\\n",
       "0            [(61077326, 61077628), (61077916, 61079231)]   \n",
       "1        [(160261682, 160261922), (160281430, 160281935)]   \n",
       "2       [(239386568, 239387227), (239492709, 239492807...   \n",
       "3            [(23191895, 23195001), (23217291, 23217502)]   \n",
       "4        [(233319834, 233321155), (233327229, 233327455)]   \n",
       "...                                                   ...   \n",
       "376337                                                NaN   \n",
       "376338                                                NaN   \n",
       "376339                                                NaN   \n",
       "376340                                                NaN   \n",
       "376341                                                NaN   \n",
       "\n",
       "                                                 category unannotated_exons  \\\n",
       "0                           GENCODE-annotated transcripts               NaN   \n",
       "1                           GENCODE-annotated transcripts               NaN   \n",
       "2                           GENCODE-annotated transcripts               NaN   \n",
       "3                           GENCODE-annotated transcripts               NaN   \n",
       "4                           GENCODE-annotated transcripts               NaN   \n",
       "...                                                   ...               ...   \n",
       "376337  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376338  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376339  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376340  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376341  Unannotated transcripts composed entirely of a...               NaN   \n",
       "\n",
       "       Strand                               summarized_category  \\\n",
       "0           +                     GENCODE-annotated transcripts   \n",
       "1           +                     GENCODE-annotated transcripts   \n",
       "2           +                     GENCODE-annotated transcripts   \n",
       "3           -                     GENCODE-annotated transcripts   \n",
       "4           -                     GENCODE-annotated transcripts   \n",
       "...       ...                                               ...   \n",
       "376337      +  Unannotated transcripts with all exons annotated   \n",
       "376338      +  Unannotated transcripts with all exons annotated   \n",
       "376339      +  Unannotated transcripts with all exons annotated   \n",
       "376340      +  Unannotated transcripts with all exons annotated   \n",
       "376341      +  Unannotated transcripts with all exons annotated   \n",
       "\n",
       "                   ref_id  stripped_ref_id   Gene stable ID        Gene name  \\\n",
       "0       ENST00000699992.1  ENST00000699992  ENSG00000162599             NFIA   \n",
       "1       ENST00000756371.1  ENST00000756371  ENSG00000228606         DCAF8-DT   \n",
       "2       ENST00000676153.1  ENST00000676153  ENSG00000133019            CHRM3   \n",
       "3       ENST00000374619.2  ENST00000374619  ENSG00000179546            HTR1D   \n",
       "4       ENST00000771587.1  ENST00000771587  ENSG00000300424  ENSG00000300424   \n",
       "...                   ...              ...              ...              ...   \n",
       "376337  ENST00000651177.1  ENST00000651177  ENSG00000114374            USP9Y   \n",
       "376338  ENST00000382772.3  ENST00000382772  ENSG00000198692           EIF1AY   \n",
       "376339  ENST00000651177.1  ENST00000651177  ENSG00000114374            USP9Y   \n",
       "376340  ENST00000651177.1  ENST00000651177  ENSG00000114374            USP9Y   \n",
       "376341  ENST00000651177.1  ENST00000651177  ENSG00000114374            USP9Y   \n",
       "\n",
       "             Gene type   gene_category                           peak_name  \\\n",
       "0       protein_coding  protein_coding    chr1:61077221-61077380_Peak_3111   \n",
       "1               lncRNA          lncRNA  chr1:160261726-160261763_Peak_5604   \n",
       "2       protein_coding  protein_coding  chr1:239386576-239386587_Peak_7900   \n",
       "3       protein_coding  protein_coding    chr1:23217498-23217499_Peak_1212   \n",
       "4               lncRNA          lncRNA  chr1:233327323-233327461_Peak_7654   \n",
       "...                ...             ...                                 ...   \n",
       "376337  protein_coding  protein_coding   chrY:12662270-12662588_Peak_80721   \n",
       "376338  protein_coding  protein_coding   chrY:20575721-20575776_Peak_80778   \n",
       "376339  protein_coding  protein_coding   chrY:12662270-12662588_Peak_80721   \n",
       "376340  protein_coding  protein_coding   chrY:12662270-12662588_Peak_80721   \n",
       "376341  protein_coding  protein_coding   chrY:12662270-12662588_Peak_80721   \n",
       "\n",
       "         peak_type  mean_TPM_ref  log2_tx_exp  gene_exp_ref  relative_tx_exp  \\\n",
       "0       Permissive      0.178215    -2.488310      4.820350         0.036971   \n",
       "1       Permissive      0.358414    -1.480300      0.743874         0.481821   \n",
       "2           Robust      0.468766    -1.093061      3.127170         0.149901   \n",
       "3       Permissive      0.064644    -3.951339      0.116743         0.553728   \n",
       "4       Permissive      0.616169    -0.698603      2.291033         0.268948   \n",
       "...            ...           ...          ...           ...              ...   \n",
       "376337      Robust      0.035458    -4.817728      8.744087         0.004055   \n",
       "376338      Robust      3.717255     1.894237     20.208525         0.183945   \n",
       "376339      Robust      0.248549    -2.008399      8.744087         0.028425   \n",
       "376340      Robust      0.033453    -4.901732      8.744087         0.003826   \n",
       "376341      Robust      0.020601    -5.601169      8.744087         0.002356   \n",
       "\n",
       "        TSS_TPM_per_gene  Relative TSS usage    TSS_type  TSS_start  \\\n",
       "0               0.213765            0.044346   minor TSS   61077221   \n",
       "1               0.743874            1.000000  unique TSS  160261726   \n",
       "2               1.421490            0.454561    main TSS  239386576   \n",
       "3               0.116743            1.000000  unique TSS   23217498   \n",
       "4               2.093885            0.913948    main TSS  233327323   \n",
       "...                  ...                 ...         ...        ...   \n",
       "376337          8.744087            1.000000  unique TSS   12662270   \n",
       "376338         20.208525            1.000000  unique TSS   20575721   \n",
       "376339          8.744087            1.000000  unique TSS   12662270   \n",
       "376340          8.744087            1.000000  unique TSS   12662270   \n",
       "376341          8.744087            1.000000  unique TSS   12662270   \n",
       "\n",
       "          TSS_end  distance_to_closest_TSS closest_TSS_ref_id  \\\n",
       "0        61077380                        6  ENST00000699964.1   \n",
       "1       160261763                        0  ENST00000442130.1   \n",
       "2       239386587                        8  ENST00000676153.1   \n",
       "3        23217499                        3  ENST00000374619.2   \n",
       "4       233327461                        3  ENST00000771586.1   \n",
       "...           ...                      ...                ...   \n",
       "376337   12662588                       51  ENST00000417071.1   \n",
       "376338   20575776                        0  ENST00000361365.7   \n",
       "376339   12662588                       51  ENST00000417071.1   \n",
       "376340   12662588                       51  ENST00000417071.1   \n",
       "376341   12662588                       51  ENST00000417071.1   \n",
       "\n",
       "       TSS distance category  TSS_name  \n",
       "0                    <=100bp   P3111_A  \n",
       "1                    <=100bp   P5604_A  \n",
       "2                    <=100bp   P7900_A  \n",
       "3                    <=100bp   P1212_A  \n",
       "4                    <=100bp   P7654_A  \n",
       "...                      ...       ...  \n",
       "376337               <=100bp  P80721_A  \n",
       "376338               <=100bp  P80778_A  \n",
       "376339               <=100bp  P80721_A  \n",
       "376340               <=100bp  P80721_A  \n",
       "376341               <=100bp  P80721_A  \n",
       "\n",
       "[376342 rows x 29 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "6065f408-509e-41d2-9d75-a9c38bda2afa",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv('../data_processed/curated_HIT.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e8d657f-16e0-4da1-8147-9d03e204fc4e",
   "metadata": {},
   "source": [
    "## Renaming transcript IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "bae91378-810f-4700-bfb0-f9cf7cab0065",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>closest_TSS_ref_id</th>\n",
       "      <th>TSS distance category</th>\n",
       "      <th>TSS_name</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(61077274, 61077628), (61077916, 61079203)]</td>\n",
       "      <td>[(61077326, 61077628), (61077916, 61079231)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000699992.1</td>\n",
       "      <td>ENST00000699992</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:61077221-61077380_Peak_3111</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.178215</td>\n",
       "      <td>-2.488310</td>\n",
       "      <td>4.820350</td>\n",
       "      <td>0.036971</td>\n",
       "      <td>0.213765</td>\n",
       "      <td>0.044346</td>\n",
       "      <td>minor TSS</td>\n",
       "      <td>61077221</td>\n",
       "      <td>61077380</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000699964.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P3111_A</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(160261734, 160261922), (160281430, 160281892)]</td>\n",
       "      <td>[(160261682, 160261922), (160281430, 160281935)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000756371.1</td>\n",
       "      <td>ENST00000756371</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:160261726-160261763_Peak_5604</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.358414</td>\n",
       "      <td>-1.480300</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>0.481821</td>\n",
       "      <td>0.743874</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>160261726</td>\n",
       "      <td>160261763</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000442130.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P5604_A</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>[(239386568, 239387227), (239492709, 239492807...</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>ENST00000676153</td>\n",
       "      <td>ENSG00000133019</td>\n",
       "      <td>CHRM3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:239386576-239386587_Peak_7900</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.468766</td>\n",
       "      <td>-1.093061</td>\n",
       "      <td>3.127170</td>\n",
       "      <td>0.149901</td>\n",
       "      <td>1.421490</td>\n",
       "      <td>0.454561</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>239386576</td>\n",
       "      <td>239386587</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000676153.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P7900_A</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCT00000022754.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(23191893, 23195001), (23217291, 23217499)]</td>\n",
       "      <td>[(23191895, 23195001), (23217291, 23217502)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>ENST00000374619</td>\n",
       "      <td>ENSG00000179546</td>\n",
       "      <td>HTR1D</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1:23217498-23217499_Peak_1212</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.064644</td>\n",
       "      <td>-3.951339</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>0.553728</td>\n",
       "      <td>0.116743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>23217498</td>\n",
       "      <td>23217499</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000374619.2</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P1212_A</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCT00000039204.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(233319835, 233321155), (233327229, 233327357)]</td>\n",
       "      <td>[(233319834, 233321155), (233327229, 233327455)]</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE-annotated transcripts</td>\n",
       "      <td>ENST00000771587.1</td>\n",
       "      <td>ENST00000771587</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>ENSG00000300424</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr1:233327323-233327461_Peak_7654</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.616169</td>\n",
       "      <td>-0.698603</td>\n",
       "      <td>2.291033</td>\n",
       "      <td>0.268948</td>\n",
       "      <td>2.093885</td>\n",
       "      <td>0.913948</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>233327323</td>\n",
       "      <td>233327461</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000771586.1</td>\n",
       "      <td>&lt;=100bp</td>\n",
       "      <td>P7654_A</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id   chr                                              exons  \\\n",
       "0  ENCT00000006334.1  chr1       [(61077274, 61077628), (61077916, 61079203)]   \n",
       "1  ENCT00000013056.1  chr1   [(160261734, 160261922), (160281430, 160281892)]   \n",
       "2  ENCT00000019620.1  chr1  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3  ENCT00000022754.1  chr1       [(23191893, 23195001), (23217291, 23217499)]   \n",
       "4  ENCT00000039204.1  chr1   [(233319835, 233321155), (233327229, 233327357)]   \n",
       "\n",
       "                                          matched_to  \\\n",
       "0       [(61077326, 61077628), (61077916, 61079231)]   \n",
       "1   [(160261682, 160261922), (160281430, 160281935)]   \n",
       "2  [(239386568, 239387227), (239492709, 239492807...   \n",
       "3       [(23191895, 23195001), (23217291, 23217502)]   \n",
       "4   [(233319834, 233321155), (233327229, 233327455)]   \n",
       "\n",
       "                        category unannotated_exons Strand  \\\n",
       "0  GENCODE-annotated transcripts               NaN      +   \n",
       "1  GENCODE-annotated transcripts               NaN      +   \n",
       "2  GENCODE-annotated transcripts               NaN      +   \n",
       "3  GENCODE-annotated transcripts               NaN      -   \n",
       "4  GENCODE-annotated transcripts               NaN      -   \n",
       "\n",
       "             summarized_category             ref_id  stripped_ref_id  \\\n",
       "0  GENCODE-annotated transcripts  ENST00000699992.1  ENST00000699992   \n",
       "1  GENCODE-annotated transcripts  ENST00000756371.1  ENST00000756371   \n",
       "2  GENCODE-annotated transcripts  ENST00000676153.1  ENST00000676153   \n",
       "3  GENCODE-annotated transcripts  ENST00000374619.2  ENST00000374619   \n",
       "4  GENCODE-annotated transcripts  ENST00000771587.1  ENST00000771587   \n",
       "\n",
       "    Gene stable ID        Gene name       Gene type   gene_category  \\\n",
       "0  ENSG00000162599             NFIA  protein_coding  protein_coding   \n",
       "1  ENSG00000228606         DCAF8-DT          lncRNA          lncRNA   \n",
       "2  ENSG00000133019            CHRM3  protein_coding  protein_coding   \n",
       "3  ENSG00000179546            HTR1D  protein_coding  protein_coding   \n",
       "4  ENSG00000300424  ENSG00000300424          lncRNA          lncRNA   \n",
       "\n",
       "                            peak_name   peak_type  mean_TPM_ref  log2_tx_exp  \\\n",
       "0    chr1:61077221-61077380_Peak_3111  Permissive      0.178215    -2.488310   \n",
       "1  chr1:160261726-160261763_Peak_5604  Permissive      0.358414    -1.480300   \n",
       "2  chr1:239386576-239386587_Peak_7900      Robust      0.468766    -1.093061   \n",
       "3    chr1:23217498-23217499_Peak_1212  Permissive      0.064644    -3.951339   \n",
       "4  chr1:233327323-233327461_Peak_7654  Permissive      0.616169    -0.698603   \n",
       "\n",
       "   gene_exp_ref  relative_tx_exp  TSS_TPM_per_gene  Relative TSS usage  \\\n",
       "0      4.820350         0.036971          0.213765            0.044346   \n",
       "1      0.743874         0.481821          0.743874            1.000000   \n",
       "2      3.127170         0.149901          1.421490            0.454561   \n",
       "3      0.116743         0.553728          0.116743            1.000000   \n",
       "4      2.291033         0.268948          2.093885            0.913948   \n",
       "\n",
       "     TSS_type  TSS_start    TSS_end  distance_to_closest_TSS  \\\n",
       "0   minor TSS   61077221   61077380                        6   \n",
       "1  unique TSS  160261726  160261763                        0   \n",
       "2    main TSS  239386576  239386587                        8   \n",
       "3  unique TSS   23217498   23217499                        3   \n",
       "4    main TSS  233327323  233327461                        3   \n",
       "\n",
       "  closest_TSS_ref_id TSS distance category TSS_name      source  \n",
       "0  ENST00000699964.1               <=100bp  P3111_A  FANTOM_cat  \n",
       "1  ENST00000442130.1               <=100bp  P5604_A  FANTOM_cat  \n",
       "2  ENST00000676153.1               <=100bp  P7900_A  FANTOM_cat  \n",
       "3  ENST00000374619.2               <=100bp  P1212_A  FANTOM_cat  \n",
       "4  ENST00000771586.1               <=100bp  P7654_A  FANTOM_cat  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx_source_map = HIT_hg38_df[['source', 'transcript_id']].drop_duplicates()\n",
    "combined_df = combined_df.merge(tx_source_map, on='transcript_id', how='left')\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "923e8894-56fe-4b39-9cb0-50f6b84ff102",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "prefix_map = {\n",
    "    'FANTOM_cat': 'FTMT',\n",
    "    'PacBio': 'PBT',\n",
    "    'StringTie': 'STRT',\n",
    "    'ENSEMBL': 'NCBIT',\n",
    "}\n",
    "\n",
    "combined_df[\"new_tx_id\"] = None\n",
    "\n",
    "# assign GENCODE-annotated transcripts\n",
    "mask = combined_df[\"category\"] == \"GENCODE-annotated transcripts\"\n",
    "combined_df.loc[mask, \"new_tx_id\"] = combined_df.loc[mask, \"matched_ref_id\"]\n",
    "\n",
    "# generate IDs based on source\n",
    "remaining = combined_df.loc[~mask].copy()\n",
    "\n",
    "for source_name, prefix in prefix_map.items():\n",
    "    source_mask = remaining[\"source\"] == source_name\n",
    "    idx = remaining[source_mask].index\n",
    "    for i, row_idx in enumerate(idx, start=1):\n",
    "        combined_df.at[row_idx, \"new_tx_id\"] = f\"{prefix}{str(i).zfill(7)}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4b72b0ce-545e-4212-8238-27fc45a1db35",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>matched_ref_id</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>best_match_ref_id</th>\n",
       "      <th>largest_overlap_gencode_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>...</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>source</th>\n",
       "      <th>new_tx_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38927</th>\n",
       "      <td>ENST00000235329.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(11980197, 11980484), (11981970, 11982114), (...</td>\n",
       "      <td>[(11980209, 11980484), (11981970, 11982114), (...</td>\n",
       "      <td>ENST00000675817.1</td>\n",
       "      <td>Unannotated transcripts fully contained within...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:11980204-11980451_Peak_666</td>\n",
       "      <td>3.586003</td>\n",
       "      <td>1.842377</td>\n",
       "      <td>15.968124</td>\n",
       "      <td>0.224573</td>\n",
       "      <td>13.568596</td>\n",
       "      <td>0.849730</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38928</th>\n",
       "      <td>ENST00000236914.3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(200027634, 200027911), (200043774, 200043892...</td>\n",
       "      <td>[(200027710, 200027911), (200039658, 200039795...</td>\n",
       "      <td>ENST00000367362.8</td>\n",
       "      <td>Unannotated transcripts fully contained within...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:200027537-200027864_Peak_6480</td>\n",
       "      <td>4.262182</td>\n",
       "      <td>2.091592</td>\n",
       "      <td>21.202160</td>\n",
       "      <td>0.201026</td>\n",
       "      <td>12.154987</td>\n",
       "      <td>0.573290</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38929</th>\n",
       "      <td>ENST00000239457.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(173824659, 173825356), (173826687, 173826786...</td>\n",
       "      <td>[(173824673, 173825356), (173826687, 173826786...</td>\n",
       "      <td>ENST00000649689.2</td>\n",
       "      <td>Unannotated transcripts fully contained within...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:173824474-173824687_Peak_6035</td>\n",
       "      <td>0.288137</td>\n",
       "      <td>-1.795171</td>\n",
       "      <td>6.000629</td>\n",
       "      <td>0.048018</td>\n",
       "      <td>5.583196</td>\n",
       "      <td>0.930435</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38930</th>\n",
       "      <td>ENST00000253251.8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(10032959, 10033694), (10072028, 10072214), (...</td>\n",
       "      <td>[(10032958, 10033694), (10072028, 10072214), (...</td>\n",
       "      <td>ENST00000343090.11</td>\n",
       "      <td>Unannotated transcripts fully contained within...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:10032954-10033166_Peak_545</td>\n",
       "      <td>5.618578</td>\n",
       "      <td>2.490205</td>\n",
       "      <td>10.907193</td>\n",
       "      <td>0.515126</td>\n",
       "      <td>10.885175</td>\n",
       "      <td>0.997981</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38931</th>\n",
       "      <td>ENST00000261443.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(114716920, 114718216), (114718613, 114718745...</td>\n",
       "      <td>[(114716913, 114718216), (114718613, 114718745...</td>\n",
       "      <td>ENST00000339438.10</td>\n",
       "      <td>Unannotated transcripts fully contained within...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:114757960-114757994_Peak_4265</td>\n",
       "      <td>3.338205</td>\n",
       "      <td>1.739073</td>\n",
       "      <td>193.694624</td>\n",
       "      <td>0.017234</td>\n",
       "      <td>137.989329</td>\n",
       "      <td>0.712407</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375357</th>\n",
       "      <td>ENST00000543642.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(37349330, 37349469), (37406209, 37406259), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['ENST00000491253.1', 'ENST00000470290.1', 'EN...</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000378628.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX:37349326-37349369_Peak_78599</td>\n",
       "      <td>0.504692</td>\n",
       "      <td>-0.986525</td>\n",
       "      <td>2.708327</td>\n",
       "      <td>0.186348</td>\n",
       "      <td>2.708327</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0014842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375358</th>\n",
       "      <td>ENST00000545566.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(13653123, 13653188), (13655289, 13655373), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['ENST00000696126.2', 'ENST00000380600.2', 'EN...</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000696128.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX:13653136-13653193_Peak_78335</td>\n",
       "      <td>0.097810</td>\n",
       "      <td>-3.353878</td>\n",
       "      <td>1.731590</td>\n",
       "      <td>0.056486</td>\n",
       "      <td>1.731590</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0014843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375359</th>\n",
       "      <td>ENST00000545618.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(64916370, 64917896), (64919042, 64919204), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['ENST00000374839.8', 'ENST00000488831.5', 'EN...</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000703136.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX:64976443-64976455_Peak_79235</td>\n",
       "      <td>0.071311</td>\n",
       "      <td>-3.809740</td>\n",
       "      <td>3.810785</td>\n",
       "      <td>0.018713</td>\n",
       "      <td>2.025059</td>\n",
       "      <td>0.531402</td>\n",
       "      <td>main TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0014844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375360</th>\n",
       "      <td>ENST00000602791.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(77910631, 77910835), (77962623, 77964890)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['ENST00000341514.11', 'ENST00000602791.2', 'E...</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000602791.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX:77910639-77910745_Peak_79619</td>\n",
       "      <td>0.409518</td>\n",
       "      <td>-1.288001</td>\n",
       "      <td>0.586949</td>\n",
       "      <td>0.697707</td>\n",
       "      <td>0.586949</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0014845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376301</th>\n",
       "      <td>ENST00000538878.1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(13323047, 13323368), (13323555, 13323760), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['ENST00000329134.9', 'ENST00000612274.5', 'EN...</td>\n",
       "      <td>Unannotated transcripts composed entirely of a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000329134.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>chrY:13479608-13479629_Peak_80736</td>\n",
       "      <td>0.028947</td>\n",
       "      <td>-5.110419</td>\n",
       "      <td>5.252041</td>\n",
       "      <td>0.005512</td>\n",
       "      <td>0.373901</td>\n",
       "      <td>0.071192</td>\n",
       "      <td>minor TSS</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NCBIT0014846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14846 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id   chr  \\\n",
       "38927   ENST00000235329.5  chr1   \n",
       "38928   ENST00000236914.3  chr1   \n",
       "38929   ENST00000239457.5  chr1   \n",
       "38930   ENST00000253251.8  chr1   \n",
       "38931   ENST00000261443.5  chr1   \n",
       "...                   ...   ...   \n",
       "375357  ENST00000543642.1  chrX   \n",
       "375358  ENST00000545566.1  chrX   \n",
       "375359  ENST00000545618.1  chrX   \n",
       "375360  ENST00000602791.1  chrX   \n",
       "376301  ENST00000538878.1  chrY   \n",
       "\n",
       "                                                    exons  \\\n",
       "38927   [(11980197, 11980484), (11981970, 11982114), (...   \n",
       "38928   [(200027634, 200027911), (200043774, 200043892...   \n",
       "38929   [(173824659, 173825356), (173826687, 173826786...   \n",
       "38930   [(10032959, 10033694), (10072028, 10072214), (...   \n",
       "38931   [(114716920, 114718216), (114718613, 114718745...   \n",
       "...                                                   ...   \n",
       "375357  [(37349330, 37349469), (37406209, 37406259), (...   \n",
       "375358  [(13653123, 13653188), (13655289, 13655373), (...   \n",
       "375359  [(64916370, 64917896), (64919042, 64919204), (...   \n",
       "375360       [(77910631, 77910835), (77962623, 77964890)]   \n",
       "376301  [(13323047, 13323368), (13323555, 13323760), (...   \n",
       "\n",
       "                                               matched_to  \\\n",
       "38927   [(11980209, 11980484), (11981970, 11982114), (...   \n",
       "38928   [(200027710, 200027911), (200039658, 200039795...   \n",
       "38929   [(173824673, 173825356), (173826687, 173826786...   \n",
       "38930   [(10032958, 10033694), (10072028, 10072214), (...   \n",
       "38931   [(114716913, 114718216), (114718613, 114718745...   \n",
       "...                                                   ...   \n",
       "375357                                                NaN   \n",
       "375358                                                NaN   \n",
       "375359                                                NaN   \n",
       "375360                                                NaN   \n",
       "376301                                                NaN   \n",
       "\n",
       "                                           matched_ref_id  \\\n",
       "38927                                   ENST00000675817.1   \n",
       "38928                                   ENST00000367362.8   \n",
       "38929                                   ENST00000649689.2   \n",
       "38930                                  ENST00000343090.11   \n",
       "38931                                  ENST00000339438.10   \n",
       "...                                                   ...   \n",
       "375357  ['ENST00000491253.1', 'ENST00000470290.1', 'EN...   \n",
       "375358  ['ENST00000696126.2', 'ENST00000380600.2', 'EN...   \n",
       "375359  ['ENST00000374839.8', 'ENST00000488831.5', 'EN...   \n",
       "375360  ['ENST00000341514.11', 'ENST00000602791.2', 'E...   \n",
       "376301  ['ENST00000329134.9', 'ENST00000612274.5', 'EN...   \n",
       "\n",
       "                                                 category unannotated_exons  \\\n",
       "38927   Unannotated transcripts fully contained within...               NaN   \n",
       "38928   Unannotated transcripts fully contained within...               NaN   \n",
       "38929   Unannotated transcripts fully contained within...               NaN   \n",
       "38930   Unannotated transcripts fully contained within...               NaN   \n",
       "38931   Unannotated transcripts fully contained within...               NaN   \n",
       "...                                                   ...               ...   \n",
       "375357  Unannotated transcripts composed entirely of a...               NaN   \n",
       "375358  Unannotated transcripts composed entirely of a...               NaN   \n",
       "375359  Unannotated transcripts composed entirely of a...               NaN   \n",
       "375360  Unannotated transcripts composed entirely of a...               NaN   \n",
       "376301  Unannotated transcripts composed entirely of a...               NaN   \n",
       "\n",
       "        best_match_ref_id largest_overlap_gencode_id Strand  ...  \\\n",
       "38927                 NaN                        NaN      +  ...   \n",
       "38928                 NaN                        NaN      +  ...   \n",
       "38929                 NaN                        NaN      +  ...   \n",
       "38930                 NaN                        NaN      +  ...   \n",
       "38931                 NaN                        NaN      -  ...   \n",
       "...                   ...                        ...    ...  ...   \n",
       "375357  ENST00000378628.9                        NaN      +  ...   \n",
       "375358  ENST00000696128.1                        NaN      +  ...   \n",
       "375359  ENST00000703136.1                        NaN      -  ...   \n",
       "375360  ENST00000602791.2                        NaN      +  ...   \n",
       "376301  ENST00000329134.9                        NaN      -  ...   \n",
       "\n",
       "                                 peak_name mean_TPM_ref log2_tx_exp  \\\n",
       "38927      chr1:11980204-11980451_Peak_666     3.586003    1.842377   \n",
       "38928   chr1:200027537-200027864_Peak_6480     4.262182    2.091592   \n",
       "38929   chr1:173824474-173824687_Peak_6035     0.288137   -1.795171   \n",
       "38930      chr1:10032954-10033166_Peak_545     5.618578    2.490205   \n",
       "38931   chr1:114757960-114757994_Peak_4265     3.338205    1.739073   \n",
       "...                                    ...          ...         ...   \n",
       "375357   chrX:37349326-37349369_Peak_78599     0.504692   -0.986525   \n",
       "375358   chrX:13653136-13653193_Peak_78335     0.097810   -3.353878   \n",
       "375359   chrX:64976443-64976455_Peak_79235     0.071311   -3.809740   \n",
       "375360   chrX:77910639-77910745_Peak_79619     0.409518   -1.288001   \n",
       "376301   chrY:13479608-13479629_Peak_80736     0.028947   -5.110419   \n",
       "\n",
       "       gene_exp_ref relative_tx_exp TSS_TPM_per_gene Relative TSS usage  \\\n",
       "38927     15.968124        0.224573        13.568596           0.849730   \n",
       "38928     21.202160        0.201026        12.154987           0.573290   \n",
       "38929      6.000629        0.048018         5.583196           0.930435   \n",
       "38930     10.907193        0.515126        10.885175           0.997981   \n",
       "38931    193.694624        0.017234       137.989329           0.712407   \n",
       "...             ...             ...              ...                ...   \n",
       "375357     2.708327        0.186348         2.708327           1.000000   \n",
       "375358     1.731590        0.056486         1.731590           1.000000   \n",
       "375359     3.810785        0.018713         2.025059           0.531402   \n",
       "375360     0.586949        0.697707         0.586949           1.000000   \n",
       "376301     5.252041        0.005512         0.373901           0.071192   \n",
       "\n",
       "          TSS_type   source     new_tx_id  \n",
       "38927     main TSS  ENSEMBL  NCBIT0000001  \n",
       "38928     main TSS  ENSEMBL  NCBIT0000002  \n",
       "38929     main TSS  ENSEMBL  NCBIT0000003  \n",
       "38930     main TSS  ENSEMBL  NCBIT0000004  \n",
       "38931     main TSS  ENSEMBL  NCBIT0000005  \n",
       "...            ...      ...           ...  \n",
       "375357  unique TSS  ENSEMBL  NCBIT0014842  \n",
       "375358  unique TSS  ENSEMBL  NCBIT0014843  \n",
       "375359    main TSS  ENSEMBL  NCBIT0014844  \n",
       "375360  unique TSS  ENSEMBL  NCBIT0014845  \n",
       "376301   minor TSS  ENSEMBL  NCBIT0014846  \n",
       "\n",
       "[14846 rows x 28 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[combined_df[\"new_tx_id\"].str.startswith(\"NCBIT\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1d5181e-a345-4a6d-ad8c-ecfee63bafbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combined_df.to_csv('HIT_w_new_tx_ID.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64db3cf0-b1a5-4d5e-80ce-7b05a8799346",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (annotation_project)",
   "language": "python",
   "name": "annotation_project"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
