{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9e028ed1-2a91-4572-a2f1-c965751076b5",
   "metadata": {},
   "source": [
    "# Do novel exons contain T2D GWAS variants?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "318c5f44-791e-4e43-b3b3-4a523bdc71e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e56c5b2-8d43-4745-a95e-4b80e480bc6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load CDS output of non-overlapping exons in predicted CDS\n",
    "with open(\"../data_processed/ORFanage/orfanage_output_with_CDS_no_stop.pkl\", \"rb\") as f:\n",
    "    orfanage_output_with_CDS = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d9c18d74-00ba-4928-b2af-c6e3373bf012",
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_exons = orfanage_output_with_CDS[orfanage_output_with_CDS[\"is_non_overlapping\"] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9394c6d-2388-4041-a350-b065210aed24",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/19452335/ipykernel_4083057/1444729924.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  novel_exons['exon_key'] = novel_exons['Chromosome'].astype(str) + '_' + \\\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>is_non_overlapping</th>\n",
       "      <th>is_unannotated</th>\n",
       "      <th>CDS_matches_exon</th>\n",
       "      <th>is_stop_in_cds</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>exon_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>2932983</td>\n",
       "      <td>2933528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00439837</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ARSL</td>\n",
       "      <td>chrX_2932983_2933528_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>2933334</td>\n",
       "      <td>2933528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02233988</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ARSL</td>\n",
       "      <td>chrX_2933334_2933528_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>7918808</td>\n",
       "      <td>7918984</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02228666</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PNPLA4</td>\n",
       "      <td>chrX_7918808_7918984_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>7918864</td>\n",
       "      <td>7918984</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02237178</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PNPLA4</td>\n",
       "      <td>chrX_7918864_7918984_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>7918904</td>\n",
       "      <td>7918984</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02253319</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PNPLA4</td>\n",
       "      <td>chrX_7918904_7918984_-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  source feature    Start      End score Strand frame  \\\n",
       "1        chrX  custom    exon  2932983  2933528     .      -     .   \n",
       "21       chrX  custom    exon  2933334  2933528     .      -     .   \n",
       "43       chrX  custom    exon  7918808  7918984     .      -     .   \n",
       "59       chrX  custom    exon  7918864  7918984     .      -     .   \n",
       "73       chrX  custom    exon  7918904  7918984     .      -     .   \n",
       "\n",
       "   transcript_id  is_non_overlapping  is_unannotated  CDS_matches_exon  \\\n",
       "1    PBT00439837                True            True             False   \n",
       "21  STRT02233988                True            True             False   \n",
       "43  STRT02228666                True            True             False   \n",
       "59  STRT02237178                True            True             False   \n",
       "73  STRT02253319                True            True             False   \n",
       "\n",
       "    is_stop_in_cds Gene name                exon_key  \n",
       "1            False      ARSL  chrX_2932983_2933528_-  \n",
       "21           False      ARSL  chrX_2933334_2933528_-  \n",
       "43           False    PNPLA4  chrX_7918808_7918984_-  \n",
       "59           False    PNPLA4  chrX_7918864_7918984_-  \n",
       "73           False    PNPLA4  chrX_7918904_7918984_-  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_exons['exon_key'] = novel_exons['Chromosome'].astype(str) + '_' + \\\n",
    "                        novel_exons['Start'].astype(str) + '_' + \\\n",
    "                        novel_exons['End'].astype(str) + '_' + \\\n",
    "                        novel_exons['Strand'].astype(str)\n",
    "novel_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e5e9f638-2e41-4e57-a8ab-19ad52b151b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10847"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_exons[\"exon_key\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8189e319-582a-42d7-8e2d-fd446da6e8b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>is_non_overlapping</th>\n",
       "      <th>is_unannotated</th>\n",
       "      <th>CDS_matches_exon</th>\n",
       "      <th>is_stop_in_cds</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>exon_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>13715720</td>\n",
       "      <td>13715794</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02287699</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>OFD1</td>\n",
       "      <td>chrX_13715720_13715794_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>13715720</td>\n",
       "      <td>13715794</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02244133</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>OFD1</td>\n",
       "      <td>chrX_13715720_13715794_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>15753960</td>\n",
       "      <td>15754004</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02274688</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CA5B</td>\n",
       "      <td>chrX_15753960_15754004_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>15753960</td>\n",
       "      <td>15754004</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02271968</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CA5B</td>\n",
       "      <td>chrX_15753960_15754004_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>18930763</td>\n",
       "      <td>18930847</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02275181</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PHKA2</td>\n",
       "      <td>chrX_18930763_18930847_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272264</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50276134</td>\n",
       "      <td>50276188</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00285527</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50276134_50276188_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272299</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50295377</td>\n",
       "      <td>50295429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00285527</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50295377_50295429_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272332</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50276134</td>\n",
       "      <td>50276188</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00285553</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50276134_50276188_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272396</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50276134</td>\n",
       "      <td>50276227</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00285567</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50276134_50276227_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272462</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50276134</td>\n",
       "      <td>50276227</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00285569</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50276134_50276227_-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3354 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  source feature     Start       End score Strand frame  \\\n",
       "265          chrX  custom    exon  13715720  13715794     .      +     .   \n",
       "313          chrX  custom    exon  13715720  13715794     .      +     .   \n",
       "519          chrX  custom    exon  15753960  15754004     .      +     .   \n",
       "543          chrX  custom    exon  15753960  15754004     .      +     .   \n",
       "701          chrX  custom    exon  18930763  18930847     .      -     .   \n",
       "...           ...     ...     ...       ...       ...   ...    ...   ...   \n",
       "272264      chr22  custom    exon  50276134  50276188     .      -     .   \n",
       "272299      chr22  custom    exon  50295377  50295429     .      -     .   \n",
       "272332      chr22  custom    exon  50276134  50276188     .      -     .   \n",
       "272396      chr22  custom    exon  50276134  50276227     .      -     .   \n",
       "272462      chr22  custom    exon  50276134  50276227     .      -     .   \n",
       "\n",
       "       transcript_id  is_non_overlapping  is_unannotated  CDS_matches_exon  \\\n",
       "265     STRT02287699                True            True             False   \n",
       "313     STRT02244133                True            True             False   \n",
       "519     STRT02274688                True            True             False   \n",
       "543     STRT02271968                True            True             False   \n",
       "701     STRT02275181                True            True             False   \n",
       "...              ...                 ...             ...               ...   \n",
       "272264   PBT00285527                True            True             False   \n",
       "272299   PBT00285527                True            True             False   \n",
       "272332   PBT00285553                True            True             False   \n",
       "272396   PBT00285567                True            True             False   \n",
       "272462   PBT00285569                True            True             False   \n",
       "\n",
       "        is_stop_in_cds Gene name                   exon_key  \n",
       "265              False      OFD1   chrX_13715720_13715794_+  \n",
       "313              False      OFD1   chrX_13715720_13715794_+  \n",
       "519              False      CA5B   chrX_15753960_15754004_+  \n",
       "543              False      CA5B   chrX_15753960_15754004_+  \n",
       "701              False     PHKA2   chrX_18930763_18930847_-  \n",
       "...                ...       ...                        ...  \n",
       "272264           False    PLXNB2  chr22_50276134_50276188_-  \n",
       "272299           False    PLXNB2  chr22_50295377_50295429_-  \n",
       "272332           False    PLXNB2  chr22_50276134_50276188_-  \n",
       "272396           False    PLXNB2  chr22_50276134_50276227_-  \n",
       "272462           False    PLXNB2  chr22_50276134_50276227_-  \n",
       "\n",
       "[3354 rows x 15 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# same novel exon in multiple transcripts\n",
    "duplicates_mask = novel_exons[\"exon_key\"].duplicated(keep=False)\n",
    "novel_exons[duplicates_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "039b8e16-b259-4431-90a7-2b913b3b0d59",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>is_non_overlapping</th>\n",
       "      <th>is_unannotated</th>\n",
       "      <th>CDS_matches_exon</th>\n",
       "      <th>is_stop_in_cds</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>exon_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>9466385</td>\n",
       "      <td>9466457</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02262527</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TBL1X</td>\n",
       "      <td>chrX_9466385_9466457_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>9719997</td>\n",
       "      <td>9721277</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02262527</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TBL1X</td>\n",
       "      <td>chrX_9719997_9721277_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>16653206</td>\n",
       "      <td>16653389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02255865</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CTPS2</td>\n",
       "      <td>chrX_16653206_16653389_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>16707409</td>\n",
       "      <td>16707503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02255865</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CTPS2</td>\n",
       "      <td>chrX_16707409_16707503_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>chrX</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>23832331</td>\n",
       "      <td>23832407</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02274094</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>APOO</td>\n",
       "      <td>chrX_23832331_23832407_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272041</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50204919</td>\n",
       "      <td>50205024</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01398653</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>SELENOO</td>\n",
       "      <td>chr22_50204919_50205024_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272196</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50276134</td>\n",
       "      <td>50276227</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01405431</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50276134_50276227_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272231</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50295377</td>\n",
       "      <td>50295429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01405431</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50295377_50295429_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272264</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50276134</td>\n",
       "      <td>50276188</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00285527</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50276134_50276188_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272299</th>\n",
       "      <td>chr22</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>50295377</td>\n",
       "      <td>50295429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PBT00285527</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLXNB2</td>\n",
       "      <td>chr22_50295377_50295429_-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3579 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  source feature     Start       End score Strand frame  \\\n",
       "169          chrX  custom    exon   9466385   9466457     .      +     .   \n",
       "186          chrX  custom    exon   9719997   9721277     .      +     .   \n",
       "572          chrX  custom    exon  16653206  16653389     .      -     .   \n",
       "585          chrX  custom    exon  16707409  16707503     .      -     .   \n",
       "1161         chrX  custom    exon  23832331  23832407     .      -     .   \n",
       "...           ...     ...     ...       ...       ...   ...    ...   ...   \n",
       "272041      chr22  custom    exon  50204919  50205024     .      +     .   \n",
       "272196      chr22  custom    exon  50276134  50276227     .      -     .   \n",
       "272231      chr22  custom    exon  50295377  50295429     .      -     .   \n",
       "272264      chr22  custom    exon  50276134  50276188     .      -     .   \n",
       "272299      chr22  custom    exon  50295377  50295429     .      -     .   \n",
       "\n",
       "       transcript_id  is_non_overlapping  is_unannotated  CDS_matches_exon  \\\n",
       "169     STRT02262527                True            True             False   \n",
       "186     STRT02262527                True            True             False   \n",
       "572     STRT02255865                True            True             False   \n",
       "585     STRT02255865                True            True             False   \n",
       "1161    STRT02274094                True            True             False   \n",
       "...              ...                 ...             ...               ...   \n",
       "272041  STRT01398653                True            True             False   \n",
       "272196  STRT01405431                True            True             False   \n",
       "272231  STRT01405431                True            True             False   \n",
       "272264   PBT00285527                True            True             False   \n",
       "272299   PBT00285527                True            True             False   \n",
       "\n",
       "        is_stop_in_cds Gene name                   exon_key  \n",
       "169              False     TBL1X     chrX_9466385_9466457_+  \n",
       "186              False     TBL1X     chrX_9719997_9721277_+  \n",
       "572              False     CTPS2   chrX_16653206_16653389_-  \n",
       "585              False     CTPS2   chrX_16707409_16707503_-  \n",
       "1161             False      APOO   chrX_23832331_23832407_-  \n",
       "...                ...       ...                        ...  \n",
       "272041           False   SELENOO  chr22_50204919_50205024_+  \n",
       "272196           False    PLXNB2  chr22_50276134_50276227_-  \n",
       "272231           False    PLXNB2  chr22_50295377_50295429_-  \n",
       "272264           False    PLXNB2  chr22_50276134_50276188_-  \n",
       "272299           False    PLXNB2  chr22_50295377_50295429_-  \n",
       "\n",
       "[3579 rows x 15 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# one transcript with multiple novel exons\n",
    "duplicates_mask = novel_exons[\"transcript_id\"].duplicated(keep=False)\n",
    "novel_exons[duplicates_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fc9c9f12-08d3-4446-9bde-7767d8720004",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>229537208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133273813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133274084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133270061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133271745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        pos\n",
       "0  chr1  229537208\n",
       "1  chr9  133273813\n",
       "2  chr9  133274084\n",
       "3  chr9  133270061\n",
       "4  chr9  133271745"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load variant positions\n",
    "column_names = ['chr','pos']\n",
    "variants = pd.read_csv(\"../data_raw/credset_liftover_positions.tsv\", sep=\"\\t\", header=None, names=column_names)\n",
    "variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cd3458da-c879-4568-81d3-393598846359",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7902"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2a300d3b-1206-425a-88fd-e748e52144c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.merge(variants, novel_exons, left_on=\"chr\", right_on=\"Chromosome\", how=\"left\")\n",
    "merged[\"in_exon\"] = (merged[\"pos\"] >= merged[\"Start\"]) & (merged[\"pos\"] <= merged[\"End\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3f76ecf5-9bcc-4dd4-b5e5-d9d68543ef9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>is_non_overlapping</th>\n",
       "      <th>is_unannotated</th>\n",
       "      <th>CDS_matches_exon</th>\n",
       "      <th>is_stop_in_cds</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>exon_key</th>\n",
       "      <th>in_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>112454</th>\n",
       "      <td>chr10</td>\n",
       "      <td>97230382</td>\n",
       "      <td>chr10</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>97230320</td>\n",
       "      <td>97230387</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT00277921</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ARHGAP19</td>\n",
       "      <td>chr10_97230320_97230387_-</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519840</th>\n",
       "      <td>chr17</td>\n",
       "      <td>67896391</td>\n",
       "      <td>chr17</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>67895489</td>\n",
       "      <td>67898418</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT00929838</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>BPTF</td>\n",
       "      <td>chr17_67895489_67898418_+</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521724</th>\n",
       "      <td>chr17</td>\n",
       "      <td>67896903</td>\n",
       "      <td>chr17</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>67895489</td>\n",
       "      <td>67898418</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT00929838</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>BPTF</td>\n",
       "      <td>chr17_67895489_67898418_+</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522195</th>\n",
       "      <td>chr17</td>\n",
       "      <td>67896636</td>\n",
       "      <td>chr17</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>67895489</td>\n",
       "      <td>67898418</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT00929838</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>BPTF</td>\n",
       "      <td>chr17_67895489_67898418_+</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525483</th>\n",
       "      <td>chr17</td>\n",
       "      <td>67951854</td>\n",
       "      <td>chr17</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>67951730</td>\n",
       "      <td>67951922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>HBMT00000608352.1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>BPTF</td>\n",
       "      <td>chr17_67951730_67951922_+</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5182746</th>\n",
       "      <td>chr8</td>\n",
       "      <td>115567092</td>\n",
       "      <td>chr8</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>115518120</td>\n",
       "      <td>115586630</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02087866</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TRPS1</td>\n",
       "      <td>chr8_115518120_115586630_-</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5183350</th>\n",
       "      <td>chr8</td>\n",
       "      <td>115566821</td>\n",
       "      <td>chr8</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>115518120</td>\n",
       "      <td>115586630</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02087866</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TRPS1</td>\n",
       "      <td>chr8_115518120_115586630_-</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5184558</th>\n",
       "      <td>chr8</td>\n",
       "      <td>115571890</td>\n",
       "      <td>chr8</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>115518120</td>\n",
       "      <td>115586630</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02087866</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TRPS1</td>\n",
       "      <td>chr8_115518120_115586630_-</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5185766</th>\n",
       "      <td>chr8</td>\n",
       "      <td>115571642</td>\n",
       "      <td>chr8</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>115518120</td>\n",
       "      <td>115586630</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02087866</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TRPS1</td>\n",
       "      <td>chr8_115518120_115586630_-</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5350501</th>\n",
       "      <td>chr4</td>\n",
       "      <td>6295919</td>\n",
       "      <td>chr4</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>6295916</td>\n",
       "      <td>6296402</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01577036</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>WFS1</td>\n",
       "      <td>chr4_6295916_6296402_+</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr        pos Chromosome  source feature      Start        End  \\\n",
       "112454   chr10   97230382      chr10  custom    exon   97230320   97230387   \n",
       "519840   chr17   67896391      chr17  custom    exon   67895489   67898418   \n",
       "521724   chr17   67896903      chr17  custom    exon   67895489   67898418   \n",
       "522195   chr17   67896636      chr17  custom    exon   67895489   67898418   \n",
       "525483   chr17   67951854      chr17  custom    exon   67951730   67951922   \n",
       "...        ...        ...        ...     ...     ...        ...        ...   \n",
       "5182746   chr8  115567092       chr8  custom    exon  115518120  115586630   \n",
       "5183350   chr8  115566821       chr8  custom    exon  115518120  115586630   \n",
       "5184558   chr8  115571890       chr8  custom    exon  115518120  115586630   \n",
       "5185766   chr8  115571642       chr8  custom    exon  115518120  115586630   \n",
       "5350501   chr4    6295919       chr4  custom    exon    6295916    6296402   \n",
       "\n",
       "        score Strand frame      transcript_id  is_non_overlapping  \\\n",
       "112454      .      -     .       STRT00277921                True   \n",
       "519840      .      +     .       STRT00929838                True   \n",
       "521724      .      +     .       STRT00929838                True   \n",
       "522195      .      +     .       STRT00929838                True   \n",
       "525483      .      +     .  HBMT00000608352.1                True   \n",
       "...       ...    ...   ...                ...                 ...   \n",
       "5182746     .      -     .       STRT02087866                True   \n",
       "5183350     .      -     .       STRT02087866                True   \n",
       "5184558     .      -     .       STRT02087866                True   \n",
       "5185766     .      -     .       STRT02087866                True   \n",
       "5350501     .      +     .       STRT01577036                True   \n",
       "\n",
       "         is_unannotated  CDS_matches_exon  is_stop_in_cds Gene name  \\\n",
       "112454             True             False           False  ARHGAP19   \n",
       "519840             True             False           False      BPTF   \n",
       "521724             True             False           False      BPTF   \n",
       "522195             True             False           False      BPTF   \n",
       "525483             True             False           False      BPTF   \n",
       "...                 ...               ...             ...       ...   \n",
       "5182746            True             False           False     TRPS1   \n",
       "5183350            True             False           False     TRPS1   \n",
       "5184558            True             False           False     TRPS1   \n",
       "5185766            True             False           False     TRPS1   \n",
       "5350501            True             False           False      WFS1   \n",
       "\n",
       "                           exon_key  in_exon  \n",
       "112454    chr10_97230320_97230387_-     True  \n",
       "519840    chr17_67895489_67898418_+     True  \n",
       "521724    chr17_67895489_67898418_+     True  \n",
       "522195    chr17_67895489_67898418_+     True  \n",
       "525483    chr17_67951730_67951922_+     True  \n",
       "...                             ...      ...  \n",
       "5182746  chr8_115518120_115586630_-     True  \n",
       "5183350  chr8_115518120_115586630_-     True  \n",
       "5184558  chr8_115518120_115586630_-     True  \n",
       "5185766  chr8_115518120_115586630_-     True  \n",
       "5350501      chr4_6295916_6296402_+     True  \n",
       "\n",
       "[67 rows x 18 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits = merged[merged[\"in_exon\"]]\n",
    "hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d19a08aa-ddae-4183-b4e4-9a8488d59d95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "67"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits[\"pos\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4c8cbb12-2ed8-4df0-9a50-82b2b938f116",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits[\"exon_key\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1adb0197-6c6f-457f-b485-09245ce907cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ARHGAP19', 'BPTF', 'PCGF3', 'FAF1', 'DEUP1', 'SLC9B2', 'DPEP1',\n",
       "       'TCF7L2', 'TRPS1', 'WFS1'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits[\"Gene name\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "56e1a037-ba06-41e9-90ec-728e8f8bb470",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr10_97230320_97230387_-', 'chr17_67895489_67898418_+',\n",
       "       'chr17_67951730_67951922_+', 'chr4_747902_749477_+',\n",
       "       'chr1_50779947_50781044_-', 'chr11_93397259_93398926_+',\n",
       "       'chr4_103035389_103036518_-', 'chr16_89640594_89641859_+',\n",
       "       'chr10_112968330_112981419_+', 'chr8_115518120_115586630_-',\n",
       "       'chr4_6295916_6296402_+'], dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits[\"exon_key\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5e00ff76-05ac-4933-8aa8-88500f289f07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon_key\n",
       "chr10_112968330_112981419_+     5\n",
       "chr10_97230320_97230387_-       1\n",
       "chr11_93397259_93398926_+       1\n",
       "chr16_89640594_89641859_+       1\n",
       "chr17_67895489_67898418_+       5\n",
       "chr17_67951730_67951922_+       1\n",
       "chr1_50779947_50781044_-        1\n",
       "chr4_103035389_103036518_-      1\n",
       "chr4_6295916_6296402_+          1\n",
       "chr4_747902_749477_+            2\n",
       "chr8_115518120_115586630_-     48\n",
       "Name: pos, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits.groupby(\"exon_key\")[\"pos\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "99c81afd-ee40-4311-b70c-6d418dbcc07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "87ac53c1-2e9a-4958-9eca-f7a930b33d17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_variants_in_CDS</th>\n",
       "      <th>variant_positions_in_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARHGAP19</td>\n",
       "      <td>STRT00277921</td>\n",
       "      <td>1</td>\n",
       "      <td>[97230382]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>HBMT00000608352.1</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00930465</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00931774</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00955719</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00955720</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00936211</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00941653</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00952645</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00955391</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>STRT00929838</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCGF3</td>\n",
       "      <td>ENCT00000315849.1</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PCGF3</td>\n",
       "      <td>STRT01574493</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PCGF3</td>\n",
       "      <td>STRT01601181</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PCGF3</td>\n",
       "      <td>STRT01642293</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PCGF3</td>\n",
       "      <td>STRT01622000</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00173939</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00132128</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00089753</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00091411</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00151331</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00202875</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00165318</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00204082</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00151332</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00165319</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00208065</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00175886</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FAF1</td>\n",
       "      <td>STRT00119806</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00376646</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00381588</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00370288</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00427895</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00424550</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00413137</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00370785</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>DEUP1</td>\n",
       "      <td>STRT00354660</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SLC9B2</td>\n",
       "      <td>STRT01650612</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SLC9B2</td>\n",
       "      <td>STRT01586604</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>SLC9B2</td>\n",
       "      <td>STRT01608571</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>SLC9B2</td>\n",
       "      <td>STRT01585615</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>DPEP1</td>\n",
       "      <td>STRT00822061</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>DPEP1</td>\n",
       "      <td>STRT00798764</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>STRT00314500</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>FTMT23900013247.1</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>TRPS1</td>\n",
       "      <td>STRT02087866</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>WFS1</td>\n",
       "      <td>STRT01577036</td>\n",
       "      <td>1</td>\n",
       "      <td>[6295919]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene name      transcript_id  num_variants_in_CDS variant_positions_in_CDS\n",
       "0   ARHGAP19       STRT00277921                    1               [97230382]\n",
       "1       BPTF  HBMT00000608352.1                    0                       []\n",
       "2       BPTF       STRT00930465                    0                       []\n",
       "3       BPTF       STRT00931774                    0                       []\n",
       "4       BPTF       STRT00955719                    0                       []\n",
       "5       BPTF       STRT00955720                    0                       []\n",
       "6       BPTF       STRT00936211                    0                       []\n",
       "7       BPTF       STRT00941653                    0                       []\n",
       "8       BPTF       STRT00952645                    0                       []\n",
       "9       BPTF       STRT00955391                    0                       []\n",
       "10      BPTF       STRT00929838                    0                       []\n",
       "11     PCGF3  ENCT00000315849.1                    0                       []\n",
       "12     PCGF3       STRT01574493                    0                       []\n",
       "13     PCGF3       STRT01601181                    0                       []\n",
       "14     PCGF3       STRT01642293                    0                       []\n",
       "15     PCGF3       STRT01622000                    0                       []\n",
       "16      FAF1       STRT00173939                    0                       []\n",
       "17      FAF1       STRT00132128                    0                       []\n",
       "18      FAF1       STRT00089753                    0                       []\n",
       "19      FAF1       STRT00091411                    0                       []\n",
       "20      FAF1       STRT00151331                    0                       []\n",
       "21      FAF1       STRT00202875                    0                       []\n",
       "22      FAF1       STRT00165318                    0                       []\n",
       "23      FAF1       STRT00204082                    0                       []\n",
       "24      FAF1       STRT00151332                    0                       []\n",
       "25      FAF1       STRT00165319                    0                       []\n",
       "26      FAF1       STRT00208065                    0                       []\n",
       "27      FAF1       STRT00175886                    0                       []\n",
       "28      FAF1       STRT00119806                    0                       []\n",
       "29     DEUP1       STRT00376646                    0                       []\n",
       "30     DEUP1       STRT00381588                    0                       []\n",
       "31     DEUP1       STRT00370288                    0                       []\n",
       "32     DEUP1       STRT00427895                    0                       []\n",
       "33     DEUP1       STRT00424550                    0                       []\n",
       "34     DEUP1       STRT00413137                    0                       []\n",
       "35     DEUP1       STRT00370785                    0                       []\n",
       "36     DEUP1       STRT00354660                    0                       []\n",
       "37    SLC9B2       STRT01650612                    0                       []\n",
       "38    SLC9B2       STRT01586604                    0                       []\n",
       "39    SLC9B2       STRT01608571                    0                       []\n",
       "40    SLC9B2       STRT01585615                    0                       []\n",
       "41     DPEP1       STRT00822061                    0                       []\n",
       "42     DPEP1       STRT00798764                    0                       []\n",
       "43    TCF7L2       STRT00314500                    0                       []\n",
       "44    TCF7L2  FTMT23900013247.1                    0                       []\n",
       "45     TRPS1       STRT02087866                    0                       []\n",
       "46      WFS1       STRT01577036                    1                [6295919]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Are the variants actually in the predicted CDS regions?\n",
    "\n",
    "results = []  \n",
    "\n",
    "for gene in hits[\"Gene name\"].unique():\n",
    "    gene_hits = hits[hits[\"Gene name\"] == gene]\n",
    "    \n",
    "    # find all tx for that gene \n",
    "    transcripts = orfanage_output_with_CDS.loc[\n",
    "        orfanage_output_with_CDS[\"Gene name\"] == gene, \"transcript_id\"\n",
    "    ].unique()\n",
    "    \n",
    "    for tx in transcripts:\n",
    "        # CDS regions that match novel exon\n",
    "        cds_rows = orfanage_output_with_CDS[\n",
    "            (orfanage_output_with_CDS[\"transcript_id\"] == tx)\n",
    "            & (orfanage_output_with_CDS[\"CDS_matches_exon\"] == True)\n",
    "        ]\n",
    "        cds_ranges = list(zip(cds_rows[\"Start\"], cds_rows[\"End\"]))\n",
    "        \n",
    "        # mark which variants fall inside CDS\n",
    "        gene_hits_tx = gene_hits.copy()\n",
    "        gene_hits_tx[\"in_CDS\"] = gene_hits_tx[\"pos\"].apply(\n",
    "            lambda p: any(start <= p <= end for start, end in cds_ranges)\n",
    "        )\n",
    "        \n",
    "        # keep only variants inside CDS\n",
    "        hits_in_cds = gene_hits_tx[gene_hits_tx[\"in_CDS\"]]\n",
    "        \n",
    "        results.append({\n",
    "            \"Gene name\": gene,\n",
    "            \"transcript_id\": tx,\n",
    "            \"num_variants_in_CDS\": len(hits_in_cds),\n",
    "            \"variant_positions_in_CDS\": hits_in_cds[\"pos\"].tolist()\n",
    "        })\n",
    "\n",
    "cds_hits = pd.DataFrame(results)\n",
    "cds_hits\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "39ba98c1-2f7e-422f-bfda-cc176fcd426d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>is_non_overlapping</th>\n",
       "      <th>is_unannotated</th>\n",
       "      <th>CDS_matches_exon</th>\n",
       "      <th>is_stop_in_cds</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>exon_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>170137</th>\n",
       "      <td>chr10</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>97230320</td>\n",
       "      <td>97230387</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT00277921</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ARHGAP19</td>\n",
       "      <td>chr10_97230320_97230387_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170145</th>\n",
       "      <td>chr10</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>97265242</td>\n",
       "      <td>97265334</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT00277921</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ARHGAP19</td>\n",
       "      <td>chr10_97265242_97265334_-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  source feature     Start       End score Strand frame  \\\n",
       "170137      chr10  custom    exon  97230320  97230387     .      -     .   \n",
       "170145      chr10  custom    exon  97265242  97265334     .      -     .   \n",
       "\n",
       "       transcript_id  is_non_overlapping  is_unannotated  CDS_matches_exon  \\\n",
       "170137  STRT00277921                True            True             False   \n",
       "170145  STRT00277921                True            True             False   \n",
       "\n",
       "        is_stop_in_cds Gene name                   exon_key  \n",
       "170137           False  ARHGAP19  chr10_97230320_97230387_-  \n",
       "170145           False  ARHGAP19  chr10_97265242_97265334_-  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_exons[novel_exons[\"Gene name\"] == \"ARHGAP19\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bffc12dd-e9f0-4da1-81ad-b797f0572f6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>is_non_overlapping</th>\n",
       "      <th>is_unannotated</th>\n",
       "      <th>CDS_matches_exon</th>\n",
       "      <th>is_stop_in_cds</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>exon_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>75806</th>\n",
       "      <td>chr4</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>6295916</td>\n",
       "      <td>6296402</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01577036</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>WFS1</td>\n",
       "      <td>chr4_6295916_6296402_+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome  source feature    Start      End score Strand frame  \\\n",
       "75806       chr4  custom    exon  6295916  6296402     .      +     .   \n",
       "\n",
       "      transcript_id  is_non_overlapping  is_unannotated  CDS_matches_exon  \\\n",
       "75806  STRT01577036                True            True             False   \n",
       "\n",
       "       is_stop_in_cds Gene name                exon_key  \n",
       "75806           False      WFS1  chr4_6295916_6296402_+  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_exons[novel_exons[\"Gene name\"] == \"WFS1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64163609-b5e8-476b-b89e-eeea7c0ffb8b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (annotation_project)",
   "language": "python",
   "name": "annotation_project"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
