{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9cf98f48-d87b-402f-84b9-eb9e9c21261a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/jferrer/jmidgley/annotation_project/lib64/python3.9/site-packages/sorted_nearest/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import pyranges as pr\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import csv\n",
    "import re\n",
    "from itertools import combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1a15da3f-e258-4322-bd21-9fe350b374cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', None) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "befcf50c-1a6d-40f8-8302-72ea87d49dc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/19438950/ipykernel_1775176/558083981.py:1: DtypeWarning: Columns (3,5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  combined_df = pd.read_csv( '../data_processed/curated_HIT.tsv', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "combined_df = pd.read_csv( '../data_processed/curated_HIT.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "25836ab5-574a-4f31-9631-6a849aeae9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert to GTF format\n",
    "import ast\n",
    "\n",
    "# Convert stringified exon lists to actual lists of tuples\n",
    "def parse_exons(exon_str):\n",
    "    if isinstance(exon_str, str):\n",
    "        try:\n",
    "            return ast.literal_eval(exon_str)\n",
    "        except:\n",
    "            return None \n",
    "    return exon_str  # already valid\n",
    "\n",
    "combined_df[\"exons\"] = combined_df[\"exons\"].apply(parse_exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5ba05794-50a6-467f-b1bb-e01e835cb95d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>61077274</td>\n",
       "      <td>61077628</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>61077916</td>\n",
       "      <td>61079203</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>160261734</td>\n",
       "      <td>160261922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>160281430</td>\n",
       "      <td>160281892</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>239386568</td>\n",
       "      <td>239387227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000133019\"; transcript_id \"ENCT00000019620.1\"; gene_name \"CHRM3\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3966117</th>\n",
       "      <td>chrY</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>12847239</td>\n",
       "      <td>12847327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3966118</th>\n",
       "      <td>chrY</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>12856340</td>\n",
       "      <td>12856496</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3966119</th>\n",
       "      <td>chrY</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>12856633</td>\n",
       "      <td>12856845</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3966120</th>\n",
       "      <td>chrY</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>12857566</td>\n",
       "      <td>12857661</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3966121</th>\n",
       "      <td>chrY</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>12859279</td>\n",
       "      <td>12860832</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3966122 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome  source feature      start        end score strand frame  \\\n",
       "0             chr1  custom    exon   61077274   61077628     .      +     .   \n",
       "1             chr1  custom    exon   61077916   61079203     .      +     .   \n",
       "2             chr1  custom    exon  160261734  160261922     .      +     .   \n",
       "3             chr1  custom    exon  160281430  160281892     .      +     .   \n",
       "4             chr1  custom    exon  239386568  239387227     .      +     .   \n",
       "...            ...     ...     ...        ...        ...   ...    ...   ...   \n",
       "3966117       chrY  custom    exon   12847239   12847327     .      +     .   \n",
       "3966118       chrY  custom    exon   12856340   12856496     .      +     .   \n",
       "3966119       chrY  custom    exon   12856633   12856845     .      +     .   \n",
       "3966120       chrY  custom    exon   12857566   12857661     .      +     .   \n",
       "3966121       chrY  custom    exon   12859279   12860832     .      +     .   \n",
       "\n",
       "                                                                                                             attribute  \n",
       "0         gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";   \n",
       "1         gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";   \n",
       "2             gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";   \n",
       "3             gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";   \n",
       "4        gene_id \"ENSG00000133019\"; transcript_id \"ENCT00000019620.1\"; gene_name \"CHRM3\"; gene_type \"protein_coding\";   \n",
       "...                                                                                                                ...  \n",
       "3966117       gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";   \n",
       "3966118       gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";   \n",
       "3966119       gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";   \n",
       "3966120       gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";   \n",
       "3966121       gene_id \"ENSG00000114374\"; transcript_id \"STRT02296304\"; gene_name \"USP9Y\"; gene_type \"protein_coding\";   \n",
       "\n",
       "[3966122 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_rows = []\n",
    "\n",
    "for _, row in combined_df.iterrows():\n",
    "    transcript_id = row[\"transcript_id\"]\n",
    "    gene_id = row.get(\"Gene stable ID\", \".\")\n",
    "    gene_name = row.get(\"Gene name\", \".\")\n",
    "    gene_type = row.get(\"Gene type\", \".\")\n",
    "    chromosome = row[\"chr\"]\n",
    "    strand = row[\"Strand\"]\n",
    "    source = \"custom\"\n",
    "    score = \".\"\n",
    "    frame = \".\"\n",
    "\n",
    "    exons = row.get(\"exons\")\n",
    "\n",
    "    # Skip missing or malformed exon lists\n",
    "    if not isinstance(exons, list) or len(exons) == 0:\n",
    "        continue\n",
    "\n",
    "    for i, (start, end) in enumerate(exons, 1):\n",
    "        if not isinstance((start, end), tuple) or len((start, end)) != 2:\n",
    "            continue\n",
    "\n",
    "        attributes = (\n",
    "            f'gene_id \"{gene_id}\"; '\n",
    "            f'transcript_id \"{transcript_id}\"; '\n",
    "            f'gene_name \"{gene_name}\"; '\n",
    "            f'gene_type \"{gene_type}\"; '\n",
    "            #f'exon_number \"{i}\";'\n",
    "        )\n",
    "\n",
    "        gtf_rows.append([\n",
    "            chromosome,\n",
    "            source,\n",
    "            \"exon\",\n",
    "            start,\n",
    "            end,\n",
    "            score,\n",
    "            strand,\n",
    "            frame,\n",
    "            attributes\n",
    "        ])\n",
    "\n",
    "# Convert to DataFrame\n",
    "gtf_df = pd.DataFrame(gtf_rows, columns=[\n",
    "    \"chromosome\", \"source\", \"feature\", \"start\", \"end\",\n",
    "    \"score\", \"strand\", \"frame\", \"attribute\"\n",
    "])\n",
    "gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7b88cbab-e74b-42af-874a-7a923d6aa925",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to file\n",
    "gtf_df.to_csv(\"../data_processed/curated_HIT.gtf\", sep=\"\\t\", index=False, header=False, quoting=csv.QUOTE_NONE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (annotation_project)",
   "language": "python",
   "name": "annotation_project"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
