{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ee9701cd-2571-42ad-bbca-82cdb5d08610",
   "metadata": {},
   "source": [
    "# Comparing HIT annotation to GENCODE using GFFCompare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9abb948b-9642-4ad4-af00-42f76906b596",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/jferrer/jmidgley/annotation_project/lib64/python3.9/site-packages/sorted_nearest/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "from matplotlib import font_manager\n",
    "import csv\n",
    "import pyranges as pr\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'retina'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e7229b04-2c73-4619-90a8-42782e304dc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_column_names = ['chromosome','source','feature','start','end','score','strand','frame','attribute']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7a4c3144-1fdf-4326-bab5-53e118f13df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_attributes(df, *attribute_type):\n",
    "    \"\"\"split attribute of a gtf into separate columns\"\"\"\n",
    "    try: \n",
    "        l_attributes = df.strip(';').split('; ')\n",
    "        attribute_position = [i for i, x in enumerate(l_attributes) if x.startswith(attribute_type)]\n",
    "        attribute = l_attributes[attribute_position[0]].split(' ')[1].strip('\"')\n",
    "        return attribute\n",
    "    except Exception:\n",
    "        return np.nan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19cb10de-29bf-47a8-9051-906b9acc9812",
   "metadata": {},
   "source": [
    "## Comparison with hg38"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b93b19-604d-4627-8bc8-64f27556c7da",
   "metadata": {},
   "source": [
    "Ran command: gffcompare -r gencode.v47.annotation.gtf -o hg38 Human_Islet_Transcriptome_v2.2.HG38.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "794e7d25-f47d-4ba0-81d0-e04689a4dbaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ENST00000416931.1\"; gene_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ENST00000416931.1\"; gene_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>629083</td>\n",
       "      <td>634376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"FTMT20400000028.1\"; gene_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>exon</td>\n",
       "      <td>629083</td>\n",
       "      <td>634376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"FTMT20400000028.1\"; gene_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>629640</td>\n",
       "      <td>630683</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ENST00000457540.1\"; gene_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome      source     feature   start     end score strand frame  \\\n",
       "0       chr1     ENSEMBL  transcript  629062  629433     .      +     .   \n",
       "1       chr1     ENSEMBL        exon  629062  629433     .      +     .   \n",
       "2       chr1  FANTOM_cat  transcript  629083  634376     .      +     .   \n",
       "3       chr1  FANTOM_cat        exon  629083  634376     .      +     .   \n",
       "4       chr1     ENSEMBL  transcript  629640  630683     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  transcript_id \"ENST00000416931.1\"; gene_id \"EN...  \n",
       "1  transcript_id \"ENST00000416931.1\"; gene_id \"EN...  \n",
       "2  transcript_id \"FTMT20400000028.1\"; gene_id \"EN...  \n",
       "3  transcript_id \"FTMT20400000028.1\"; gene_id \"EN...  \n",
       "4  transcript_id \"ENST00000457540.1\"; gene_id \"EN...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load output from gffcompare\n",
    "hg38_gffcompare_annotation = pd.read_csv( '../data_raw/hg38.annotated.gtf',\n",
    "                           sep='\\t', header=None, names=gtf_column_names)\n",
    "hg38_gffcompare_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "198ad3ab-4e8d-44b5-8b75-74f5a0a1d856",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "376727\n",
      "376587\n"
     ]
    }
   ],
   "source": [
    "# filter the transcripts (remove exons)\n",
    "hg38_gffcompare_annotation = hg38_gffcompare_annotation[hg38_gffcompare_annotation['feature'] == 'transcript']\n",
    "print(len(hg38_gffcompare_annotation))\n",
    "\n",
    "# remove non-standard chromosomes\n",
    "standard_chromosomes = {f'chr{i}' for i in range(1, 23)}.union({'chrX', 'chrY'})\n",
    "hg38_gffcompare_annotation = hg38_gffcompare_annotation[hg38_gffcompare_annotation['chromosome'].isin(standard_chromosomes)]\n",
    "print(len(hg38_gffcompare_annotation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ccbc4dbf-336e-4642-8d31-6503aa095ef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_gffcompare_annotation.loc[:, 'transcript_id'] = hg38_gffcompare_annotation['attribute'].apply(split_attributes, args=(\"transcript_id\",))\n",
    "hg38_gffcompare_annotation.loc[:, 'gene_id'] = hg38_gffcompare_annotation['attribute'].apply(split_attributes, args=(\"gene_id\",))\n",
    "hg38_gffcompare_annotation.loc[:, 'class_code'] = hg38_gffcompare_annotation['attribute'].apply(split_attributes, args=(\"class_code\",))\n",
    "hg38_gffcompare_annotation.loc[:, 'cmp_ref'] = hg38_gffcompare_annotation['attribute'].apply(split_attributes, args=(\"cmp_ref\",))        # closest ref transcript\n",
    "hg38_gffcompare_annotation.loc[:, 'ref_gene_id'] = hg38_gffcompare_annotation['attribute'].apply(split_attributes, args=(\"ref_gene_id\",))\n",
    "hg38_gffcompare_annotation.loc[:, 'ref_gene_name'] = hg38_gffcompare_annotation['attribute'].apply(split_attributes, args=(\"ref_gene_name\",))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e9183563-fb41-4fee-a8b3-fe887ceab785",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>cmp_ref</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ENST00000416931.1\"; gene_id \"EN...</td>\n",
       "      <td>ENST00000416931.1</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>k</td>\n",
       "      <td>ENST00000416931.1</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>629083</td>\n",
       "      <td>634376</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"FTMT20400000028.1\"; gene_id \"EN...</td>\n",
       "      <td>FTMT20400000028.1</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>k</td>\n",
       "      <td>ENST00000414273.1</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>MTCO1P12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>629640</td>\n",
       "      <td>630683</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ENST00000457540.1\"; gene_id \"EN...</td>\n",
       "      <td>ENST00000457540.1</td>\n",
       "      <td>ENSG00000225630</td>\n",
       "      <td>k</td>\n",
       "      <td>ENST00000457540.1</td>\n",
       "      <td>ENSG00000225630</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>631074</td>\n",
       "      <td>632616</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ENST00000414273.1\"; gene_id \"EN...</td>\n",
       "      <td>ENST00000414273.1</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>k</td>\n",
       "      <td>ENST00000414273.1</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>MTCO1P12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>631074</td>\n",
       "      <td>707804</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"STRT00160641\"; gene_id \"ENSG000...</td>\n",
       "      <td>STRT00160641</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>k</td>\n",
       "      <td>ENST00000414273.1</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>MTCO1P12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome      source     feature   start     end score strand frame  \\\n",
       "0       chr1     ENSEMBL  transcript  629062  629433     .      +     .   \n",
       "2       chr1  FANTOM_cat  transcript  629083  634376     .      +     .   \n",
       "4       chr1     ENSEMBL  transcript  629640  630683     .      +     .   \n",
       "6       chr1     ENSEMBL  transcript  631074  632616     .      +     .   \n",
       "8       chr1   StringTie  transcript  631074  707804     .      +     .   \n",
       "\n",
       "                                           attribute      transcript_id  \\\n",
       "0  transcript_id \"ENST00000416931.1\"; gene_id \"EN...  ENST00000416931.1   \n",
       "2  transcript_id \"FTMT20400000028.1\"; gene_id \"EN...  FTMT20400000028.1   \n",
       "4  transcript_id \"ENST00000457540.1\"; gene_id \"EN...  ENST00000457540.1   \n",
       "6  transcript_id \"ENST00000414273.1\"; gene_id \"EN...  ENST00000414273.1   \n",
       "8  transcript_id \"STRT00160641\"; gene_id \"ENSG000...       STRT00160641   \n",
       "\n",
       "           gene_id class_code            cmp_ref      ref_gene_id  \\\n",
       "0  ENSG00000225972          k  ENST00000416931.1  ENSG00000225972   \n",
       "2  ENSG00000237973          k  ENST00000414273.1  ENSG00000237973   \n",
       "4  ENSG00000225630          k  ENST00000457540.1  ENSG00000225630   \n",
       "6  ENSG00000237973          k  ENST00000414273.1  ENSG00000237973   \n",
       "8  ENSG00000237973          k  ENST00000414273.1  ENSG00000237973   \n",
       "\n",
       "  ref_gene_name  \n",
       "0           NaN  \n",
       "2      MTCO1P12  \n",
       "4           NaN  \n",
       "6      MTCO1P12  \n",
       "8      MTCO1P12  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg38_gffcompare_annotation['ref_gene_id'] = hg38_gffcompare_annotation['ref_gene_id'].str.replace(r'\\..*', '', regex=True)\n",
    "hg38_gffcompare_annotation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62e12495-0f84-4de5-8632-b4818e9ef956",
   "metadata": {},
   "source": [
    "## Compare number of transcripts in gffcompare annotation vs original transcriptome GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b30f09c0-87dd-4bd6-aea0-fb2cdd44d3ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "376587\n"
     ]
    }
   ],
   "source": [
    "# Total number of transcripts in gffcompare annotation\n",
    "print(len(hg38_gffcompare_annotation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d0c11c78-afc2-40c8-9793-4957e3692ee5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376465"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of UNIQUE transcripts in gffcompare annotation\n",
    "len(hg38_gffcompare_annotation['transcript_id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7a27ff08-b37c-409a-9517-610fd3667b5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "122"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of duplicated transcripts\n",
    "len(hg38_gffcompare_annotation) - len(hg38_gffcompare_annotation['transcript_id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "db43102b-2a41-4965-8000-3ba27c12b939",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>cmp_ref</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336729</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144930269</td>\n",
       "      <td>144995200</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000010979.1</td>\n",
       "      <td>ENSG00000196369</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000619678.2</td>\n",
       "      <td>ENSG00000196369</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119343</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>143972661</td>\n",
       "      <td>143975153</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000010979.1</td>\n",
       "      <td>ENSG00000196369</td>\n",
       "      <td>=</td>\n",
       "      <td>ENST00000685922.2</td>\n",
       "      <td>ENSG00000289318</td>\n",
       "      <td>ENSG00000289318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336256</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120176307</td>\n",
       "      <td>120176520</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000011102.1</td>\n",
       "      <td>ENSG00000223380</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000578049.4</td>\n",
       "      <td>ENSG00000265808</td>\n",
       "      <td>SEC22B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340612</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148787530</td>\n",
       "      <td>148787807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000011102.1</td>\n",
       "      <td>ENSG00000223380</td>\n",
       "      <td>u</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119178</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120851843</td>\n",
       "      <td>120851857</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000011617.1</td>\n",
       "      <td>FTMG001538</td>\n",
       "      <td>p</td>\n",
       "      <td>ENST00000834689.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049817</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>67082692</td>\n",
       "      <td>67201219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02161951</td>\n",
       "      <td>ENSG00000154529</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000792558.1</td>\n",
       "      <td>ENSG00000303182</td>\n",
       "      <td>ENSG00000303182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049404</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65675928</td>\n",
       "      <td>65720092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02165508</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000382405.8</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>ZNG1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144616</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65318522</td>\n",
       "      <td>65323015</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02165508</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000445695.1</td>\n",
       "      <td>ENSG00000204778</td>\n",
       "      <td>ZNG1DP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049350</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65675928</td>\n",
       "      <td>65703014</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02167594</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000430059.6</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>ZNG1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144607</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65287981</td>\n",
       "      <td>65321223</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02167594</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000445695.1</td>\n",
       "      <td>ENSG00000204778</td>\n",
       "      <td>ZNG1DP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome      source     feature      start        end score strand  \\\n",
       "336729        chr1  FANTOM_cat  transcript  144930269  144995200     .      -   \n",
       "119343        chr1  FANTOM_cat  transcript  143972661  143975153     .      +   \n",
       "336256        chr1  FANTOM_cat  transcript  120176307  120176520     .      -   \n",
       "340612        chr1  FANTOM_cat  transcript  148787530  148787807     .      -   \n",
       "119178        chr1  FANTOM_cat  transcript  120851843  120851857     .      +   \n",
       "...            ...         ...         ...        ...        ...   ...    ...   \n",
       "4049817       chr9   StringTie  transcript   67082692   67201219     .      +   \n",
       "4049404       chr9   StringTie  transcript   65675928   65720092     .      +   \n",
       "4144616       chr9   StringTie  transcript   65318522   65323015     .      -   \n",
       "4049350       chr9   StringTie  transcript   65675928   65703014     .      +   \n",
       "4144607       chr9   StringTie  transcript   65287981   65321223     .      -   \n",
       "\n",
       "        frame      transcript_id          gene_id class_code  \\\n",
       "336729      .  ENCT00000010979.1  ENSG00000196369          j   \n",
       "119343      .  ENCT00000010979.1  ENSG00000196369          =   \n",
       "336256      .  ENCT00000011102.1  ENSG00000223380          c   \n",
       "340612      .  ENCT00000011102.1  ENSG00000223380          u   \n",
       "119178      .  ENCT00000011617.1       FTMG001538          p   \n",
       "...       ...                ...              ...        ...   \n",
       "4049817     .       STRT02161951  ENSG00000154529          j   \n",
       "4049404     .       STRT02165508  ENSG00000147996          c   \n",
       "4144616     .       STRT02165508  ENSG00000147996          c   \n",
       "4049350     .       STRT02167594  ENSG00000147996          c   \n",
       "4144607     .       STRT02167594  ENSG00000147996          c   \n",
       "\n",
       "                   cmp_ref      ref_gene_id    ref_gene_name  \n",
       "336729   ENST00000619678.2  ENSG00000196369              NaN  \n",
       "119343   ENST00000685922.2  ENSG00000289318  ENSG00000289318  \n",
       "336256   ENST00000578049.4  ENSG00000265808           SEC22B  \n",
       "340612                 NaN              NaN              NaN  \n",
       "119178   ENST00000834689.1              NaN              NaN  \n",
       "...                    ...              ...              ...  \n",
       "4049817  ENST00000792558.1  ENSG00000303182  ENSG00000303182  \n",
       "4049404  ENST00000382405.8  ENSG00000147996            ZNG1E  \n",
       "4144616  ENST00000445695.1  ENSG00000204778           ZNG1DP  \n",
       "4049350  ENST00000430059.6  ENSG00000147996            ZNG1E  \n",
       "4144607  ENST00000445695.1  ENSG00000204778           ZNG1DP  \n",
       "\n",
       "[244 rows x 14 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Which transcripts are duplicated in gffcompare annotation?\n",
    "dup_ids = hg38_gffcompare_annotation['transcript_id'][hg38_gffcompare_annotation['transcript_id'].duplicated()]\n",
    "duplicated_transcripts = hg38_gffcompare_annotation[hg38_gffcompare_annotation['transcript_id'].isin(dup_ids)].sort_values(by='transcript_id').drop(columns=[\"attribute\"])\n",
    "\n",
    "# Save to TSV\n",
    "duplicated_transcripts.to_csv('../data_processed/duplicated_transcripts.tsv', sep='\\t', index=False)\n",
    "duplicated_transcripts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "995eb15e-3909-4bef-8415-84bb7a39b14c",
   "metadata": {},
   "source": [
    "### Why are there duplicates?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "11e4e6d9-59ae-43a9-a391-8f64dd8b1ebc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>cmp_ref</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336729</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144930269</td>\n",
       "      <td>144995200</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000010979.1</td>\n",
       "      <td>ENSG00000196369</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000619678.2</td>\n",
       "      <td>ENSG00000196369</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119343</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>143972661</td>\n",
       "      <td>143975153</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000010979.1</td>\n",
       "      <td>ENSG00000196369</td>\n",
       "      <td>=</td>\n",
       "      <td>ENST00000685922.2</td>\n",
       "      <td>ENSG00000289318</td>\n",
       "      <td>ENSG00000289318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2235922</th>\n",
       "      <td>chr2</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91578477</td>\n",
       "      <td>91580863</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000226871.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000797278.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336725</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144809581</td>\n",
       "      <td>144809666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000226871.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783910.1</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336723</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144809577</td>\n",
       "      <td>144809726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000226875.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783910.1</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049817</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>67082692</td>\n",
       "      <td>67201219</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02161951</td>\n",
       "      <td>ENSG00000154529</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000792558.1</td>\n",
       "      <td>ENSG00000303182</td>\n",
       "      <td>ENSG00000303182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049404</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65675928</td>\n",
       "      <td>65720092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02165508</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000382405.8</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>ZNG1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144616</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65318522</td>\n",
       "      <td>65323015</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02165508</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000445695.1</td>\n",
       "      <td>ENSG00000204778</td>\n",
       "      <td>ZNG1DP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049350</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65675928</td>\n",
       "      <td>65703014</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02167594</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000430059.6</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>ZNG1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144607</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65287981</td>\n",
       "      <td>65321223</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02167594</td>\n",
       "      <td>ENSG00000147996</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000445695.1</td>\n",
       "      <td>ENSG00000204778</td>\n",
       "      <td>ZNG1DP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome      source     feature      start        end score strand  \\\n",
       "336729        chr1  FANTOM_cat  transcript  144930269  144995200     .      -   \n",
       "119343        chr1  FANTOM_cat  transcript  143972661  143975153     .      +   \n",
       "2235922       chr2  FANTOM_cat  transcript   91578477   91580863     .      +   \n",
       "336725        chr1  FANTOM_cat  transcript  144809581  144809666     .      -   \n",
       "336723        chr1  FANTOM_cat  transcript  144809577  144809726     .      -   \n",
       "...            ...         ...         ...        ...        ...   ...    ...   \n",
       "4049817       chr9   StringTie  transcript   67082692   67201219     .      +   \n",
       "4049404       chr9   StringTie  transcript   65675928   65720092     .      +   \n",
       "4144616       chr9   StringTie  transcript   65318522   65323015     .      -   \n",
       "4049350       chr9   StringTie  transcript   65675928   65703014     .      +   \n",
       "4144607       chr9   StringTie  transcript   65287981   65321223     .      -   \n",
       "\n",
       "        frame      transcript_id          gene_id class_code  \\\n",
       "336729      .  ENCT00000010979.1  ENSG00000196369          j   \n",
       "119343      .  ENCT00000010979.1  ENSG00000196369          =   \n",
       "2235922     .  ENCT00000226871.1  ENSG00000261600          c   \n",
       "336725      .  ENCT00000226871.1  ENSG00000261600          c   \n",
       "336723      .  ENCT00000226875.1  ENSG00000261600          c   \n",
       "...       ...                ...              ...        ...   \n",
       "4049817     .       STRT02161951  ENSG00000154529          j   \n",
       "4049404     .       STRT02165508  ENSG00000147996          c   \n",
       "4144616     .       STRT02165508  ENSG00000147996          c   \n",
       "4049350     .       STRT02167594  ENSG00000147996          c   \n",
       "4144607     .       STRT02167594  ENSG00000147996          c   \n",
       "\n",
       "                   cmp_ref      ref_gene_id    ref_gene_name  \n",
       "336729   ENST00000619678.2  ENSG00000196369              NaN  \n",
       "119343   ENST00000685922.2  ENSG00000289318  ENSG00000289318  \n",
       "2235922  ENST00000797278.1  ENSG00000261600  ENSG00000261600  \n",
       "336725   ENST00000783910.1  ENSG00000302082  ENSG00000302082  \n",
       "336723   ENST00000783910.1  ENSG00000302082  ENSG00000302082  \n",
       "...                    ...              ...              ...  \n",
       "4049817  ENST00000792558.1  ENSG00000303182  ENSG00000303182  \n",
       "4049404  ENST00000382405.8  ENSG00000147996            ZNG1E  \n",
       "4144616  ENST00000445695.1  ENSG00000204778           ZNG1DP  \n",
       "4049350  ENST00000430059.6  ENSG00000147996            ZNG1E  \n",
       "4144607  ENST00000445695.1  ENSG00000204778           ZNG1DP  \n",
       "\n",
       "[168 rows x 14 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show only instances of transcripts with + and -\n",
    "strand_conflicts = duplicated_transcripts.groupby(\"transcript_id\").filter(\n",
    "    lambda g: set(g[\"strand\"]) == {'+', '-'} and len(g) == 2)\n",
    "strand_conflicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c6331870-038e-4dde-8d1d-bd06852fa77a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>cmp_ref</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2944049</th>\n",
       "      <td>chr3</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48211057</td>\n",
       "      <td>48211110</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000213115.1</td>\n",
       "      <td>ENSG00000231205</td>\n",
       "      <td>u</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2126392</th>\n",
       "      <td>chr19</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>20408950</td>\n",
       "      <td>20424956</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000213115.1</td>\n",
       "      <td>ENSG00000231205</td>\n",
       "      <td>=</td>\n",
       "      <td>ENST00000731957.1</td>\n",
       "      <td>ENSG00000291130</td>\n",
       "      <td>ENSG00000291130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2235922</th>\n",
       "      <td>chr2</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91578477</td>\n",
       "      <td>91580863</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000226871.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000797278.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336725</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144809581</td>\n",
       "      <td>144809666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000226871.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783910.1</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336723</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144809577</td>\n",
       "      <td>144809726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000226875.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783910.1</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2235916</th>\n",
       "      <td>chr2</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91578477</td>\n",
       "      <td>91580863</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000226875.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000797278.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2944045</th>\n",
       "      <td>chr3</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48210879</td>\n",
       "      <td>48211110</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01039144</td>\n",
       "      <td>ENSG00000231205</td>\n",
       "      <td>u</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2126383</th>\n",
       "      <td>chr19</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>20408950</td>\n",
       "      <td>20424931</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01039144</td>\n",
       "      <td>ENSG00000231205</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000615684.4</td>\n",
       "      <td>ENSG00000291130</td>\n",
       "      <td>ENSG00000291130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2944047</th>\n",
       "      <td>chr3</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48210956</td>\n",
       "      <td>48211110</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01091226</td>\n",
       "      <td>ENSG00000231205</td>\n",
       "      <td>u</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2126376</th>\n",
       "      <td>chr19</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>20408950</td>\n",
       "      <td>20424927</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01091226</td>\n",
       "      <td>ENSG00000231205</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000615684.4</td>\n",
       "      <td>ENSG00000291130</td>\n",
       "      <td>ENSG00000291130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336727</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144809581</td>\n",
       "      <td>144809726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01132284</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783910.1</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2235914</th>\n",
       "      <td>chr2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91578477</td>\n",
       "      <td>91578585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01132284</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>k</td>\n",
       "      <td>ENST00000454518.1</td>\n",
       "      <td>ENSG00000230964</td>\n",
       "      <td>ENSG00000230964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336719</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144809489</td>\n",
       "      <td>144809726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01151362</td>\n",
       "      <td>ENSG00000233991</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783910.1</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2235910</th>\n",
       "      <td>chr2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91571610</td>\n",
       "      <td>91571683</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01151362</td>\n",
       "      <td>ENSG00000233991</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000797279.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336721</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144809490</td>\n",
       "      <td>144809726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01193804</td>\n",
       "      <td>ENSG00000233991</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783910.1</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "      <td>ENSG00000302082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2235912</th>\n",
       "      <td>chr2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91571648</td>\n",
       "      <td>91571683</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT01193804</td>\n",
       "      <td>ENSG00000233991</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000797279.1</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "      <td>ENSG00000261600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome      source     feature      start        end score strand  \\\n",
       "2944049       chr3  FANTOM_cat  transcript   48211057   48211110     .      -   \n",
       "2126392      chr19  FANTOM_cat  transcript   20408950   20424956     .      -   \n",
       "2235922       chr2  FANTOM_cat  transcript   91578477   91580863     .      +   \n",
       "336725        chr1  FANTOM_cat  transcript  144809581  144809666     .      -   \n",
       "336723        chr1  FANTOM_cat  transcript  144809577  144809726     .      -   \n",
       "2235916       chr2  FANTOM_cat  transcript   91578477   91580863     .      +   \n",
       "2944045       chr3   StringTie  transcript   48210879   48211110     .      -   \n",
       "2126383      chr19   StringTie  transcript   20408950   20424931     .      -   \n",
       "2944047       chr3   StringTie  transcript   48210956   48211110     .      -   \n",
       "2126376      chr19   StringTie  transcript   20408950   20424927     .      -   \n",
       "336727        chr1   StringTie  transcript  144809581  144809726     .      -   \n",
       "2235914       chr2   StringTie  transcript   91578477   91578585     .      +   \n",
       "336719        chr1   StringTie  transcript  144809489  144809726     .      -   \n",
       "2235910       chr2   StringTie  transcript   91571610   91571683     .      +   \n",
       "336721        chr1   StringTie  transcript  144809490  144809726     .      -   \n",
       "2235912       chr2   StringTie  transcript   91571648   91571683     .      +   \n",
       "\n",
       "        frame      transcript_id          gene_id class_code  \\\n",
       "2944049     .  ENCT00000213115.1  ENSG00000231205          u   \n",
       "2126392     .  ENCT00000213115.1  ENSG00000231205          =   \n",
       "2235922     .  ENCT00000226871.1  ENSG00000261600          c   \n",
       "336725      .  ENCT00000226871.1  ENSG00000261600          c   \n",
       "336723      .  ENCT00000226875.1  ENSG00000261600          c   \n",
       "2235916     .  ENCT00000226875.1  ENSG00000261600          c   \n",
       "2944045     .       STRT01039144  ENSG00000231205          u   \n",
       "2126383     .       STRT01039144  ENSG00000231205          c   \n",
       "2944047     .       STRT01091226  ENSG00000231205          u   \n",
       "2126376     .       STRT01091226  ENSG00000231205          c   \n",
       "336727      .       STRT01132284  ENSG00000261600          c   \n",
       "2235914     .       STRT01132284  ENSG00000261600          k   \n",
       "336719      .       STRT01151362  ENSG00000233991          c   \n",
       "2235910     .       STRT01151362  ENSG00000233991          c   \n",
       "336721      .       STRT01193804  ENSG00000233991          c   \n",
       "2235912     .       STRT01193804  ENSG00000233991          c   \n",
       "\n",
       "                   cmp_ref      ref_gene_id    ref_gene_name  \n",
       "2944049                NaN              NaN              NaN  \n",
       "2126392  ENST00000731957.1  ENSG00000291130  ENSG00000291130  \n",
       "2235922  ENST00000797278.1  ENSG00000261600  ENSG00000261600  \n",
       "336725   ENST00000783910.1  ENSG00000302082  ENSG00000302082  \n",
       "336723   ENST00000783910.1  ENSG00000302082  ENSG00000302082  \n",
       "2235916  ENST00000797278.1  ENSG00000261600  ENSG00000261600  \n",
       "2944045                NaN              NaN              NaN  \n",
       "2126383  ENST00000615684.4  ENSG00000291130  ENSG00000291130  \n",
       "2944047                NaN              NaN              NaN  \n",
       "2126376  ENST00000615684.4  ENSG00000291130  ENSG00000291130  \n",
       "336727   ENST00000783910.1  ENSG00000302082  ENSG00000302082  \n",
       "2235914  ENST00000454518.1  ENSG00000230964  ENSG00000230964  \n",
       "336719   ENST00000783910.1  ENSG00000302082  ENSG00000302082  \n",
       "2235910  ENST00000797279.1  ENSG00000261600  ENSG00000261600  \n",
       "336721   ENST00000783910.1  ENSG00000302082  ENSG00000302082  \n",
       "2235912  ENST00000797279.1  ENSG00000261600  ENSG00000261600  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show only instances of transcripts with different chromosomes\n",
    "chrom_conflicts = duplicated_transcripts.groupby(\"transcript_id\").filter(\n",
    "    lambda g: len(g[\"chromosome\"].unique()) > 1 and len(g) == 2)\n",
    "chrom_conflicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3bddab2e-c89c-4e6c-8059-c4ab9d3900b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>cmp_ref</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336256</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120176307</td>\n",
       "      <td>120176520</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000011102.1</td>\n",
       "      <td>ENSG00000223380</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000578049.4</td>\n",
       "      <td>ENSG00000265808</td>\n",
       "      <td>SEC22B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340612</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148787530</td>\n",
       "      <td>148787807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000011102.1</td>\n",
       "      <td>ENSG00000223380</td>\n",
       "      <td>u</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119178</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120851843</td>\n",
       "      <td>120851857</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000011617.1</td>\n",
       "      <td>FTMG001538</td>\n",
       "      <td>p</td>\n",
       "      <td>ENST00000834689.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119311</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>143729963</td>\n",
       "      <td>143730438</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000011617.1</td>\n",
       "      <td>FTMG001538</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000783555.1</td>\n",
       "      <td>ENSG00000232721</td>\n",
       "      <td>ENSG00000232721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336278</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM_cat</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120850897</td>\n",
       "      <td>120850985</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCT00000031689.1</td>\n",
       "      <td>ENSG00000212544</td>\n",
       "      <td>c</td>\n",
       "      <td>ENST00000664186.1</td>\n",
       "      <td>ENSG00000287979</td>\n",
       "      <td>ENSG00000287979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340542</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148769633</td>\n",
       "      <td>148784591</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT00204587</td>\n",
       "      <td>ENSG00000223380</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000764483.1</td>\n",
       "      <td>ENSG00000299543</td>\n",
       "      <td>ENSG00000299543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4143669</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>41131678</td>\n",
       "      <td>41189273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02160141</td>\n",
       "      <td>ENSG00000204790</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000456520.5</td>\n",
       "      <td>ENSG00000215126</td>\n",
       "      <td>ZNG1F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144572</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>63974762</td>\n",
       "      <td>64002564</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02160141</td>\n",
       "      <td>ENSG00000204790</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000485382.3</td>\n",
       "      <td>ENSG00000204790</td>\n",
       "      <td>ENSG00000204790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4143686</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>41131678</td>\n",
       "      <td>41189307</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02161837</td>\n",
       "      <td>ENSG00000204790</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000456520.5</td>\n",
       "      <td>ENSG00000215126</td>\n",
       "      <td>ZNG1F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144564</th>\n",
       "      <td>chr9</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>63974762</td>\n",
       "      <td>64002564</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>STRT02161837</td>\n",
       "      <td>ENSG00000204790</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000738329.1</td>\n",
       "      <td>ENSG00000296339</td>\n",
       "      <td>ENSG00000296339</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome      source     feature      start        end score strand  \\\n",
       "336256        chr1  FANTOM_cat  transcript  120176307  120176520     .      -   \n",
       "340612        chr1  FANTOM_cat  transcript  148787530  148787807     .      -   \n",
       "119178        chr1  FANTOM_cat  transcript  120851843  120851857     .      +   \n",
       "119311        chr1  FANTOM_cat  transcript  143729963  143730438     .      +   \n",
       "336278        chr1  FANTOM_cat  transcript  120850897  120850985     .      -   \n",
       "...            ...         ...         ...        ...        ...   ...    ...   \n",
       "340542        chr1   StringTie  transcript  148769633  148784591     .      -   \n",
       "4143669       chr9   StringTie  transcript   41131678   41189273     .      -   \n",
       "4144572       chr9   StringTie  transcript   63974762   64002564     .      -   \n",
       "4143686       chr9   StringTie  transcript   41131678   41189307     .      -   \n",
       "4144564       chr9   StringTie  transcript   63974762   64002564     .      -   \n",
       "\n",
       "        frame      transcript_id          gene_id class_code  \\\n",
       "336256      .  ENCT00000011102.1  ENSG00000223380          c   \n",
       "340612      .  ENCT00000011102.1  ENSG00000223380          u   \n",
       "119178      .  ENCT00000011617.1       FTMG001538          p   \n",
       "119311      .  ENCT00000011617.1       FTMG001538          c   \n",
       "336278      .  ENCT00000031689.1  ENSG00000212544          c   \n",
       "...       ...                ...              ...        ...   \n",
       "340542      .       STRT00204587  ENSG00000223380          j   \n",
       "4143669     .       STRT02160141  ENSG00000204790          j   \n",
       "4144572     .       STRT02160141  ENSG00000204790          j   \n",
       "4143686     .       STRT02161837  ENSG00000204790          j   \n",
       "4144564     .       STRT02161837  ENSG00000204790          j   \n",
       "\n",
       "                   cmp_ref      ref_gene_id    ref_gene_name  \n",
       "336256   ENST00000578049.4  ENSG00000265808           SEC22B  \n",
       "340612                 NaN              NaN              NaN  \n",
       "119178   ENST00000834689.1              NaN              NaN  \n",
       "119311   ENST00000783555.1  ENSG00000232721  ENSG00000232721  \n",
       "336278   ENST00000664186.1  ENSG00000287979  ENSG00000287979  \n",
       "...                    ...              ...              ...  \n",
       "340542   ENST00000764483.1  ENSG00000299543  ENSG00000299543  \n",
       "4143669  ENST00000456520.5  ENSG00000215126            ZNG1F  \n",
       "4144572  ENST00000485382.3  ENSG00000204790  ENSG00000204790  \n",
       "4143686  ENST00000456520.5  ENSG00000215126            ZNG1F  \n",
       "4144564  ENST00000738329.1  ENSG00000296339  ENSG00000296339  \n",
       "\n",
       "[70 rows x 14 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show all remaining instances\n",
    "conflict_ids = set(chrom_conflicts[\"transcript_id\"]) | set(strand_conflicts[\"transcript_id\"])\n",
    "other_conflicts = duplicated_transcripts[~duplicated_transcripts[\"transcript_id\"].isin(conflict_ids)]\n",
    "#other_conflicts.to_csv('other_conflicts.tsv', sep='\\t', index=False)\n",
    "other_conflicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "650b7f39-964f-499e-b894-ceae52bdd67f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22870886\n",
      "28656077\n"
     ]
    }
   ],
   "source": [
    "distances = []\n",
    "for tid, group in other_conflicts.groupby(\"transcript_id\"):\n",
    "    min_start = group[\"start\"].min()\n",
    "    max_end = group[\"end\"].max()\n",
    "    dist = int(max_end) - int(min_start)\n",
    "    distances.append(dist)\n",
    "\n",
    "print(min(distances))\n",
    "print(max(distances))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4411db95-4d50-45ce-97c2-a7be91242ae8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcript IDs in both strand and chromosome conflicts:\n",
      "ENCT00000226871.1\n",
      "ENCT00000226875.1\n",
      "STRT01132284\n",
      "STRT01151362\n",
      "STRT01193804\n"
     ]
    }
   ],
   "source": [
    "# Which transcripts have both different strands and different chromosomes\n",
    "conflict_overlap = set(strand_conflicts[\"transcript_id\"]) & set(chrom_conflicts[\"transcript_id\"])\n",
    "\n",
    "print(\"Transcript IDs in both strand and chromosome conflicts:\")\n",
    "for tid in sorted(conflict_overlap):\n",
    "    print(tid)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (annotation_project)",
   "language": "python",
   "name": "annotation_project"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
