{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bacfbc9a-75d6-47d4-bf17-5a124e7fb8e2",
   "metadata": {},
   "source": [
    "# Add liftover of TSS coordinates to Anthony's df of TSS info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b2d1019-c9a6-4d7a-b216-33830aa4c899",
   "metadata": {},
   "source": [
    "## Load Anthony's TSS data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c03991a5-2096-4727-9c1f-866eaaa69871",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    ".libPaths(\"/no_backup/jferrer/jmidgley/R_libs\")\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(stringr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "03f3ff88-fac8-4d01-bb8e-52d5fce523cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 28</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chromosome</th><th scope=col>tx_start</th><th scope=col>tx_end</th><th scope=col>strand</th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>gene_name</th><th scope=col>Gene class</th><th scope=col>gene_type</th><th scope=col>transcript_type</th><th scope=col>⋯</th><th scope=col>high_tpm</th><th scope=col>low_tpm</th><th scope=col>peak</th><th scope=col>TSS expression high glucose (Salmon, TPM)</th><th scope=col>Relative TSS usage</th><th scope=col>TSS_type</th><th scope=col>distance_to_closest_TSS</th><th scope=col>Distance to closest Gencode TSS</th><th scope=col>name</th><th scope=col>new_1st_exon</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>14362</td><td>29370</td><td>-</td><td>ENST00000423562.1</td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.58083</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td></tr>\n",
       "\t<tr><td>chr1</td><td>14359</td><td>29350</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037199      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.58083</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td></tr>\n",
       "\t<tr><td>chr1</td><td>14359</td><td>29350</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037200      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.58083</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td></tr>\n",
       "\t<tr><td>chr1</td><td>14360</td><td>29364</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037201      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.58083</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td></tr>\n",
       "\t<tr><td>chr1</td><td>14362</td><td>29346</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037202      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.58083</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td></tr>\n",
       "\t<tr><td>chr1</td><td>14365</td><td>29347</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037205      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.58083</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 28\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " chromosome & tx\\_start & tx\\_end & strand & transcript\\_id & gene\\_id & gene\\_name & Gene class & gene\\_type & transcript\\_type & ⋯ & high\\_tpm & low\\_tpm & peak & TSS expression high glucose (Salmon, TPM) & Relative TSS usage & TSS\\_type & distance\\_to\\_closest\\_TSS & Distance to closest Gencode TSS & name & new\\_1st\\_exon\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <chr> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t chr1 & 14362 & 29370 & - & ENST00000423562.1 & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.58083 & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes\\\\\n",
       "\t chr1 & 14359 & 29350 & - & PBT00037199       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.58083 & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes\\\\\n",
       "\t chr1 & 14359 & 29350 & - & PBT00037200       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.58083 & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes\\\\\n",
       "\t chr1 & 14360 & 29364 & - & PBT00037201       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.58083 & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes\\\\\n",
       "\t chr1 & 14362 & 29346 & - & PBT00037202       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.58083 & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes\\\\\n",
       "\t chr1 & 14365 & 29347 & - & PBT00037205       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.58083 & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 28\n",
       "\n",
       "| chromosome &lt;chr&gt; | tx_start &lt;dbl&gt; | tx_end &lt;dbl&gt; | strand &lt;chr&gt; | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | gene_name &lt;chr&gt; | Gene class &lt;chr&gt; | gene_type &lt;chr&gt; | transcript_type &lt;chr&gt; | ⋯ ⋯ | high_tpm &lt;dbl&gt; | low_tpm &lt;dbl&gt; | peak &lt;chr&gt; | TSS expression high glucose (Salmon, TPM) &lt;dbl&gt; | Relative TSS usage &lt;dbl&gt; | TSS_type &lt;chr&gt; | distance_to_closest_TSS &lt;dbl&gt; | Distance to closest Gencode TSS &lt;chr&gt; | name &lt;chr&gt; | new_1st_exon &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 14362 | 29370 | - | ENST00000423562.1 | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.58083 | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes |\n",
       "| chr1 | 14359 | 29350 | - | PBT00037199       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.58083 | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes |\n",
       "| chr1 | 14359 | 29350 | - | PBT00037200       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.58083 | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes |\n",
       "| chr1 | 14360 | 29364 | - | PBT00037201       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.58083 | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes |\n",
       "| chr1 | 14362 | 29346 | - | PBT00037202       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.58083 | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes |\n",
       "| chr1 | 14365 | 29347 | - | PBT00037205       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.58083 | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes |\n",
       "\n"
      ],
      "text/plain": [
       "  chromosome tx_start tx_end strand transcript_id     gene_id         gene_name\n",
       "1 chr1       14362    29370  -      ENST00000423562.1 ENSG00000227232 WASH7P   \n",
       "2 chr1       14359    29350  -      PBT00037199       ENSG00000227232 WASH7P   \n",
       "3 chr1       14359    29350  -      PBT00037200       ENSG00000227232 WASH7P   \n",
       "4 chr1       14360    29364  -      PBT00037201       ENSG00000227232 WASH7P   \n",
       "5 chr1       14362    29346  -      PBT00037202       ENSG00000227232 WASH7P   \n",
       "6 chr1       14365    29347  -      PBT00037205       ENSG00000227232 WASH7P   \n",
       "  Gene class gene_type  transcript_type ⋯ high_tpm low_tpm  peak  \n",
       "1 misc RNA   pseudogene new_transcript  ⋯ 17.58083 17.85643 Robust\n",
       "2 misc RNA   pseudogene new_transcript  ⋯ 17.58083 17.85643 Robust\n",
       "3 misc RNA   pseudogene new_transcript  ⋯ 17.58083 17.85643 Robust\n",
       "4 misc RNA   pseudogene new_transcript  ⋯ 17.58083 17.85643 Robust\n",
       "5 misc RNA   pseudogene new_transcript  ⋯ 17.58083 17.85643 Robust\n",
       "6 misc RNA   pseudogene new_transcript  ⋯ 17.58083 17.85643 Robust\n",
       "  TSS expression high glucose (Salmon, TPM) Relative TSS usage TSS_type  \n",
       "1 27.34853                                  1                  unique TSS\n",
       "2 27.34853                                  1                  unique TSS\n",
       "3 27.34853                                  1                  unique TSS\n",
       "4 27.34853                                  1                  unique TSS\n",
       "5 27.34853                                  1                  unique TSS\n",
       "6 27.34853                                  1                  unique TSS\n",
       "  distance_to_closest_TSS Distance to closest Gencode TSS\n",
       "1 211                     >100bp                         \n",
       "2 211                     >100bp                         \n",
       "3 211                     >100bp                         \n",
       "4 211                     >100bp                         \n",
       "5 211                     >100bp                         \n",
       "6 211                     >100bp                         \n",
       "  name                    new_1st_exon\n",
       "1 chr1:29336-29359_Peak_1 yes         \n",
       "2 chr1:29336-29359_Peak_1 yes         \n",
       "3 chr1:29336-29359_Peak_1 yes         \n",
       "4 chr1:29336-29359_Peak_1 yes         \n",
       "5 chr1:29336-29359_Peak_1 yes         \n",
       "6 chr1:29336-29359_Peak_1 yes         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load table with TSS info - NOTE: uses hg19 positions\n",
    "TSS_df <- read_tsv('/users/jferrer/public-docs/tracksDirectory/mplanas/HI_transcriptome_v2.2.tsv.gz', show_col_types = FALSE)\n",
    "head(TSS_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b77ebcf-a032-4be8-8b5e-24d2f4b9f00f",
   "metadata": {},
   "source": [
    "### Load LiftOver TSS coordinates and add to TSS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bc15e6d-e99a-457e-b177-ed9a1115f357",
   "metadata": {},
   "source": [
    "Used UCSC Liftover tool to create BED file of new start and end positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b82df6d-ded1-45c6-a9af-e2e012b65058",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>new_start</th><th scope=col>new_end</th><th scope=col>old_coords</th><th scope=col>score</th><th scope=col>strand</th><th scope=col>old_base</th><th scope=col>new_coords</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 29335</td><td> 29359</td><td>chr1:29336-29359:.  </td><td>1</td><td>-</td><td>chr1:29336-29359  </td><td>chr1:29336-29359  </td></tr>\n",
       "\t<tr><td>chr1</td><td>629079</td><td>629136</td><td>chr1:564460-564516:.</td><td>1</td><td>+</td><td>chr1:564460-564516</td><td>chr1:629080-629136</td></tr>\n",
       "\t<tr><td>chr1</td><td>629638</td><td>629739</td><td>chr1:565019-565119:.</td><td>1</td><td>+</td><td>chr1:565019-565119</td><td>chr1:629639-629739</td></tr>\n",
       "\t<tr><td>chr1</td><td>631069</td><td>631078</td><td>chr1:566450-566458:.</td><td>1</td><td>+</td><td>chr1:566450-566458</td><td>chr1:631070-631078</td></tr>\n",
       "\t<tr><td>chr1</td><td>632535</td><td>632616</td><td>chr1:567916-567996:.</td><td>1</td><td>-</td><td>chr1:567916-567996</td><td>chr1:632536-632616</td></tr>\n",
       "\t<tr><td>chr1</td><td>632597</td><td>632876</td><td>chr1:567978-568256:.</td><td>1</td><td>+</td><td>chr1:567978-568256</td><td>chr1:632598-632876</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " chr & new\\_start & new\\_end & old\\_coords & score & strand & old\\_base & new\\_coords\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t chr1 &  29335 &  29359 & chr1:29336-29359:.   & 1 & - & chr1:29336-29359   & chr1:29336-29359  \\\\\n",
       "\t chr1 & 629079 & 629136 & chr1:564460-564516:. & 1 & + & chr1:564460-564516 & chr1:629080-629136\\\\\n",
       "\t chr1 & 629638 & 629739 & chr1:565019-565119:. & 1 & + & chr1:565019-565119 & chr1:629639-629739\\\\\n",
       "\t chr1 & 631069 & 631078 & chr1:566450-566458:. & 1 & + & chr1:566450-566458 & chr1:631070-631078\\\\\n",
       "\t chr1 & 632535 & 632616 & chr1:567916-567996:. & 1 & - & chr1:567916-567996 & chr1:632536-632616\\\\\n",
       "\t chr1 & 632597 & 632876 & chr1:567978-568256:. & 1 & + & chr1:567978-568256 & chr1:632598-632876\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 8\n",
       "\n",
       "| chr &lt;chr&gt; | new_start &lt;dbl&gt; | new_end &lt;dbl&gt; | old_coords &lt;chr&gt; | score &lt;dbl&gt; | strand &lt;chr&gt; | old_base &lt;chr&gt; | new_coords &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  29335 |  29359 | chr1:29336-29359:.   | 1 | - | chr1:29336-29359   | chr1:29336-29359   |\n",
       "| chr1 | 629079 | 629136 | chr1:564460-564516:. | 1 | + | chr1:564460-564516 | chr1:629080-629136 |\n",
       "| chr1 | 629638 | 629739 | chr1:565019-565119:. | 1 | + | chr1:565019-565119 | chr1:629639-629739 |\n",
       "| chr1 | 631069 | 631078 | chr1:566450-566458:. | 1 | + | chr1:566450-566458 | chr1:631070-631078 |\n",
       "| chr1 | 632535 | 632616 | chr1:567916-567996:. | 1 | - | chr1:567916-567996 | chr1:632536-632616 |\n",
       "| chr1 | 632597 | 632876 | chr1:567978-568256:. | 1 | + | chr1:567978-568256 | chr1:632598-632876 |\n",
       "\n"
      ],
      "text/plain": [
       "  chr  new_start new_end old_coords           score strand old_base          \n",
       "1 chr1  29335     29359  chr1:29336-29359:.   1     -      chr1:29336-29359  \n",
       "2 chr1 629079    629136  chr1:564460-564516:. 1     +      chr1:564460-564516\n",
       "3 chr1 629638    629739  chr1:565019-565119:. 1     +      chr1:565019-565119\n",
       "4 chr1 631069    631078  chr1:566450-566458:. 1     +      chr1:566450-566458\n",
       "5 chr1 632535    632616  chr1:567916-567996:. 1     -      chr1:567916-567996\n",
       "6 chr1 632597    632876  chr1:567978-568256:. 1     +      chr1:567978-568256\n",
       "  new_coords        \n",
       "1 chr1:29336-29359  \n",
       "2 chr1:629080-629136\n",
       "3 chr1:629639-629739\n",
       "4 chr1:631070-631078\n",
       "5 chr1:632536-632616\n",
       "6 chr1:632598-632876"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "liftover_tss <- read_tsv(\"./HumanIsletTranscriptome/data_raw/tss_liftover.bed\", col_names = FALSE, show_col_types = FALSE) %>%\n",
    "  setNames(c(\"chr\", \"new_start\", \"new_end\", \"old_coords\", \"score\", \"strand\")) %>%\n",
    "  mutate(\n",
    "    old_base = str_extract(old_coords, \"^chr[\\\\w]+:[0-9]+-[0-9]+\"),\n",
    "    new_coords = paste0(chr, \":\", new_start + 1, \"-\", new_end)\n",
    "  ) %>%\n",
    "  distinct(old_base, .keep_all = TRUE)\n",
    "head(liftover_tss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ea606a2c-4d6b-41c9-a776-31d3a14ca836",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 29</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chromosome</th><th scope=col>tx_start</th><th scope=col>tx_end</th><th scope=col>strand</th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>gene_name</th><th scope=col>Gene class</th><th scope=col>gene_type</th><th scope=col>transcript_type</th><th scope=col>⋯</th><th scope=col>low_tpm</th><th scope=col>peak</th><th scope=col>TSS expression high glucose (Salmon, TPM)</th><th scope=col>Relative TSS usage</th><th scope=col>TSS_type</th><th scope=col>distance_to_closest_TSS</th><th scope=col>Distance to closest Gencode TSS</th><th scope=col>name</th><th scope=col>new_1st_exon</th><th scope=col>new_name</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>14362</td><td>29370</td><td>-</td><td>ENST00000423562.1</td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td><td>chr1:29336-29359_Peak_1</td></tr>\n",
       "\t<tr><td>chr1</td><td>14359</td><td>29350</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037199      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td><td>chr1:29336-29359_Peak_1</td></tr>\n",
       "\t<tr><td>chr1</td><td>14359</td><td>29350</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037200      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td><td>chr1:29336-29359_Peak_1</td></tr>\n",
       "\t<tr><td>chr1</td><td>14360</td><td>29364</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037201      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td><td>chr1:29336-29359_Peak_1</td></tr>\n",
       "\t<tr><td>chr1</td><td>14362</td><td>29346</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037202      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td><td>chr1:29336-29359_Peak_1</td></tr>\n",
       "\t<tr><td>chr1</td><td>14365</td><td>29347</td><td>-</td><td><span style=white-space:pre-wrap>PBT00037205      </span></td><td>ENSG00000227232</td><td>WASH7P</td><td>misc RNA</td><td>pseudogene</td><td>new_transcript</td><td>⋯</td><td>17.85643</td><td>Robust</td><td>27.34853</td><td>1</td><td>unique TSS</td><td>211</td><td>&gt;100bp</td><td>chr1:29336-29359_Peak_1</td><td>yes</td><td>chr1:29336-29359_Peak_1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 29\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " chromosome & tx\\_start & tx\\_end & strand & transcript\\_id & gene\\_id & gene\\_name & Gene class & gene\\_type & transcript\\_type & ⋯ & low\\_tpm & peak & TSS expression high glucose (Salmon, TPM) & Relative TSS usage & TSS\\_type & distance\\_to\\_closest\\_TSS & Distance to closest Gencode TSS & name & new\\_1st\\_exon & new\\_name\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <chr> & <dbl> & <dbl> & <chr> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t chr1 & 14362 & 29370 & - & ENST00000423562.1 & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes & chr1:29336-29359\\_Peak\\_1\\\\\n",
       "\t chr1 & 14359 & 29350 & - & PBT00037199       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes & chr1:29336-29359\\_Peak\\_1\\\\\n",
       "\t chr1 & 14359 & 29350 & - & PBT00037200       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes & chr1:29336-29359\\_Peak\\_1\\\\\n",
       "\t chr1 & 14360 & 29364 & - & PBT00037201       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes & chr1:29336-29359\\_Peak\\_1\\\\\n",
       "\t chr1 & 14362 & 29346 & - & PBT00037202       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes & chr1:29336-29359\\_Peak\\_1\\\\\n",
       "\t chr1 & 14365 & 29347 & - & PBT00037205       & ENSG00000227232 & WASH7P & misc RNA & pseudogene & new\\_transcript & ⋯ & 17.85643 & Robust & 27.34853 & 1 & unique TSS & 211 & >100bp & chr1:29336-29359\\_Peak\\_1 & yes & chr1:29336-29359\\_Peak\\_1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 29\n",
       "\n",
       "| chromosome &lt;chr&gt; | tx_start &lt;dbl&gt; | tx_end &lt;dbl&gt; | strand &lt;chr&gt; | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | gene_name &lt;chr&gt; | Gene class &lt;chr&gt; | gene_type &lt;chr&gt; | transcript_type &lt;chr&gt; | ⋯ ⋯ | low_tpm &lt;dbl&gt; | peak &lt;chr&gt; | TSS expression high glucose (Salmon, TPM) &lt;dbl&gt; | Relative TSS usage &lt;dbl&gt; | TSS_type &lt;chr&gt; | distance_to_closest_TSS &lt;dbl&gt; | Distance to closest Gencode TSS &lt;chr&gt; | name &lt;chr&gt; | new_1st_exon &lt;chr&gt; | new_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 14362 | 29370 | - | ENST00000423562.1 | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes | chr1:29336-29359_Peak_1 |\n",
       "| chr1 | 14359 | 29350 | - | PBT00037199       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes | chr1:29336-29359_Peak_1 |\n",
       "| chr1 | 14359 | 29350 | - | PBT00037200       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes | chr1:29336-29359_Peak_1 |\n",
       "| chr1 | 14360 | 29364 | - | PBT00037201       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes | chr1:29336-29359_Peak_1 |\n",
       "| chr1 | 14362 | 29346 | - | PBT00037202       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes | chr1:29336-29359_Peak_1 |\n",
       "| chr1 | 14365 | 29347 | - | PBT00037205       | ENSG00000227232 | WASH7P | misc RNA | pseudogene | new_transcript | ⋯ | 17.85643 | Robust | 27.34853 | 1 | unique TSS | 211 | &gt;100bp | chr1:29336-29359_Peak_1 | yes | chr1:29336-29359_Peak_1 |\n",
       "\n"
      ],
      "text/plain": [
       "  chromosome tx_start tx_end strand transcript_id     gene_id         gene_name\n",
       "1 chr1       14362    29370  -      ENST00000423562.1 ENSG00000227232 WASH7P   \n",
       "2 chr1       14359    29350  -      PBT00037199       ENSG00000227232 WASH7P   \n",
       "3 chr1       14359    29350  -      PBT00037200       ENSG00000227232 WASH7P   \n",
       "4 chr1       14360    29364  -      PBT00037201       ENSG00000227232 WASH7P   \n",
       "5 chr1       14362    29346  -      PBT00037202       ENSG00000227232 WASH7P   \n",
       "6 chr1       14365    29347  -      PBT00037205       ENSG00000227232 WASH7P   \n",
       "  Gene class gene_type  transcript_type ⋯ low_tpm  peak  \n",
       "1 misc RNA   pseudogene new_transcript  ⋯ 17.85643 Robust\n",
       "2 misc RNA   pseudogene new_transcript  ⋯ 17.85643 Robust\n",
       "3 misc RNA   pseudogene new_transcript  ⋯ 17.85643 Robust\n",
       "4 misc RNA   pseudogene new_transcript  ⋯ 17.85643 Robust\n",
       "5 misc RNA   pseudogene new_transcript  ⋯ 17.85643 Robust\n",
       "6 misc RNA   pseudogene new_transcript  ⋯ 17.85643 Robust\n",
       "  TSS expression high glucose (Salmon, TPM) Relative TSS usage TSS_type  \n",
       "1 27.34853                                  1                  unique TSS\n",
       "2 27.34853                                  1                  unique TSS\n",
       "3 27.34853                                  1                  unique TSS\n",
       "4 27.34853                                  1                  unique TSS\n",
       "5 27.34853                                  1                  unique TSS\n",
       "6 27.34853                                  1                  unique TSS\n",
       "  distance_to_closest_TSS Distance to closest Gencode TSS\n",
       "1 211                     >100bp                         \n",
       "2 211                     >100bp                         \n",
       "3 211                     >100bp                         \n",
       "4 211                     >100bp                         \n",
       "5 211                     >100bp                         \n",
       "6 211                     >100bp                         \n",
       "  name                    new_1st_exon new_name               \n",
       "1 chr1:29336-29359_Peak_1 yes          chr1:29336-29359_Peak_1\n",
       "2 chr1:29336-29359_Peak_1 yes          chr1:29336-29359_Peak_1\n",
       "3 chr1:29336-29359_Peak_1 yes          chr1:29336-29359_Peak_1\n",
       "4 chr1:29336-29359_Peak_1 yes          chr1:29336-29359_Peak_1\n",
       "5 chr1:29336-29359_Peak_1 yes          chr1:29336-29359_Peak_1\n",
       "6 chr1:29336-29359_Peak_1 yes          chr1:29336-29359_Peak_1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Add hg38 coordinates to original df\n",
    "TSS_df <- TSS_df %>%\n",
    "  mutate(\n",
    "    old_base = str_extract(name, \"^chr[\\\\w]+:[0-9]+-[0-9]+\"),\n",
    "    peak_suffix = str_extract(name, \"Peak_\\\\d+$\")\n",
    "  ) %>%\n",
    "  left_join(liftover_tss %>% dplyr::select(old_base, new_coords), by = \"old_base\") %>%\n",
    "  mutate(\n",
    "    new_name = if_else(\n",
    "      is.na(new_coords),\n",
    "      paste0(\"conversion_failed_\", peak_suffix),\n",
    "      paste0(new_coords, \"_\", peak_suffix)\n",
    "    )\n",
    "  ) %>% dplyr::select(-old_base, -peak_suffix, -new_coords)\n",
    "head(TSS_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "57df3903-e415-41db-a0c2-95e53d96f619",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(TSS_df, file = \"./HumanIsletTranscriptome/data_processed/TSS_info_w_lift.tsv\", sep = \"\\t\", quote = FALSE, row.names = FALSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.4",
   "language": "R",
   "name": "r44"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
