{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9cf98f48-d87b-402f-84b9-eb9e9c21261a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/jferrer/jmidgley/annotation_project/lib64/python3.9/site-packages/sorted_nearest/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import pyranges as pr\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import csv\n",
    "import re\n",
    "from itertools import combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1a15da3f-e258-4322-bd21-9fe350b374cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', None) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "befcf50c-1a6d-40f8-8302-72ea87d49dc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/19159811/ipykernel_493454/4232622582.py:1: DtypeWarning: Columns (3,6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  combined_df = pd.read_csv( 'jessies_gffcompare_v3.0.tsv', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "combined_df = pd.read_csv( 'jessies_gffcompare_with_TSS_name.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7874dd81-2940-4697-b5cc-caff5b07ea26",
   "metadata": {},
   "outputs": [],
   "source": [
    "unannotated_genes = combined_df[combined_df[\"Gene name\"].str.startswith(\"Peak\")].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3a52bb64-ec13-43e9-b928-54d1d6a041d3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>matched_ref_id</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>best_match_ref_id</th>\n",
       "      <th>largest_overlap_gencode_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>og_gene_id</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>closest_ref_id</th>\n",
       "      <th>TSS distance category</th>\n",
       "      <th>TSS_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336566</th>\n",
       "      <td>ENCT00000000219.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(1162877, 1163719), (1163993, 1164042)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>FTMG000031</td>\n",
       "      <td>chr1:1162789-1162942_Peak_111</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.018685</td>\n",
       "      <td>-5.741955</td>\n",
       "      <td>0.339126</td>\n",
       "      <td>0.055098</td>\n",
       "      <td>0.339126</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>1162789.0</td>\n",
       "      <td>1162942.0</td>\n",
       "      <td>4162.0</td>\n",
       "      <td>ENST00000384997.3</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P111_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336567</th>\n",
       "      <td>ENCT00000000276.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(1349559, 1355076), (1355263, 1355740), (1357214, 1358279)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_145</td>\n",
       "      <td>Peak_145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>FTMG000038</td>\n",
       "      <td>chr1:1349517-1349586_Peak_145</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.836916</td>\n",
       "      <td>-0.256846</td>\n",
       "      <td>0.867767</td>\n",
       "      <td>0.964448</td>\n",
       "      <td>0.867767</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>1349517.0</td>\n",
       "      <td>1349586.0</td>\n",
       "      <td>679.0</td>\n",
       "      <td>ENST00000806659.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P145_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336568</th>\n",
       "      <td>ENCT00000002194.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(17439849, 17449625)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_910</td>\n",
       "      <td>Peak_910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>FTMG000288</td>\n",
       "      <td>chr1:17439844-17439937_Peak_910</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.085408</td>\n",
       "      <td>-3.549480</td>\n",
       "      <td>0.085408</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.085408</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>17439844.0</td>\n",
       "      <td>17439937.0</td>\n",
       "      <td>20004.0</td>\n",
       "      <td>ENST00000835437.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P910_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336569</th>\n",
       "      <td>ENCT00000002302.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(19485741, 19489397)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_996</td>\n",
       "      <td>Peak_996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>FTMG000306</td>\n",
       "      <td>chr1:19485730-19485780_Peak_996</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.114061</td>\n",
       "      <td>-3.132117</td>\n",
       "      <td>0.168185</td>\n",
       "      <td>0.678191</td>\n",
       "      <td>0.168185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>19485730.0</td>\n",
       "      <td>19485780.0</td>\n",
       "      <td>1327.0</td>\n",
       "      <td>ENST00000648702.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P996_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336570</th>\n",
       "      <td>ENCT00000002410.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[(20508236, 20510041)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>FTMG000320</td>\n",
       "      <td>chr1:20508218-20508233_Peak_1039</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.059096</td>\n",
       "      <td>-4.080791</td>\n",
       "      <td>0.059096</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.059096</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>20508218.0</td>\n",
       "      <td>20508233.0</td>\n",
       "      <td>17756.0</td>\n",
       "      <td>ENST00000443195.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P1039_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339852</th>\n",
       "      <td>STRT02284050</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(74034396, 74034874)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_79446</td>\n",
       "      <td>Peak_79446</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>STRG270089</td>\n",
       "      <td>chrX:74034838-74034839_Peak_79446</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>1.715846</td>\n",
       "      <td>0.778920</td>\n",
       "      <td>1.715846</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.715846</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>74034838.0</td>\n",
       "      <td>74034839.0</td>\n",
       "      <td>21893.0</td>\n",
       "      <td>ENST00000638437.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P79446_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339853</th>\n",
       "      <td>STRT02289282</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(51805614, 51805945)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_79008</td>\n",
       "      <td>Peak_79008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>STRG264175</td>\n",
       "      <td>chrX:51805624-51805625_Peak_79008</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.318309</td>\n",
       "      <td>-1.651502</td>\n",
       "      <td>0.318309</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.318309</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>51805624.0</td>\n",
       "      <td>51805625.0</td>\n",
       "      <td>2617.0</td>\n",
       "      <td>ENST00000375772.7</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P79008_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339854</th>\n",
       "      <td>STRT02289408</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(57617393, 57618659)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_79202</td>\n",
       "      <td>Peak_79202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>STRG264636</td>\n",
       "      <td>chrX:57617355-57617356_Peak_79202</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.129846</td>\n",
       "      <td>-2.945122</td>\n",
       "      <td>0.129846</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.129846</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>57617355.0</td>\n",
       "      <td>57617356.0</td>\n",
       "      <td>23432.0</td>\n",
       "      <td>ENST00000818758.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P79202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339855</th>\n",
       "      <td>STRT02290336</td>\n",
       "      <td>chrX</td>\n",
       "      <td>[(19990302, 19990805)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_78445</td>\n",
       "      <td>Peak_78445</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>ENSG00000173681</td>\n",
       "      <td>chrX:19990802-19990828_Peak_78445</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.101023</td>\n",
       "      <td>-3.307240</td>\n",
       "      <td>0.260752</td>\n",
       "      <td>0.387430</td>\n",
       "      <td>0.260752</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>19990802.0</td>\n",
       "      <td>19990828.0</td>\n",
       "      <td>233.0</td>\n",
       "      <td>ENST00000379682.9</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P78445_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339856</th>\n",
       "      <td>PBT00442330</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[(18987729, 18991364)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG057570</td>\n",
       "      <td>chrY:18991177-18991443_Peak_80755</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.372451</td>\n",
       "      <td>-1.424877</td>\n",
       "      <td>0.372451</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.372451</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>18991177.0</td>\n",
       "      <td>18991443.0</td>\n",
       "      <td>1266.0</td>\n",
       "      <td>ENST00000382840.3</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P80755_U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3291 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id   chr  \\\n",
       "336566  ENCT00000000219.1  chr1   \n",
       "336567  ENCT00000000276.1  chr1   \n",
       "336568  ENCT00000002194.1  chr1   \n",
       "336569  ENCT00000002302.1  chr1   \n",
       "336570  ENCT00000002410.1  chr1   \n",
       "...                   ...   ...   \n",
       "339852       STRT02284050  chrX   \n",
       "339853       STRT02289282  chrX   \n",
       "339854       STRT02289408  chrX   \n",
       "339855       STRT02290336  chrX   \n",
       "339856        PBT00442330  chrY   \n",
       "\n",
       "                                                               exons  \\\n",
       "336566                      [(1162877, 1163719), (1163993, 1164042)]   \n",
       "336567  [(1349559, 1355076), (1355263, 1355740), (1357214, 1358279)]   \n",
       "336568                                        [(17439849, 17449625)]   \n",
       "336569                                        [(19485741, 19489397)]   \n",
       "336570                                        [(20508236, 20510041)]   \n",
       "...                                                              ...   \n",
       "339852                                        [(74034396, 74034874)]   \n",
       "339853                                        [(51805614, 51805945)]   \n",
       "339854                                        [(57617393, 57618659)]   \n",
       "339855                                        [(19990302, 19990805)]   \n",
       "339856                                        [(18987729, 18991364)]   \n",
       "\n",
       "       matched_to matched_ref_id  \\\n",
       "336566        NaN            NaN   \n",
       "336567        NaN            NaN   \n",
       "336568        NaN            NaN   \n",
       "336569        NaN            NaN   \n",
       "336570        NaN            NaN   \n",
       "...           ...            ...   \n",
       "339852        NaN            NaN   \n",
       "339853        NaN            NaN   \n",
       "339854        NaN            NaN   \n",
       "339855        NaN            NaN   \n",
       "339856        NaN            NaN   \n",
       "\n",
       "                                                           category  \\\n",
       "336566  Fully unannotated transcripts without any overlapping exons   \n",
       "336567  Fully unannotated transcripts without any overlapping exons   \n",
       "336568  Fully unannotated transcripts without any overlapping exons   \n",
       "336569  Fully unannotated transcripts without any overlapping exons   \n",
       "336570  Fully unannotated transcripts without any overlapping exons   \n",
       "...                                                             ...   \n",
       "339852  Fully unannotated transcripts without any overlapping exons   \n",
       "339853  Fully unannotated transcripts without any overlapping exons   \n",
       "339854  Fully unannotated transcripts without any overlapping exons   \n",
       "339855  Fully unannotated transcripts without any overlapping exons   \n",
       "339856  Fully unannotated transcripts without any overlapping exons   \n",
       "\n",
       "       unannotated_exons best_match_ref_id largest_overlap_gencode_id Strand  \\\n",
       "336566               NaN               NaN                        NaN      +   \n",
       "336567               NaN               NaN                        NaN      +   \n",
       "336568               NaN               NaN                        NaN      +   \n",
       "336569               NaN               NaN                        NaN      +   \n",
       "336570               NaN               NaN                        NaN      +   \n",
       "...                  ...               ...                        ...    ...   \n",
       "339852               NaN               NaN                        NaN      -   \n",
       "339853               NaN               NaN                        NaN      +   \n",
       "339854               NaN               NaN                        NaN      +   \n",
       "339855               NaN               NaN                        NaN      -   \n",
       "339856               NaN               NaN                        NaN      -   \n",
       "\n",
       "                                        summarized_category ref_id  \\\n",
       "336566  Unannotated transcripts without any annotated exons    NaN   \n",
       "336567  Unannotated transcripts without any annotated exons    NaN   \n",
       "336568  Unannotated transcripts without any annotated exons    NaN   \n",
       "336569  Unannotated transcripts without any annotated exons    NaN   \n",
       "336570  Unannotated transcripts without any annotated exons    NaN   \n",
       "...                                                     ...    ...   \n",
       "339852  Unannotated transcripts without any annotated exons    NaN   \n",
       "339853  Unannotated transcripts without any annotated exons    NaN   \n",
       "339854  Unannotated transcripts without any annotated exons    NaN   \n",
       "339855  Unannotated transcripts without any annotated exons    NaN   \n",
       "339856  Unannotated transcripts without any annotated exons    NaN   \n",
       "\n",
       "       stripped_ref_id Gene stable ID   Gene name Gene type gene_category  \\\n",
       "336566             NaN       Peak_111    Peak_111       NaN      unmapped   \n",
       "336567             NaN       Peak_145    Peak_145       NaN      unmapped   \n",
       "336568             NaN       Peak_910    Peak_910       NaN      unmapped   \n",
       "336569             NaN       Peak_996    Peak_996       NaN      unmapped   \n",
       "336570             NaN      Peak_1039   Peak_1039       NaN      unmapped   \n",
       "...                ...            ...         ...       ...           ...   \n",
       "339852             NaN     Peak_79446  Peak_79446       NaN      unmapped   \n",
       "339853             NaN     Peak_79008  Peak_79008       NaN      unmapped   \n",
       "339854             NaN     Peak_79202  Peak_79202       NaN      unmapped   \n",
       "339855             NaN     Peak_78445  Peak_78445       NaN      unmapped   \n",
       "339856             NaN     Peak_80755  Peak_80755       NaN      unmapped   \n",
       "\n",
       "             og_gene_id                          peak_name   peak_type  \\\n",
       "336566       FTMG000031      chr1:1162789-1162942_Peak_111      Robust   \n",
       "336567       FTMG000038      chr1:1349517-1349586_Peak_145  Permissive   \n",
       "336568       FTMG000288    chr1:17439844-17439937_Peak_910  Permissive   \n",
       "336569       FTMG000306    chr1:19485730-19485780_Peak_996  Permissive   \n",
       "336570       FTMG000320   chr1:20508218-20508233_Peak_1039  Permissive   \n",
       "...                 ...                                ...         ...   \n",
       "339852       STRG270089  chrX:74034838-74034839_Peak_79446  Permissive   \n",
       "339853       STRG264175  chrX:51805624-51805625_Peak_79008  Permissive   \n",
       "339854       STRG264636  chrX:57617355-57617356_Peak_79202  Permissive   \n",
       "339855  ENSG00000173681  chrX:19990802-19990828_Peak_78445  Permissive   \n",
       "339856        PBG057570  chrY:18991177-18991443_Peak_80755  Permissive   \n",
       "\n",
       "        mean_TPM_ref  log2_tx_exp  gene_exp_ref  relative_tx_exp  \\\n",
       "336566      0.018685    -5.741955      0.339126         0.055098   \n",
       "336567      0.836916    -0.256846      0.867767         0.964448   \n",
       "336568      0.085408    -3.549480      0.085408         1.000000   \n",
       "336569      0.114061    -3.132117      0.168185         0.678191   \n",
       "336570      0.059096    -4.080791      0.059096         1.000000   \n",
       "...              ...          ...           ...              ...   \n",
       "339852      1.715846     0.778920      1.715846         1.000000   \n",
       "339853      0.318309    -1.651502      0.318309         1.000000   \n",
       "339854      0.129846    -2.945122      0.129846         1.000000   \n",
       "339855      0.101023    -3.307240      0.260752         0.387430   \n",
       "339856      0.372451    -1.424877      0.372451         1.000000   \n",
       "\n",
       "        TSS_TPM_per_gene  Relative TSS usage    TSS_type   TSS_start  \\\n",
       "336566          0.339126                 1.0  unique TSS   1162789.0   \n",
       "336567          0.867767                 1.0  unique TSS   1349517.0   \n",
       "336568          0.085408                 1.0  unique TSS  17439844.0   \n",
       "336569          0.168185                 1.0  unique TSS  19485730.0   \n",
       "336570          0.059096                 1.0  unique TSS  20508218.0   \n",
       "...                  ...                 ...         ...         ...   \n",
       "339852          1.715846                 1.0  unique TSS  74034838.0   \n",
       "339853          0.318309                 1.0  unique TSS  51805624.0   \n",
       "339854          0.129846                 1.0  unique TSS  57617355.0   \n",
       "339855          0.260752                 1.0  unique TSS  19990802.0   \n",
       "339856          0.372451                 1.0  unique TSS  18991177.0   \n",
       "\n",
       "           TSS_end  distance_to_closest_TSS     closest_ref_id  \\\n",
       "336566   1162942.0                   4162.0  ENST00000384997.3   \n",
       "336567   1349586.0                    679.0  ENST00000806659.1   \n",
       "336568  17439937.0                  20004.0  ENST00000835437.1   \n",
       "336569  19485780.0                   1327.0  ENST00000648702.1   \n",
       "336570  20508233.0                  17756.0  ENST00000443195.1   \n",
       "...            ...                      ...                ...   \n",
       "339852  74034839.0                  21893.0  ENST00000638437.1   \n",
       "339853  51805625.0                   2617.0  ENST00000375772.7   \n",
       "339854  57617356.0                  23432.0  ENST00000818758.1   \n",
       "339855  19990828.0                    233.0  ENST00000379682.9   \n",
       "339856  18991443.0                   1266.0  ENST00000382840.3   \n",
       "\n",
       "       TSS distance category  TSS_name  \n",
       "336566                >100bp    P111_U  \n",
       "336567                >100bp    P145_U  \n",
       "336568                >100bp    P910_U  \n",
       "336569                >100bp    P996_U  \n",
       "336570                >100bp   P1039_U  \n",
       "...                      ...       ...  \n",
       "339852                >100bp  P79446_U  \n",
       "339853                >100bp  P79008_U  \n",
       "339854                >100bp  P79202_U  \n",
       "339855                >100bp  P78445_U  \n",
       "339856                >100bp  P80755_U  \n",
       "\n",
       "[3291 rows x 33 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unannotated_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "76bfe92c-33a9-4e4b-a5a2-4d2cc8565c47",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def get_tx_coords(exon_str):\n",
    "    # Find all numeric pairs\n",
    "    matches = re.findall(r\"\\((\\d+),\\s*(\\d+)\\)\", exon_str)\n",
    "    if not matches:\n",
    "        return (None, None)\n",
    "    starts = [int(s) for s, _ in matches]\n",
    "    ends = [int(e) for _, e in matches]\n",
    "    return (min(starts), max(ends))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9975770e-390b-47b8-8dc5-c3543dc17462",
   "metadata": {},
   "outputs": [],
   "source": [
    "unannotated_genes[[\"tx_start\", \"tx_end\"]] = unannotated_genes[\"exons\"].apply(\n",
    "    lambda x: pd.Series(get_tx_coords(x))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "91ca1d07-96bc-4018-8607-5c6b0384c8ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tx_start</th>\n",
       "      <th>tx_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336566</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>ENCT00000000219.1</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1164042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336567</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_145</td>\n",
       "      <td>ENCT00000000276.1</td>\n",
       "      <td>1349559</td>\n",
       "      <td>1358279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336568</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_910</td>\n",
       "      <td>ENCT00000002194.1</td>\n",
       "      <td>17439849</td>\n",
       "      <td>17449625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336569</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_996</td>\n",
       "      <td>ENCT00000002302.1</td>\n",
       "      <td>19485741</td>\n",
       "      <td>19489397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336570</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>ENCT00000002410.1</td>\n",
       "      <td>20508236</td>\n",
       "      <td>20510041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339852</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_79446</td>\n",
       "      <td>STRT02284050</td>\n",
       "      <td>74034396</td>\n",
       "      <td>74034874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339853</th>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_79008</td>\n",
       "      <td>STRT02289282</td>\n",
       "      <td>51805614</td>\n",
       "      <td>51805945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339854</th>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_79202</td>\n",
       "      <td>STRT02289408</td>\n",
       "      <td>57617393</td>\n",
       "      <td>57618659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339855</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_78445</td>\n",
       "      <td>STRT02290336</td>\n",
       "      <td>19990302</td>\n",
       "      <td>19990805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339856</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>PBT00442330</td>\n",
       "      <td>18987729</td>\n",
       "      <td>18991364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3291 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr Strand   Gene name      transcript_id  tx_start    tx_end\n",
       "336566  chr1      +    Peak_111  ENCT00000000219.1   1162877   1164042\n",
       "336567  chr1      +    Peak_145  ENCT00000000276.1   1349559   1358279\n",
       "336568  chr1      +    Peak_910  ENCT00000002194.1  17439849  17449625\n",
       "336569  chr1      +    Peak_996  ENCT00000002302.1  19485741  19489397\n",
       "336570  chr1      +   Peak_1039  ENCT00000002410.1  20508236  20510041\n",
       "...      ...    ...         ...                ...       ...       ...\n",
       "339852  chrX      -  Peak_79446       STRT02284050  74034396  74034874\n",
       "339853  chrX      +  Peak_79008       STRT02289282  51805614  51805945\n",
       "339854  chrX      +  Peak_79202       STRT02289408  57617393  57618659\n",
       "339855  chrX      -  Peak_78445       STRT02290336  19990302  19990805\n",
       "339856  chrY      -  Peak_80755        PBT00442330  18987729  18991364\n",
       "\n",
       "[3291 rows x 6 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_coords = unannotated_genes[[\"chr\", \"Strand\", \"Gene name\", \"transcript_id\", \"tx_start\", \"tx_end\"]]\n",
    "transcript_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6513ecf4-e670-48cc-b907-cd93c44fec7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1704"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_coords[\"Gene name\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2bc22a80-8dd9-4c96-9d89-9efe14e9868c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tx_start</th>\n",
       "      <th>tx_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336566</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>ENCT00000000219.1</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1164042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336614</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>HBMT00000000427.1</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1166499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr Strand Gene name      transcript_id  tx_start   tx_end\n",
       "336566  chr1      +  Peak_111  ENCT00000000219.1   1162877  1164042\n",
       "336614  chr1      +  Peak_111  HBMT00000000427.1   1162877  1166499"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_coords[transcript_coords[\"Gene name\"] == \"Peak_111\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a46d083d-9a5b-4dec-afd9-cd9b3c480fb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_coords = (\n",
    "    transcript_coords\n",
    "    .groupby([\"chr\", \"Strand\", \"Gene name\"], as_index=False)\n",
    "    .agg(min_start=(\"tx_start\", \"min\"), max_end=(\"tx_end\", \"max\"))\n",
    ")\n",
    "gene_coords = gene_coords.rename(columns={\"Gene name\": \"Gene_name\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8d6cd540-a2a4-46e1-ba03-5e705b72e4c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>min_start</th>\n",
       "      <th>max_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1039</td>\n",
       "      <td>20508236</td>\n",
       "      <td>20510041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_111</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1166499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1187</td>\n",
       "      <td>22342372</td>\n",
       "      <td>22353088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1234</td>\n",
       "      <td>23555118</td>\n",
       "      <td>23556240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>Peak_1238</td>\n",
       "      <td>23559643</td>\n",
       "      <td>23560425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1699</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80211</td>\n",
       "      <td>126472363</td>\n",
       "      <td>126472699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80214</td>\n",
       "      <td>129382617</td>\n",
       "      <td>129382957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1701</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80261</td>\n",
       "      <td>130401109</td>\n",
       "      <td>130401897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1702</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80530</td>\n",
       "      <td>153927441</td>\n",
       "      <td>153928228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1703</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>Peak_80755</td>\n",
       "      <td>18987729</td>\n",
       "      <td>18991364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1704 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr Strand   Gene_name  min_start    max_end\n",
       "0     chr1      +   Peak_1039   20508236   20510041\n",
       "1     chr1      +    Peak_111    1162877    1166499\n",
       "2     chr1      +   Peak_1187   22342372   22353088\n",
       "3     chr1      +   Peak_1234   23555118   23556240\n",
       "4     chr1      +   Peak_1238   23559643   23560425\n",
       "...    ...    ...         ...        ...        ...\n",
       "1699  chrX      -  Peak_80211  126472363  126472699\n",
       "1700  chrX      -  Peak_80214  129382617  129382957\n",
       "1701  chrX      -  Peak_80261  130401109  130401897\n",
       "1702  chrX      -  Peak_80530  153927441  153928228\n",
       "1703  chrY      -  Peak_80755   18987729   18991364\n",
       "\n",
       "[1704 rows x 5 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "45721d07-1e51-4602-9bca-13a7e73f92db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort first\n",
    "gene_coords = gene_coords.sort_values([\"chr\", \"Strand\", \"min_start\"]).reset_index(drop=True)\n",
    "\n",
    "# Compute running maximum end per chromosome/strand\n",
    "gene_coords[\"running_max_end\"] = (\n",
    "    gene_coords.groupby([\"chr\", \"Strand\"])[\"max_end\"]\n",
    "    .cummax()\n",
    ")\n",
    "\n",
    "# A new group starts when the current start > previous running max_end\n",
    "gene_coords[\"prev_running_max_end\"] = gene_coords.groupby([\"chr\", \"Strand\"])[\"running_max_end\"].shift()\n",
    "gene_coords[\"new_group\"] = (\n",
    "    (gene_coords[\"chr\"] != gene_coords[\"chr\"].shift()) |\n",
    "    (gene_coords[\"Strand\"] != gene_coords[\"Strand\"].shift()) |\n",
    "    (gene_coords[\"min_start\"] > gene_coords[\"prev_running_max_end\"])\n",
    ").astype(int)\n",
    "\n",
    "gene_coords[\"group_id\"] = gene_coords[\"new_group\"].cumsum()\n",
    "\n",
    "# Summarize groups\n",
    "groups = (\n",
    "    gene_coords.groupby([\"chr\", \"Strand\", \"group_id\"])\n",
    "    .agg(\n",
    "        group_start=(\"min_start\", \"min\"),\n",
    "        group_end=(\"max_end\", \"max\"),\n",
    "        Genes=(\"Gene_name\", list),\n",
    "        group_size=(\"Gene_name\", \"count\")\n",
    "    )\n",
    "    .reset_index()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "048b3512-9719-43ba-9d00-8e189144087c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>group_id</th>\n",
       "      <th>group_start</th>\n",
       "      <th>group_end</th>\n",
       "      <th>Genes</th>\n",
       "      <th>group_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1162877</td>\n",
       "      <td>1166499</td>\n",
       "      <td>[Peak_111]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>1349550</td>\n",
       "      <td>1358279</td>\n",
       "      <td>[Peak_145]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1375573</td>\n",
       "      <td>1378310</td>\n",
       "      <td>[Peak_150]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>2189680</td>\n",
       "      <td>2191740</td>\n",
       "      <td>[Peak_264]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>2199524</td>\n",
       "      <td>2202776</td>\n",
       "      <td>[Peak_274]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1603</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1604</td>\n",
       "      <td>126472363</td>\n",
       "      <td>126472699</td>\n",
       "      <td>[Peak_80211]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1604</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1605</td>\n",
       "      <td>129382617</td>\n",
       "      <td>129382957</td>\n",
       "      <td>[Peak_80214]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1605</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1606</td>\n",
       "      <td>130401109</td>\n",
       "      <td>130401897</td>\n",
       "      <td>[Peak_80261]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606</th>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>1607</td>\n",
       "      <td>153927441</td>\n",
       "      <td>153928228</td>\n",
       "      <td>[Peak_80530]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1607</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1608</td>\n",
       "      <td>18987729</td>\n",
       "      <td>18991364</td>\n",
       "      <td>[Peak_80755]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1608 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr Strand  group_id  group_start  group_end         Genes  group_size\n",
       "0     chr1      +         1      1162877    1166499    [Peak_111]           1\n",
       "1     chr1      +         2      1349550    1358279    [Peak_145]           1\n",
       "2     chr1      +         3      1375573    1378310    [Peak_150]           1\n",
       "3     chr1      +         4      2189680    2191740    [Peak_264]           1\n",
       "4     chr1      +         5      2199524    2202776    [Peak_274]           1\n",
       "...    ...    ...       ...          ...        ...           ...         ...\n",
       "1603  chrX      -      1604    126472363  126472699  [Peak_80211]           1\n",
       "1604  chrX      -      1605    129382617  129382957  [Peak_80214]           1\n",
       "1605  chrX      -      1606    130401109  130401897  [Peak_80261]           1\n",
       "1606  chrX      -      1607    153927441  153928228  [Peak_80530]           1\n",
       "1607  chrY      -      1608     18987729   18991364  [Peak_80755]           1\n",
       "\n",
       "[1608 rows x 7 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c4c0cf08-dedf-44a4-82e7-118b366d29e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>group_id</th>\n",
       "      <th>group_start</th>\n",
       "      <th>group_end</th>\n",
       "      <th>Genes</th>\n",
       "      <th>group_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>182</td>\n",
       "      <td>97772165</td>\n",
       "      <td>97842556</td>\n",
       "      <td>[Peak_10465, Peak_10467, Peak_10468, Peak_10469]</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr Strand  group_id  group_start  group_end  \\\n",
       "181  chr10      +       182     97772165   97842556   \n",
       "\n",
       "                                                Genes  group_size  \n",
       "181  [Peak_10465, Peak_10467, Peak_10468, Peak_10469]           4  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groups[groups[\"Genes\"].apply(lambda genes: \"Peak_10467\" in genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "773c963c-3dc8-4238-8dfe-f4f151903ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_combined_name(genes):\n",
    "    if len(genes) == 1:\n",
    "        return genes[0]\n",
    "    else:\n",
    "        # Extract numeric parts and sort\n",
    "        nums = sorted([int(g.split(\"_\")[1]) for g in genes])\n",
    "        return f\"Peak_{nums[0]}-{nums[-1]}\"\n",
    "\n",
    "# Apply to create new column\n",
    "groups[\"new_gene_name\"] = groups[\"Genes\"].apply(make_combined_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e6acba35-36ae-4021-bb55-744f5dbea70c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>group_id</th>\n",
       "      <th>group_start</th>\n",
       "      <th>group_end</th>\n",
       "      <th>Genes</th>\n",
       "      <th>group_size</th>\n",
       "      <th>new_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>chr8</td>\n",
       "      <td>+</td>\n",
       "      <td>1463</td>\n",
       "      <td>117244748</td>\n",
       "      <td>117311898</td>\n",
       "      <td>[Peak_74199, Peak_74200, Peak_74201, Peak_74202, Peak_74203]</td>\n",
       "      <td>5</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>567</td>\n",
       "      <td>57564434</td>\n",
       "      <td>57607224</td>\n",
       "      <td>[Peak_29184, Peak_29170, Peak_29174, Peak_29178, Peak_29179]</td>\n",
       "      <td>5</td>\n",
       "      <td>Peak_29170-29184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>1009</td>\n",
       "      <td>122184040</td>\n",
       "      <td>122253386</td>\n",
       "      <td>[Peak_54073, Peak_54086, Peak_54100, Peak_54102]</td>\n",
       "      <td>4</td>\n",
       "      <td>Peak_54073-54102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>182</td>\n",
       "      <td>97772165</td>\n",
       "      <td>97842556</td>\n",
       "      <td>[Peak_10465, Peak_10467, Peak_10468, Peak_10469]</td>\n",
       "      <td>4</td>\n",
       "      <td>Peak_10465-10469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>572</td>\n",
       "      <td>85445590</td>\n",
       "      <td>85465053</td>\n",
       "      <td>[Peak_30027, Peak_30030, Peak_30031]</td>\n",
       "      <td>3</td>\n",
       "      <td>Peak_30027-30031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>544</td>\n",
       "      <td>75266030</td>\n",
       "      <td>75269181</td>\n",
       "      <td>[Peak_29846]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>543</td>\n",
       "      <td>72221946</td>\n",
       "      <td>72228455</td>\n",
       "      <td>[Peak_29730]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>542</td>\n",
       "      <td>71809311</td>\n",
       "      <td>71813524</td>\n",
       "      <td>[Peak_29702]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>541</td>\n",
       "      <td>70946388</td>\n",
       "      <td>70948148</td>\n",
       "      <td>[Peak_29657]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_29657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>553</td>\n",
       "      <td>605155</td>\n",
       "      <td>611191</td>\n",
       "      <td>[Peak_27341]</td>\n",
       "      <td>1</td>\n",
       "      <td>Peak_27341</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1608 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr Strand  group_id  group_start  group_end  \\\n",
       "1462   chr8      +      1463    117244748  117311898   \n",
       "566   chr16      -       567     57564434   57607224   \n",
       "1008   chr3      +      1009    122184040  122253386   \n",
       "181   chr10      +       182     97772165   97842556   \n",
       "571   chr16      -       572     85445590   85465053   \n",
       "...     ...    ...       ...          ...        ...   \n",
       "543   chr16      +       544     75266030   75269181   \n",
       "542   chr16      +       543     72221946   72228455   \n",
       "541   chr16      +       542     71809311   71813524   \n",
       "540   chr16      +       541     70946388   70948148   \n",
       "552   chr16      -       553       605155     611191   \n",
       "\n",
       "                                                             Genes  \\\n",
       "1462  [Peak_74199, Peak_74200, Peak_74201, Peak_74202, Peak_74203]   \n",
       "566   [Peak_29184, Peak_29170, Peak_29174, Peak_29178, Peak_29179]   \n",
       "1008              [Peak_54073, Peak_54086, Peak_54100, Peak_54102]   \n",
       "181               [Peak_10465, Peak_10467, Peak_10468, Peak_10469]   \n",
       "571                           [Peak_30027, Peak_30030, Peak_30031]   \n",
       "...                                                            ...   \n",
       "543                                                   [Peak_29846]   \n",
       "542                                                   [Peak_29730]   \n",
       "541                                                   [Peak_29702]   \n",
       "540                                                   [Peak_29657]   \n",
       "552                                                   [Peak_27341]   \n",
       "\n",
       "      group_size     new_gene_name  \n",
       "1462           5  Peak_74199-74203  \n",
       "566            5  Peak_29170-29184  \n",
       "1008           4  Peak_54073-54102  \n",
       "181            4  Peak_10465-10469  \n",
       "571            3  Peak_30027-30031  \n",
       "...          ...               ...  \n",
       "543            1        Peak_29846  \n",
       "542            1        Peak_29730  \n",
       "541            1        Peak_29702  \n",
       "540            1        Peak_29657  \n",
       "552            1        Peak_27341  \n",
       "\n",
       "[1608 rows x 8 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groups.sort_values(by=\"group_size\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cf4731f7-a8c5-4446-931e-4cd4730f4c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_to_new = {}\n",
    "for _, row in groups.iterrows():\n",
    "    for gene in row[\"Genes\"]:\n",
    "        gene_to_new[gene] = row[\"new_gene_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "752dd97f-50be-4f39-bcbd-45b105d200d3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mask = combined_df[\"Gene name\"].str.startswith(\"Peak\")\n",
    "combined_df.loc[mask, \"Gene name\"] = combined_df.loc[mask, \"Gene name\"].map(gene_to_new)\n",
    "combined_df.loc[mask, \"Gene stable ID\"] = combined_df.loc[mask, \"Gene stable ID\"].map(gene_to_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "decba9ea-1417-4f6d-abcd-05ac08cfd003",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>exons</th>\n",
       "      <th>matched_to</th>\n",
       "      <th>matched_ref_id</th>\n",
       "      <th>category</th>\n",
       "      <th>unannotated_exons</th>\n",
       "      <th>best_match_ref_id</th>\n",
       "      <th>largest_overlap_gencode_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>summarized_category</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>stripped_ref_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>og_gene_id</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>mean_TPM_ref</th>\n",
       "      <th>log2_tx_exp</th>\n",
       "      <th>gene_exp_ref</th>\n",
       "      <th>relative_tx_exp</th>\n",
       "      <th>TSS_TPM_per_gene</th>\n",
       "      <th>Relative TSS usage</th>\n",
       "      <th>TSS_type</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>distance_to_closest_TSS</th>\n",
       "      <th>closest_ref_id</th>\n",
       "      <th>TSS distance category</th>\n",
       "      <th>TSS_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>339553</th>\n",
       "      <td>PBT00408642</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245448, 117245484), (117247967, 117253514)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.022827</td>\n",
       "      <td>-5.453122</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.022576</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339567</th>\n",
       "      <td>STRT02027798</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245646, 117246177), (117248158, 117248245), (117248515, 117248739), (117260709, 117260844)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.007559</td>\n",
       "      <td>-7.047650</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.007476</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339568</th>\n",
       "      <td>STRT02027799</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245646, 117246177), (117248158, 117248245), (117248518, 117248739), (117260709, 117260844)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.018302</td>\n",
       "      <td>-5.771879</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.018101</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339592</th>\n",
       "      <td>STRT02045458</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117244748, 117246177), (117248158, 117257680)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117244716-117244764_Peak_74199</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.026709</td>\n",
       "      <td>-5.226533</td>\n",
       "      <td>0.074189</td>\n",
       "      <td>0.360014</td>\n",
       "      <td>0.074189</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117244716.0</td>\n",
       "      <td>117244764.0</td>\n",
       "      <td>20453.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74199_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339597</th>\n",
       "      <td>STRT02049149</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245421, 117245484), (117247967, 117248127), (117260709, 117260807)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.039047</td>\n",
       "      <td>-4.678659</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.038618</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339604</th>\n",
       "      <td>STRT02054041</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117246105, 117246177), (117248158, 117248245), (117260709, 117260812)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117246051-117246099_Peak_74203</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.009646</td>\n",
       "      <td>-6.695865</td>\n",
       "      <td>0.009646</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.009646</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117246051.0</td>\n",
       "      <td>117246099.0</td>\n",
       "      <td>19118.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74203_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339609</th>\n",
       "      <td>STRT02057619</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245448, 117245484), (117247980, 117248127), (117248515, 117248739), (117260709, 117260803)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.186005</td>\n",
       "      <td>-2.426588</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.183962</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339617</th>\n",
       "      <td>STRT02064444</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117244801, 117246177), (117247967, 117257180)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117244716-117244764_Peak_74199</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.047480</td>\n",
       "      <td>-4.396547</td>\n",
       "      <td>0.074189</td>\n",
       "      <td>0.639986</td>\n",
       "      <td>0.074189</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117244716.0</td>\n",
       "      <td>117244764.0</td>\n",
       "      <td>20453.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74199_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339625</th>\n",
       "      <td>STRT02067854</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245673, 117248127), (117248515, 117248739), (117260709, 117263389)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.019747</td>\n",
       "      <td>-5.662213</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.019530</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339628</th>\n",
       "      <td>STRT02069586</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245448, 117245484), (117247967, 117248127), (117248515, 117248739), (117310148, 117310351), (117310977, 117311898)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.260957</td>\n",
       "      <td>-1.938116</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.258091</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339629</th>\n",
       "      <td>STRT02070645</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245448, 117245484), (117247967, 117248127), (117248515, 117248739), (117283846, 117284042), (117284647, 117284797)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.016646</td>\n",
       "      <td>-5.908663</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.016463</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339630</th>\n",
       "      <td>STRT02073316</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245448, 117245484), (117246054, 117246177), (117247967, 117248127), (117248515, 117248739), (117260709, 117261152)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.086065</td>\n",
       "      <td>-3.538433</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.085119</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339645</th>\n",
       "      <td>STRT02080482</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117244926, 117248127), (117248515, 117257681)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117244964-117245013_Peak_74200</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.243592</td>\n",
       "      <td>-2.037460</td>\n",
       "      <td>0.243592</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.243592</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117244964.0</td>\n",
       "      <td>117245013.0</td>\n",
       "      <td>20204.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74200_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339654</th>\n",
       "      <td>STRT02087941</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245421, 117245484), (117247967, 117248127), (117248515, 117248739), (117310148, 117310588)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.259964</td>\n",
       "      <td>-1.943616</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.257109</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339676</th>\n",
       "      <td>STRT02104949</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245421, 117245484), (117247967, 117248245), (117274345, 117282767)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245420-117245800_Peak_74202</td>\n",
       "      <td>Robust</td>\n",
       "      <td>0.093988</td>\n",
       "      <td>-3.411375</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>0.092956</td>\n",
       "      <td>1.011106</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245420.0</td>\n",
       "      <td>117245800.0</td>\n",
       "      <td>19417.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74202_U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339677</th>\n",
       "      <td>STRT02105127</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[(117245273, 117246177), (117247967, 117248127), (117248518, 117257184)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fully unannotated transcripts without any overlapping exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>Unannotated transcripts without any annotated exons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>Peak_74199-74203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>PBG050864</td>\n",
       "      <td>chr8:117245236-117245267_Peak_74201</td>\n",
       "      <td>Permissive</td>\n",
       "      <td>0.203012</td>\n",
       "      <td>-2.300360</td>\n",
       "      <td>0.203012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.203012</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unique TSS</td>\n",
       "      <td>117245236.0</td>\n",
       "      <td>117245267.0</td>\n",
       "      <td>19950.0</td>\n",
       "      <td>ENST00000754618.1</td>\n",
       "      <td>&gt;100bp</td>\n",
       "      <td>P74201_U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id   chr  \\\n",
       "339553   PBT00408642  chr8   \n",
       "339567  STRT02027798  chr8   \n",
       "339568  STRT02027799  chr8   \n",
       "339592  STRT02045458  chr8   \n",
       "339597  STRT02049149  chr8   \n",
       "339604  STRT02054041  chr8   \n",
       "339609  STRT02057619  chr8   \n",
       "339617  STRT02064444  chr8   \n",
       "339625  STRT02067854  chr8   \n",
       "339628  STRT02069586  chr8   \n",
       "339629  STRT02070645  chr8   \n",
       "339630  STRT02073316  chr8   \n",
       "339645  STRT02080482  chr8   \n",
       "339654  STRT02087941  chr8   \n",
       "339676  STRT02104949  chr8   \n",
       "339677  STRT02105127  chr8   \n",
       "\n",
       "                                                                                                                           exons  \\\n",
       "339553                                                                          [(117245448, 117245484), (117247967, 117253514)]   \n",
       "339567                          [(117245646, 117246177), (117248158, 117248245), (117248515, 117248739), (117260709, 117260844)]   \n",
       "339568                          [(117245646, 117246177), (117248158, 117248245), (117248518, 117248739), (117260709, 117260844)]   \n",
       "339592                                                                          [(117244748, 117246177), (117248158, 117257680)]   \n",
       "339597                                                  [(117245421, 117245484), (117247967, 117248127), (117260709, 117260807)]   \n",
       "339604                                                  [(117246105, 117246177), (117248158, 117248245), (117260709, 117260812)]   \n",
       "339609                          [(117245448, 117245484), (117247980, 117248127), (117248515, 117248739), (117260709, 117260803)]   \n",
       "339617                                                                          [(117244801, 117246177), (117247967, 117257180)]   \n",
       "339625                                                  [(117245673, 117248127), (117248515, 117248739), (117260709, 117263389)]   \n",
       "339628  [(117245448, 117245484), (117247967, 117248127), (117248515, 117248739), (117310148, 117310351), (117310977, 117311898)]   \n",
       "339629  [(117245448, 117245484), (117247967, 117248127), (117248515, 117248739), (117283846, 117284042), (117284647, 117284797)]   \n",
       "339630  [(117245448, 117245484), (117246054, 117246177), (117247967, 117248127), (117248515, 117248739), (117260709, 117261152)]   \n",
       "339645                                                                          [(117244926, 117248127), (117248515, 117257681)]   \n",
       "339654                          [(117245421, 117245484), (117247967, 117248127), (117248515, 117248739), (117310148, 117310588)]   \n",
       "339676                                                  [(117245421, 117245484), (117247967, 117248245), (117274345, 117282767)]   \n",
       "339677                                                  [(117245273, 117246177), (117247967, 117248127), (117248518, 117257184)]   \n",
       "\n",
       "       matched_to matched_ref_id  \\\n",
       "339553        NaN            NaN   \n",
       "339567        NaN            NaN   \n",
       "339568        NaN            NaN   \n",
       "339592        NaN            NaN   \n",
       "339597        NaN            NaN   \n",
       "339604        NaN            NaN   \n",
       "339609        NaN            NaN   \n",
       "339617        NaN            NaN   \n",
       "339625        NaN            NaN   \n",
       "339628        NaN            NaN   \n",
       "339629        NaN            NaN   \n",
       "339630        NaN            NaN   \n",
       "339645        NaN            NaN   \n",
       "339654        NaN            NaN   \n",
       "339676        NaN            NaN   \n",
       "339677        NaN            NaN   \n",
       "\n",
       "                                                           category  \\\n",
       "339553  Fully unannotated transcripts without any overlapping exons   \n",
       "339567  Fully unannotated transcripts without any overlapping exons   \n",
       "339568  Fully unannotated transcripts without any overlapping exons   \n",
       "339592  Fully unannotated transcripts without any overlapping exons   \n",
       "339597  Fully unannotated transcripts without any overlapping exons   \n",
       "339604  Fully unannotated transcripts without any overlapping exons   \n",
       "339609  Fully unannotated transcripts without any overlapping exons   \n",
       "339617  Fully unannotated transcripts without any overlapping exons   \n",
       "339625  Fully unannotated transcripts without any overlapping exons   \n",
       "339628  Fully unannotated transcripts without any overlapping exons   \n",
       "339629  Fully unannotated transcripts without any overlapping exons   \n",
       "339630  Fully unannotated transcripts without any overlapping exons   \n",
       "339645  Fully unannotated transcripts without any overlapping exons   \n",
       "339654  Fully unannotated transcripts without any overlapping exons   \n",
       "339676  Fully unannotated transcripts without any overlapping exons   \n",
       "339677  Fully unannotated transcripts without any overlapping exons   \n",
       "\n",
       "       unannotated_exons best_match_ref_id largest_overlap_gencode_id Strand  \\\n",
       "339553               NaN               NaN                        NaN      +   \n",
       "339567               NaN               NaN                        NaN      +   \n",
       "339568               NaN               NaN                        NaN      +   \n",
       "339592               NaN               NaN                        NaN      +   \n",
       "339597               NaN               NaN                        NaN      +   \n",
       "339604               NaN               NaN                        NaN      +   \n",
       "339609               NaN               NaN                        NaN      +   \n",
       "339617               NaN               NaN                        NaN      +   \n",
       "339625               NaN               NaN                        NaN      +   \n",
       "339628               NaN               NaN                        NaN      +   \n",
       "339629               NaN               NaN                        NaN      +   \n",
       "339630               NaN               NaN                        NaN      +   \n",
       "339645               NaN               NaN                        NaN      +   \n",
       "339654               NaN               NaN                        NaN      +   \n",
       "339676               NaN               NaN                        NaN      +   \n",
       "339677               NaN               NaN                        NaN      +   \n",
       "\n",
       "                                        summarized_category ref_id  \\\n",
       "339553  Unannotated transcripts without any annotated exons    NaN   \n",
       "339567  Unannotated transcripts without any annotated exons    NaN   \n",
       "339568  Unannotated transcripts without any annotated exons    NaN   \n",
       "339592  Unannotated transcripts without any annotated exons    NaN   \n",
       "339597  Unannotated transcripts without any annotated exons    NaN   \n",
       "339604  Unannotated transcripts without any annotated exons    NaN   \n",
       "339609  Unannotated transcripts without any annotated exons    NaN   \n",
       "339617  Unannotated transcripts without any annotated exons    NaN   \n",
       "339625  Unannotated transcripts without any annotated exons    NaN   \n",
       "339628  Unannotated transcripts without any annotated exons    NaN   \n",
       "339629  Unannotated transcripts without any annotated exons    NaN   \n",
       "339630  Unannotated transcripts without any annotated exons    NaN   \n",
       "339645  Unannotated transcripts without any annotated exons    NaN   \n",
       "339654  Unannotated transcripts without any annotated exons    NaN   \n",
       "339676  Unannotated transcripts without any annotated exons    NaN   \n",
       "339677  Unannotated transcripts without any annotated exons    NaN   \n",
       "\n",
       "       stripped_ref_id    Gene stable ID         Gene name Gene type  \\\n",
       "339553             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339567             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339568             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339592             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339597             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339604             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339609             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339617             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339625             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339628             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339629             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339630             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339645             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339654             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339676             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "339677             NaN  Peak_74199-74203  Peak_74199-74203       NaN   \n",
       "\n",
       "       gene_category og_gene_id                            peak_name  \\\n",
       "339553      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339567      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339568      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339592      unmapped  PBG050864  chr8:117244716-117244764_Peak_74199   \n",
       "339597      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339604      unmapped  PBG050864  chr8:117246051-117246099_Peak_74203   \n",
       "339609      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339617      unmapped  PBG050864  chr8:117244716-117244764_Peak_74199   \n",
       "339625      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339628      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339629      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339630      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339645      unmapped  PBG050864  chr8:117244964-117245013_Peak_74200   \n",
       "339654      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339676      unmapped  PBG050864  chr8:117245420-117245800_Peak_74202   \n",
       "339677      unmapped  PBG050864  chr8:117245236-117245267_Peak_74201   \n",
       "\n",
       "         peak_type  mean_TPM_ref  log2_tx_exp  gene_exp_ref  relative_tx_exp  \\\n",
       "339553      Robust      0.022827    -5.453122      1.011106         0.022576   \n",
       "339567      Robust      0.007559    -7.047650      1.011106         0.007476   \n",
       "339568      Robust      0.018302    -5.771879      1.011106         0.018101   \n",
       "339592  Permissive      0.026709    -5.226533      0.074189         0.360014   \n",
       "339597      Robust      0.039047    -4.678659      1.011106         0.038618   \n",
       "339604  Permissive      0.009646    -6.695865      0.009646         1.000000   \n",
       "339609      Robust      0.186005    -2.426588      1.011106         0.183962   \n",
       "339617  Permissive      0.047480    -4.396547      0.074189         0.639986   \n",
       "339625      Robust      0.019747    -5.662213      1.011106         0.019530   \n",
       "339628      Robust      0.260957    -1.938116      1.011106         0.258091   \n",
       "339629      Robust      0.016646    -5.908663      1.011106         0.016463   \n",
       "339630      Robust      0.086065    -3.538433      1.011106         0.085119   \n",
       "339645  Permissive      0.243592    -2.037460      0.243592         1.000000   \n",
       "339654      Robust      0.259964    -1.943616      1.011106         0.257109   \n",
       "339676      Robust      0.093988    -3.411375      1.011106         0.092956   \n",
       "339677  Permissive      0.203012    -2.300360      0.203012         1.000000   \n",
       "\n",
       "        TSS_TPM_per_gene  Relative TSS usage    TSS_type    TSS_start  \\\n",
       "339553          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339567          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339568          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339592          0.074189                 1.0  unique TSS  117244716.0   \n",
       "339597          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339604          0.009646                 1.0  unique TSS  117246051.0   \n",
       "339609          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339617          0.074189                 1.0  unique TSS  117244716.0   \n",
       "339625          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339628          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339629          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339630          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339645          0.243592                 1.0  unique TSS  117244964.0   \n",
       "339654          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339676          1.011106                 1.0  unique TSS  117245420.0   \n",
       "339677          0.203012                 1.0  unique TSS  117245236.0   \n",
       "\n",
       "            TSS_end  distance_to_closest_TSS     closest_ref_id  \\\n",
       "339553  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339567  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339568  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339592  117244764.0                  20453.0  ENST00000754618.1   \n",
       "339597  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339604  117246099.0                  19118.0  ENST00000754618.1   \n",
       "339609  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339617  117244764.0                  20453.0  ENST00000754618.1   \n",
       "339625  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339628  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339629  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339630  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339645  117245013.0                  20204.0  ENST00000754618.1   \n",
       "339654  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339676  117245800.0                  19417.0  ENST00000754618.1   \n",
       "339677  117245267.0                  19950.0  ENST00000754618.1   \n",
       "\n",
       "       TSS distance category  TSS_name  \n",
       "339553                >100bp  P74202_U  \n",
       "339567                >100bp  P74202_U  \n",
       "339568                >100bp  P74202_U  \n",
       "339592                >100bp  P74199_U  \n",
       "339597                >100bp  P74202_U  \n",
       "339604                >100bp  P74203_U  \n",
       "339609                >100bp  P74202_U  \n",
       "339617                >100bp  P74199_U  \n",
       "339625                >100bp  P74202_U  \n",
       "339628                >100bp  P74202_U  \n",
       "339629                >100bp  P74202_U  \n",
       "339630                >100bp  P74202_U  \n",
       "339645                >100bp  P74200_U  \n",
       "339654                >100bp  P74202_U  \n",
       "339676                >100bp  P74202_U  \n",
       "339677                >100bp  P74201_U  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[combined_df[\"Gene stable ID\"] == \"Peak_74199-74203\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a570770b-f5db-48d7-aa74-25bef8afeed0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22015"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"Gene stable ID\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "69cced52-b1cd-4dd0-aa69-6747ce789bcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combined_df.to_csv('jessies_gffcompare_v3.0.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "25836ab5-574a-4f31-9631-6a849aeae9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert to GTF format\n",
    "import ast\n",
    "\n",
    "# Convert stringified exon lists to actual lists of tuples\n",
    "def parse_exons(exon_str):\n",
    "    if isinstance(exon_str, str):\n",
    "        try:\n",
    "            return ast.literal_eval(exon_str)\n",
    "        except:\n",
    "            return None \n",
    "    return exon_str  # already valid\n",
    "\n",
    "combined_df[\"exons\"] = combined_df[\"exons\"].apply(parse_exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5ba05794-50a6-467f-b1bb-e01e835cb95d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_rows = []\n",
    "\n",
    "for _, row in combined_df.iterrows():\n",
    "    transcript_id = row[\"transcript_id\"]\n",
    "    gene_id = row.get(\"Gene stable ID\", \".\")\n",
    "    gene_name = row.get(\"Gene name\", \".\")\n",
    "    gene_type = row.get(\"Gene type\", \".\")\n",
    "    chromosome = row[\"chr\"]\n",
    "    strand = row[\"Strand\"]\n",
    "    source = \"custom\"\n",
    "    score = \".\"\n",
    "    frame = \".\"\n",
    "\n",
    "    exons = row.get(\"exons\")\n",
    "\n",
    "    # Skip missing or malformed exon lists\n",
    "    if not isinstance(exons, list) or len(exons) == 0:\n",
    "        continue\n",
    "\n",
    "    for i, (start, end) in enumerate(exons, 1):\n",
    "        if not isinstance((start, end), tuple) or len((start, end)) != 2:\n",
    "            continue\n",
    "\n",
    "        attributes = (\n",
    "            f'gene_id \"{gene_id}\"; '\n",
    "            f'transcript_id \"{transcript_id}\"; '\n",
    "            f'gene_name \"{gene_name}\"; '\n",
    "            f'gene_type \"{gene_type}\"; '\n",
    "            #f'exon_number \"{i}\";'\n",
    "        )\n",
    "\n",
    "        gtf_rows.append([\n",
    "            chromosome,\n",
    "            source,\n",
    "            \"exon\",\n",
    "            start,\n",
    "            end,\n",
    "            score,\n",
    "            strand,\n",
    "            frame,\n",
    "            attributes\n",
    "        ])\n",
    "\n",
    "# Convert to DataFrame\n",
    "gtf_df = pd.DataFrame(gtf_rows, columns=[\n",
    "    \"chromosome\", \"source\", \"feature\", \"start\", \"end\",\n",
    "    \"score\", \"strand\", \"frame\", \"attribute\"\n",
    "])\n",
    "gtf_df\n",
    "\n",
    "\n",
    "# Write to file\n",
    "#gtf_df.to_csv(\"/no_backup/jferrer/jmidgley/VASAseq/jessies_HIT_v3.0.gtf\", sep=\"\\t\", index=False, header=False, quoting=csv.QUOTE_NONE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "14e84509-fea8-422e-8b74-efc1987971a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376342"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_column_names = ['chromosome','source','feature','start','end','score','strand','frame','attribute']\n",
    "gtf_df = pd.read_csv( '/no_backup/jferrer/jmidgley/VASAseq/jessies_HIT_v3.0.gtf', sep='\\t', header=None, names=gtf_column_names)\n",
    "\n",
    "gtf_df[\"attribute\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0088c595-50f2-4284-aea1-143b57e014dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>61077274</td>\n",
       "      <td>61077628</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>61077916</td>\n",
       "      <td>61079203</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>160261734</td>\n",
       "      <td>160261922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>160281430</td>\n",
       "      <td>160281892</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>239386568</td>\n",
       "      <td>239387227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000133019\"; transcript_id \"ENCT00000019620.1\"; gene_name \"CHRM3\"; gene_type \"protein_coding\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  source feature      start        end score strand frame  \\\n",
       "0       chr1  custom    exon   61077274   61077628     .      +     .   \n",
       "1       chr1  custom    exon   61077916   61079203     .      +     .   \n",
       "2       chr1  custom    exon  160261734  160261922     .      +     .   \n",
       "3       chr1  custom    exon  160281430  160281892     .      +     .   \n",
       "4       chr1  custom    exon  239386568  239387227     .      +     .   \n",
       "\n",
       "                                                                                                       attribute  \n",
       "0   gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";   \n",
       "1   gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";   \n",
       "2       gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";   \n",
       "3       gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";   \n",
       "4  gene_id \"ENSG00000133019\"; transcript_id \"ENCT00000019620.1\"; gene_name \"CHRM3\"; gene_type \"protein_coding\";   "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c68a1f8c-6eb0-4c1e-ad0e-68c83eeb8348",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376342"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_column_names = ['chromosome','source','feature','start','end','score','strand','frame','attribute']\n",
    "gtf_df = pd.read_csv( '/no_backup/jferrer/jmidgley/VASAseq/jessies_HIT_v3.0.gtf', sep='\\t', header=None, names=gtf_column_names)\n",
    "\n",
    "gtf_df[\"attribute\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a4ad5391-42df-4782-8ff2-d2e451ea2e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_attributes(df, *attribute_type):\n",
    "    \"\"\"split attribute of a gtf into separate columns\"\"\"\n",
    "    try: \n",
    "        l_attributes = df.strip(';').split('; ')\n",
    "        attribute_position = [i for i, x in enumerate(l_attributes) if x.startswith(attribute_type)]\n",
    "        attribute = l_attributes[attribute_position[0]].split(' ')[1].strip('\"')\n",
    "        return attribute\n",
    "    except Exception:\n",
    "        return np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fca844da-1ba0-4759-bf9e-b1c824eee8ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>61077274</td>\n",
       "      <td>61077628</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";</td>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>61077916</td>\n",
       "      <td>61079203</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";</td>\n",
       "      <td>ENCT00000006334.1</td>\n",
       "      <td>ENSG00000162599</td>\n",
       "      <td>NFIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>160261734</td>\n",
       "      <td>160261922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";</td>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>160281430</td>\n",
       "      <td>160281892</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";</td>\n",
       "      <td>ENCT00000013056.1</td>\n",
       "      <td>ENSG00000228606</td>\n",
       "      <td>DCAF8-DT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>custom</td>\n",
       "      <td>exon</td>\n",
       "      <td>239386568</td>\n",
       "      <td>239387227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000133019\"; transcript_id \"ENCT00000019620.1\"; gene_name \"CHRM3\"; gene_type \"protein_coding\";</td>\n",
       "      <td>ENCT00000019620.1</td>\n",
       "      <td>ENSG00000133019</td>\n",
       "      <td>CHRM3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  source feature      start        end score strand frame  \\\n",
       "0       chr1  custom    exon   61077274   61077628     .      +     .   \n",
       "1       chr1  custom    exon   61077916   61079203     .      +     .   \n",
       "2       chr1  custom    exon  160261734  160261922     .      +     .   \n",
       "3       chr1  custom    exon  160281430  160281892     .      +     .   \n",
       "4       chr1  custom    exon  239386568  239387227     .      +     .   \n",
       "\n",
       "                                                                                                       attribute  \\\n",
       "0   gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";    \n",
       "1   gene_id \"ENSG00000162599\"; transcript_id \"ENCT00000006334.1\"; gene_name \"NFIA\"; gene_type \"protein_coding\";    \n",
       "2       gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";    \n",
       "3       gene_id \"ENSG00000228606\"; transcript_id \"ENCT00000013056.1\"; gene_name \"DCAF8-DT\"; gene_type \"lncRNA\";    \n",
       "4  gene_id \"ENSG00000133019\"; transcript_id \"ENCT00000019620.1\"; gene_name \"CHRM3\"; gene_type \"protein_coding\";    \n",
       "\n",
       "       transcript_id          gene_id gene_name  \n",
       "0  ENCT00000006334.1  ENSG00000162599      NFIA  \n",
       "1  ENCT00000006334.1  ENSG00000162599      NFIA  \n",
       "2  ENCT00000013056.1  ENSG00000228606  DCAF8-DT  \n",
       "3  ENCT00000013056.1  ENSG00000228606  DCAF8-DT  \n",
       "4  ENCT00000019620.1  ENSG00000133019     CHRM3  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.loc[:, 'transcript_id'] = gtf_df['attribute'].apply(split_attributes, args=(\"transcript_id\",))\n",
    "gtf_df.loc[:, 'gene_id'] = gtf_df['attribute'].apply(split_attributes, args=(\"gene_id\",))\n",
    "gtf_df.loc[:, 'gene_name'] = gtf_df['attribute'].apply(split_attributes, args=(\"gene_name\",))\n",
    "gtf_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b7268270-0751-453b-ad22-a5feeedc2f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gtf_df.to_csv('jessies_HIT_v3.0.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dda3292-41a6-4be9-b0e6-fb324ccee108",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (annotation_project)",
   "language": "python",
   "name": "annotation_project"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
